---
title: "ar43-setup-spieceasi"
author: "Nina Yang, Postdoctoral Investigator, WHOI"
created date: "2023-02-04"
last updated: "2026-01-12"
output: html_document
editor_options: 
  chunk_output_type: console
---

# 1. Background
This document outlines the steps taken to set up files for network analyses with the package Sparse InversE Covariance Estimation for Ecological Association inference (SPIEC-EASI). Networks were generated on WHOI's HPC. The samples was collected via CTD from the AR43 Cruise (Armstrong, March 2020) and multi-marker metabarcoding with 18S V9 and 12S MiFish were conducted.

## 1.1 Set up Environment

### 1.1.1 Install and load packages
```{r install-pkgs, include=FALSE}

if(!require(BiocManager)){install.packages("BiocManager")}
if(!require(devtools)){install.packages("devtools")}
if(!require(pairwiseQdonis)){install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")}
if(!require(phyloseq)){install.packages("phyloseq")}
if(!require(patchwork)){install.packages("patchwork")}
if(!require(tidyverse)){install.packages("tidyverse")}
if(!require(ARTool)){install_github("ARTool")}
if(!require(ggpubr)){install_github("ggpubr")}
if(!require(phyloseqCompanion)){install_github("phyloseqCompanion")}
if(!require(scales)){install.packages("scales")}
if(!require(SpiecEasi)){install_github("zdk123/SpiecEasi")}

```

```{r load-pkgs, include=FALSE}
# load packages
library(devtools); packageVersion("devtools")
library(phyloseq); packageVersion("phyloseq")
library(tidyverse); packageVersion("tidyverse")
library(patchwork); packageVersion("patchwork")
library(SpiecEasi); packageVersion("SpiecEasi")
library(ComplexHeatmap); packageVersion("ComplexHeatmap")
library(grid); packageVersion("grid")
library(gridBase); packageVersion("gridBase")

```

### 1.1.2 Setup Environment

```{r fig-setup, include = FALSE}
# Chunk options
knitr::opts_chunk$set(
 fig.width = 12,
 fig.asp = 0.5,
 out.width = "80%"
)

```

```{r theme_ggplot2, echo = FALSE}
theme_custom_ppt <- function() {
  theme_bw(
    base_size = 14) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent'),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      legend.background = element_rect(color = NA, fill='transparent'), 
      legend.box.background = element_rect(color = NA, fill='transparent'),
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = 'transparent', linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 12)
      )
}
theme_custom_poster <- function() {
  theme_bw(
    base_size = 30) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(), 
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = "white", linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 30)
      )
}
theme_custom_paper <- function() {
  theme_bw(
    base_size = 12) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent'),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      legend.background = element_rect(color = NA, fill='transparent'), 
      legend.box.background = element_rect(color = NA, fill='transparent'),
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = 'transparent', linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 10)
      )
}

# Changing the default theme
theme_set(theme_custom_paper())

```

# 2. Setup SPIEC-EASI - Aggregated - Prevalence-filtered (Core)

Trimmed unwanted ASVs (likely contaminants) and also ASVs with less than 50 reads (detection).
Trimmed rare reads based on prevalence

```{r import-asvs, eval = TRUE, message = FALSE, warning = FALSE }

load(file = "../03_diversity/updated_metadata_fordiversityanalyses.RData", verbose = TRUE)
load(file = "../03_diversity/psmelt_tables_for_networks.RData", verbose = TRUE)

# I want prot_taxa, met_taxa_order2, and fish_taxa_order
# let's make new phyloseq objects

# set up ASVs table
prot_asvs <- prot_taxa %>% select(OTU, Abundance, sample) %>% 
  pivot_wider(names_from = sample, values_from = Abundance, values_fill = 0)

head(prot_asvs)
prot_asvs_physeq <- as.matrix(prot_asvs[,-1])  # exclude OTU column
rownames(prot_asvs_physeq) <- prot_asvs$OTU
head(prot_asvs_physeq)
dim(prot_asvs_physeq) # 7046 x 69 (as expected)

met_asvs <- met_taxa_order2 %>% select(OTU, Abundance, sample) %>% 
  pivot_wider(names_from = sample, values_from = Abundance, values_fill = 0)

head(met_asvs)
met_asvs_physeq <- as.matrix(met_asvs[,-1])  # exclude OTU column
rownames(met_asvs_physeq) <- met_asvs$OTU
head(met_asvs_physeq)
dim(met_asvs_physeq) # 399 x 70 (as expected)

fish_asvs <- fish_taxa_order %>% select(OTU, Abundance, sample) %>% 
  pivot_wider(names_from = sample, values_from = Abundance, values_fill = 0)

head(fish_asvs)
fish_asvs_physeq <- as.matrix(fish_asvs[,-1])  # exclude OTU column
rownames(fish_asvs_physeq) <- fish_asvs$OTU
head(fish_asvs_physeq)
dim(fish_asvs_physeq) # 677 x 61 (as expected)

## load metadata

head(metadata_18s)
#rownames(metadata_18s) <- metadata_18s$sample

head(metadata_12s)
#rownames(metadata_12s) <- metadata_12s$sample

## taxonomy
tax_physeq_prot <- prot_taxa %>% select(OTU, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAXA) %>% unique()
dim(tax_physeq_prot)  #7046 x 11
head(tax_physeq_prot)
tax_table_prot <- as.matrix(tax_physeq_prot[,-1])
head(tax_table_prot)
rownames(tax_table_prot) <- tax_physeq_prot$OTU
head(tax_table_prot)

tax_physeq_met <- met_taxa_order2 %>% select(OTU, Domain, Kingdom, Phylum, Class, Order, Family, Genus, Species, TAXA) %>% unique() 
dim(tax_physeq_met) # 399 x 10
head(tax_physeq_met)
tax_table_met <- as.matrix(tax_physeq_met[,-1])
head(tax_table_met)
rownames(tax_table_met) <- tax_physeq_met$OTU
head(tax_table_met)

tax_physeq_fish <- fish_taxa_order %>% select(OTU, Domain, Phylum, Class, Order, Family, Genus, Species, TAXA) %>% unique() 
dim(tax_physeq_fish) # 667 x 9
head(tax_physeq_fish)
tax_table_fish <- as.matrix(tax_physeq_fish[,-1])
head(tax_table_fish)
rownames(tax_table_fish) <- tax_physeq_fish$OTU
head(tax_table_fish)
tail(tax_table)

# generate phyloseq packages
physeq_prot_network <- phyloseq(
  otu_table(prot_asvs_physeq, taxa_are_rows = T),
  tax_table(tax_table_prot),
  sample_data(metadata_18s)
)

physeq_met_network <- phyloseq(
  otu_table(met_asvs_physeq, taxa_are_rows = T),
  tax_table(tax_table_met),
  sample_data(metadata_18s)
)

physeq_fish_network <- phyloseq(
  otu_table(fish_asvs_physeq, taxa_are_rows = T),
  tax_table(tax_table_fish),
  sample_data(metadata_12s)
)

# check phyloseq objects
physeq_prot_network
physeq_met_network
physeq_fish_network

```

```{r prev-prot, eval = TRUE, message = FALSE, warning = FALSE}

# let's look at prevalence of taxa. Do I need a treshhold to remove anything?
prev_prot_all <- microbiome::prevalence(
  physeq_prot_network,
  detection = 0, # Treat any non-zero abundance as presence
  sort = TRUE, # The taxa will be sorted by how prevalent they are (most prevalent first)
  count = TRUE, # The function will return the number of samples where each taxon was detected
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(physeq_prot_network, "Class", detection = 0) + theme(legend.position = "none")

physeq_prot_core <- prune_taxa(taxa_sums(physeq_prot_network) > 50, physeq_prot_network)
physeq_prot_core # 2789 & 69 samples

# ok let's use prevalent ASVs at prevalence > 10%
physeq_prot_core_prev <- microbiome::core(physeq_prot_core, detection = 0, prevalence = 0.1)
physeq_prot_core_prev # reduces to 1977 ASVs with 10%

melt_prot <- psmelt(physeq_prot_core_prev)
head(melt_prot)
unique(melt_prot$TAXA)
melt_prot_tax <- melt_prot %>% 
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "prot") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, sep = ";"))

head(melt_prot_tax)

head(melt_prot_tax)
(unique(melt_prot_tax$sample)) # 69 samples
unique(melt_prot_tax$TAX_SHORT) 

```

```{r prev-met, eval = TRUE, message = FALSE, warning = FALSE}


# let's look at prevalence of taxa. Do I need a treshhold to remove anything?
prev_met_all <- microbiome::prevalence(
  physeq_met_network,
  detection = 0, # Treat any non-zero abundance as presence
  sort = TRUE, # The taxa will be sorted by how prevalent they are (most prevalent first)
  count = TRUE, # The function will return the number of samples where each taxon was detected
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(physeq_met_network, "Class", detection = 0) + theme(legend.position = "none")

physeq_met_core <- prune_taxa(taxa_sums(physeq_met_network) > 50, physeq_met_network)
physeq_met_core # 167 & 70 samples

# ok let's use prevalent ASVs at prevalence > 10%
physeq_met_core_prev <- microbiome::core(physeq_met_core, detection = 0, prevalence = 0.1)
physeq_met_core_prev # reduces to 66 ASVs with 10%

melt_met <- psmelt(physeq_met_core_prev)
head(melt_met)
unique(melt_met$TAXA)
melt_met_tax <- melt_met %>% 
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Kingdom, Phylum, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "met") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Kingdom, Phylum, Class, Order, Family, Genus, Species, sep = ";"))

head(melt_met_tax)

head(melt_met_tax)
(unique(melt_met_tax$sample)) # 70 samples
unique(melt_met_tax$TAX_SHORT) 

```

```{r prev-fish, eval = TRUE, message = FALSE, warning = FALSE}

# let's look at prevalence of taxa. Do I need a treshhold to remove anything?
prev_fish_all <- microbiome::prevalence(
  physeq_fish_network,
  detection = 0, # Treat any non-zero abundance as presence
  sort = TRUE, # The taxa will be sorted by how prevalent they are (most prevalent first)
  count = TRUE, # The function will return the number of samples where each taxon was detected
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(physeq_fish_network, "Family", detection = 0) + theme(legend.position = "none")

physeq_fish_core <- prune_taxa(taxa_sums(physeq_fish_network) > 50, physeq_fish_network)
physeq_fish_core # 396 ASVs & 61 samples

# ok let's use prevalent ASVs at prevalence > 10%
physeq_fish_core_prev <- microbiome::core(physeq_fish_core, detection = 0, prevalence = 0.1)
physeq_fish_core_prev # reduces to 18 ASVs with 10%

melt_fish <- psmelt(physeq_fish_core_prev)
head(melt_fish)
unique(melt_fish$TAXA)
melt_fish_tax <- melt_fish %>% 
  mutate(
    TAXA = ifelse(
      !(TAXA %in% c("Unassigned Actinopterygii", "Unassigned Perciformes")),
      paste(Family, Genus, sep = "-"), TAXA)) %>%
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Phylum, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "fish") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Phylum, Class, Order, Family, Genus, Species, sep = ";"))
    
head(melt_fish_tax)
tail(melt_fish_tax)
(unique(melt_fish_tax$sample)) # 61 samples
unique(melt_fish_tax$TAX_SHORT) 

```

```{r save-asv-rdata, eval = TRUE, message = FALSE, warning = FALSE}

# save phyloseq objects (not pruned for rare reads or prevalence-filtered)
save(physeq_prot_network, physeq_met_network, physeq_fish_network, file = "./R-outputs/physeq_taxa_network_20250410.RData")

# save phyloseq objects (pruned for rare reads but not prevalence-filtered, )
save(physeq_prot_core, physeq_met_core, physeq_fish_core, file = "./R-outputs/physeq_taxa_network_core_20250410.RData")

# Save rarefied phyloseq objects
save(physeq_prot_core_prev, physeq_met_core_prev, physeq_fish_core_prev, metadata_18s, metadata_12s, file = "./R-outputs/physeq_core_network_10perc_20250410.RData")

load("./R-outputs/physeq_core_network_10perc_20250410.RData", verbose = T)

# Save rarefied and prevalence filtered phyloseq objects (to be used if we decide to combine all depths)
physeq_prot_core_prev # 1977, 69
physeq_met_core_prev # 66, 70
physeq_fish_core_prev # 18, 61 

```

# 3. Setup SPIEC-EASI - Depths 

## 3.1. Setup prevalence-filtered, depth-specific inputs

```{r setup-tax-tables, eval = TRUE, message = FALSE, warning = FALSE}

melt_prot <- psmelt(physeq_prot_core)
head(melt_prot)
unique(melt_prot$TAXA)
melt_prot_tax <- melt_prot %>% 
    mutate(TAXA =
    case_when(
      Supergroup == "TSAR" & is.na(Division) ~ "TSAR-Unassigned",
      Supergroup == "Alveolata" & is.na(Division) ~ "Alveolata-Unassigned",
      TRUE ~ TAXA)) %>%
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "prot") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, sep = ";"))

head(melt_prot_tax)

(unique(melt_prot_tax$sample)) # 69 samples
unique(melt_prot_tax$TAX_SHORT) 

melt_met <- psmelt(physeq_met_core)
head(melt_met)
unique(melt_met$TAXA)
melt_met_tax <- melt_met %>% 
  mutate(TAXA =
    case_when(
      Phylum == "Cnidaria" & Order == "Unassigned" ~ "Cnidaria-Unassigned",
      Phylum == "Ctenophora" & Order == "Unassigned" ~ "Ctenophora-Unassigned",
      Phylum == "Arthropoda" & Order == "Unassigned" ~ "Arthropoda-Unassigned",
      TRUE ~ TAXA)) %>%
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Kingdom, Phylum, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "met") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Phylum, Class, Order, Family, Genus, Species, sep = ";"))

head(melt_met_tax)
(unique(melt_met_tax$sample)) # 70 samples
unique(melt_met_tax$TAX_SHORT) 

melt_fish <- psmelt(physeq_fish_core)
head(melt_fish)
unique(melt_fish$TAXA)
melt_fish_tax <- melt_fish %>% 
   mutate(
    TAXA = ifelse(
      !(TAXA %in% c("Unassigned Actinopterygii", "Unassigned Perciformes")),
      paste(Family, Genus, sep = "-"), TAXA)) %>%
  select(OTU, Abundance, sample, depth, depth_group, 
         Domain, Phylum, Class, Order, Family, Genus, Species, TAX_SHORT = TAXA) %>% # select columns to keep
  mutate(taxon = "fish") %>%
  mutate(OTUid = paste(taxon, OTU, sep = "_")) %>% # add a new OTU column to more easily identify what OTU it is
  mutate(TAX_full = paste(Domain, Phylum, Class, Order, Family, Genus, Species, sep = ";"))

head(melt_fish_tax)

head(melt_fish_tax)
(unique(melt_fish_tax$sample)) # 61 samples
unique(melt_fish_tax$TAX_SHORT) 

# these are the asv files with taxa, rarefied (not prevalence-filtered)
# Save multiple objects
save(melt_prot_tax, melt_met_tax, melt_fish_tax, file = "./R-outputs/alleuk_asv_melt_tax_tables_network.RData")

```

```{r make-asv-depths-input, eval = TRUE, message = FALSE, warning = FALSE}

df1 <- melt_prot_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "0-200")
df2 <- melt_met_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "0-200")
df3 <- melt_fish_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "0-200") 

# inner.join
s.list <- inner_join(unique(df1), unique(df2))
s.list <- inner_join(s.list, unique(df3))
dim(s.list) # retain 15 samples (reminder: expected is 24)

pru_sample <- s.list$sample

prot_asv_se_epi <- melt_prot_tax %>%
  select(OTUid, Abundance, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "0-200") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(prot_asv_se_epi)
dim(prot_asv_se_epi) # 2789 x 27

met_asv_se_epi <- melt_met_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "0-200") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(met_asv_se_epi)
dim(met_asv_se_epi) # 167 x 24

fish_asv_se_epi <- melt_fish_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "0-200") %>% # epipelagic samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(fish_asv_se_epi)
dim(fish_asv_se_epi) # 396 x 24

save(prot_asv_se_epi, met_asv_se_epi, fish_asv_se_epi, file = "./R-outputs/alleuk_asv_se_epi.RData")

df1 <- melt_prot_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "300-500")
df2 <- melt_met_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "300-500")
df3 <- melt_fish_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "300-500")

# inner.join
s.list <- inner_join(unique(df1), unique(df2))
s.list <- inner_join(s.list, unique(df3))
dim(s.list) # retain 21 samples (expected is 24)

pru_sample <- s.list$sample

prot_asv_se_meso_up <- melt_prot_tax %>%
  select(OTUid, Abundance, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "300-500") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(prot_asv_se_meso_up)
dim(prot_asv_se_meso_up)

met_asv_se_meso_up <- melt_met_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "300-500") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(met_asv_se_meso_up)
dim(met_asv_se_meso_up)

fish_asv_se_meso_up <- melt_fish_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "300-500") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(fish_asv_se_meso_up)
dim(fish_asv_se_meso_up)

save(prot_asv_se_meso_up, met_asv_se_meso_up, fish_asv_se_meso_up, file = "./R-outputs/alleuk_asv_se_meso_up.RData")

df1 <- melt_prot_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "800-1000")
df2 <- melt_met_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "800-1000")
df3 <- melt_fish_tax %>% select(sample, depth, depth_group) %>% filter(depth_group == "800-1000")

# inner.join
s.list <- inner_join(unique(df1), unique(df2))
s.list <- inner_join(s.list, unique(df3))
dim(s.list) # retain 22 samples (expected is 24)

pru_sample <- s.list$sample

prot_asv_se_meso_low <- melt_prot_tax %>%
  select(OTUid, Abundance, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "800-1000") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(prot_asv_se_meso_low)
dim(prot_asv_se_meso_low)

met_asv_se_meso_low <- melt_met_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "800-1000") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(met_asv_se_meso_low)
dim(met_asv_se_meso_low)

fish_asv_se_meso_low <- melt_fish_tax %>%
  select(OTUid, Abundance, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT, sample, depth_group) %>%
  filter(sample %in% pru_sample) %>%
  filter(depth_group == "800-1000") %>% #surface samples
  select(-depth_group) %>%
  pivot_wider(names_from = "sample", values_from = "Abundance", values_fill = 0) %>%
  select(order(colnames(.))) %>% 
  data.frame()

colnames(fish_asv_se_meso_low)
dim(fish_asv_se_meso_low)

save(prot_asv_se_meso_low, met_asv_se_meso_low, fish_asv_se_meso_low, file = "./R-outputs/alleuk_asv_se_meso_up.RData")

```

```{r make-physeq-depths-input, eval = TRUE, message = FALSE, warning = FALSE}

prot_asv_epi <- select(prot_asv_se_epi, -c(Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% 
  column_to_rownames(var = "OTUid") %>%
  filter(rowSums(.) != 0) %>% as.matrix()

prot_tax_epi <- as.matrix(select(prot_asv_se_epi, OTUid, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
prot_asv_epi_table <- otu_table(prot_asv_epi, taxa_are_rows = TRUE)
prot_tax_epi_table <- tax_table(prot_tax_epi)
prot_phy_epi_core <- phyloseq(prot_asv_epi_table, prot_tax_epi_table)
prot_phy_epi_core # 1123, 15

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
prot_phy_epi_prev <- microbiome::prevalence(
  prot_phy_epi_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(prot_phy_epi_core, "Class", detection = 0) + theme(legend.position = "none")

# ok let's use prevalent ASVs at prevalence > 10% (we have 15 samples so that ASVs have to be in at least 2 samples)
prot_phy_epi_core <- microbiome::core(prot_phy_epi_core, detection = 0, prevalence = .1)
prot_phy_epi_core # reduces to 991 ASVs

prot_asv_meso_up <- select(prot_asv_se_meso_up, -c(Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>%
  column_to_rownames(var = "OTUid") %>% 
  filter(rowSums(.) != 0) %>% as.matrix()
                     
prot_tax_meso_up <- as.matrix(select(prot_asv_se_meso_up, OTUid, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))

prot_asv_meso_up_table <- otu_table(prot_asv_meso_up, taxa_are_rows = TRUE)
prot_tax_meso_up_table <- tax_table(prot_tax_meso_up)
prot_phy_meso_up_core <- phyloseq(prot_asv_meso_up_table, prot_tax_meso_up_table)
prot_phy_meso_up_core # 1865, 21

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
prot_phy_meso_up_prev <- microbiome::prevalence(
  prot_phy_meso_up_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(prot_phy_meso_up_core, "Class", detection = 0) + theme(legend.position = "none")

# ok let's use prevalent ASVs at prevalence > 10% (we have 21 samples so that ASVs have to be in at least 2 samples)
prot_phy_meso_up_core <- microbiome::core(prot_phy_meso_up_core, detection = 0, prevalence = .1)
prot_phy_meso_up_core # reduces to 1345 ASVs


prot_asv_meso_low <- (select(prot_asv_se_meso_low, -c(Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>%   
  filter(rowSums(.) != 0) %>% as.matrix()

prot_tax_meso_low <- as.matrix(select(prot_asv_se_meso_low, OTUid, Domain, Supergroup, Division, Subdivision, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))

prot_asv_meso_low_table <- otu_table(prot_asv_meso_low, taxa_are_rows = TRUE)
prot_tax_meso_low_table <- tax_table(prot_tax_meso_low)
prot_phy_meso_low_core <- phyloseq(prot_asv_meso_low_table, prot_tax_meso_low_table)
prot_phy_meso_low_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
prot_phy_meso_low_prev <- microbiome::prevalence(
  prot_phy_meso_low_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(prot_phy_meso_low_core, "Class", detection = 0) + theme(legend.position = "none")

# ok let's use prevalent ASVs at prevalence > 10% (we have 22 samples so that ASVs have to be in at least 2 samples)
prot_phy_meso_low_core <- microbiome::core(prot_phy_meso_low_core, detection = 0, prevalence = .1)
prot_phy_meso_low_core # reduces to 1314 ASVs

met_asv_epi <- (select(met_asv_se_epi, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

met_tax_epi <- as.matrix(select(met_asv_se_epi, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
met_asv_epi_table <- otu_table(met_asv_epi, taxa_are_rows = TRUE)
met_tax_epi_table <- tax_table(met_tax_epi)
met_phy_epi_core <- phyloseq(met_asv_epi_table, met_tax_epi_table)
met_phy_epi_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
met_phy_epi_prev <- microbiome::prevalence(
  met_phy_epi_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(met_phy_epi_core, "Order", detection = 0) + theme(legend.position = "none")
# ok let's use prevalent ASVs at prevalence > 10% (we have 15 samples so that ASVs have to be in at least 2 samples)
met_phy_epi_core <- microbiome::core(met_phy_epi_core, detection = 0, prevalence = .1)
met_phy_epi_core # reduces to 33 ASVs

met_asv_meso_up <- (select(met_asv_se_meso_up, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

met_tax_meso_up <- as.matrix(select(met_asv_se_meso_up, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
met_asv_meso_up_table <- otu_table(met_asv_meso_up, taxa_are_rows = TRUE)
met_tax_meso_up_table <- tax_table(met_tax_meso_up)
met_phy_meso_up_core <- phyloseq(met_asv_meso_up_table, met_tax_meso_up_table)
met_phy_meso_up_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
met_phy_meso_up_prev <- microbiome::prevalence(
  met_phy_meso_up_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(met_phy_meso_up_core, "Class", detection = 0) + theme(legend.position = "none")
# ok let's use prevalent ASVs at prevalence > 10% (we have 21 samples so that ASVs have to be in at least 2 samples)
met_phy_meso_up_core <- microbiome::core(met_phy_meso_up_core, detection = 0, prevalence = .1)
met_phy_meso_up_core # reduces to 56 ASVs

met_asv_meso_low <- (select(met_asv_se_meso_low, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

met_tax_meso_low <- as.matrix(select(met_asv_se_meso_low, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
met_asv_meso_low_table <- otu_table(met_asv_meso_low, taxa_are_rows = TRUE)
met_tax_meso_low_table <- tax_table(met_tax_meso_low)
met_phy_meso_low_core <- phyloseq(met_asv_meso_low_table, met_tax_meso_low_table)
met_phy_meso_low_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
met_phy_meso_low_prev <- microbiome::prevalence(
  met_phy_meso_low_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(met_phy_meso_low_core, "Phylum", detection = 0) + theme(legend.position = "none")

# ok let's use prevalent ASVs at prevalence > 10% (we have 22 samples so that ASVs have to be in at least 2 samples)
met_phy_meso_low_core <- microbiome::core(met_phy_meso_low_core, detection = 0, prevalence = .1)
met_phy_meso_low_core # reduces to 60 ASVs


fish_asv_epi <- (select(fish_asv_se_epi, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

fish_tax_epi <- as.matrix(select(fish_asv_se_epi, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
fish_asv_epi_table <- otu_table(fish_asv_epi, taxa_are_rows = TRUE)
fish_tax_epi_table <- tax_table(fish_tax_epi)
fish_phy_epi_core <- phyloseq(fish_asv_epi_table, fish_tax_epi_table)
fish_phy_epi_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
fish_phy_epi_prev <- microbiome::prevalence(
  fish_phy_epi_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(fish_phy_epi_core, "Order", detection = 0) + theme(legend.position = "none")
# ok let's use prevalent ASVs at prevalence > 10% (we have 15 samples so that ASVs have to be in at least 2 samples)
fish_phy_epi_core <- microbiome::core(fish_phy_epi_core, detection = 0, prevalence = .1)
fish_phy_epi_core # reduces to 16 ASVs

fish_asv_meso_up <- (select(fish_asv_se_meso_up, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

fish_tax_meso_up <- as.matrix(select(fish_asv_se_meso_up, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
fish_asv_meso_up_table <- otu_table(fish_asv_meso_up, taxa_are_rows = TRUE)
fish_tax_meso_up_table <- tax_table(fish_tax_meso_up)
fish_phy_meso_up_core <- phyloseq(fish_asv_meso_up_table, fish_tax_meso_up_table)
fish_phy_meso_up_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
fish_phy_meso_up_prev <- microbiome::prevalence(
  fish_phy_meso_up_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(fish_phy_meso_up_core, "Order", detection = 0) + theme(legend.position = "none")
# ok let's use prevalent ASVs at prevalence > 10% (we have 21 samples so that ASVs have to be in at least 2 samples)
fish_phy_meso_up_core <- microbiome::core(fish_phy_meso_up_core, detection = 0, prevalence = .1)
fish_phy_meso_up_core # reduces to 23 ASVs

fish_asv_meso_low <- (select(fish_asv_se_meso_low, -c(Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT)) %>% column_to_rownames(var = "OTUid")) %>% filter(rowSums(.) != 0) %>% as.matrix()

fish_tax_meso_low <- as.matrix(select(fish_asv_se_meso_low, OTUid, Phylum, Class, Order, Family, Genus, Species, TAX_full, TAX_SHORT) %>% column_to_rownames(var = "OTUid"))
fish_asv_meso_low_table <- otu_table(fish_asv_meso_low, taxa_are_rows = TRUE)
fish_tax_meso_low_table <- tax_table(fish_tax_meso_low)
fish_phy_meso_low_core <- phyloseq(fish_asv_meso_low_table, fish_tax_meso_low_table)
fish_phy_meso_low_core

# let's look at prevalence of taxa. Do I need a threshold to remove anything?
fish_phy_meso_low_prev <- microbiome::prevalence(
  fish_phy_meso_low_core,
  detection = 0,
  sort = TRUE,
  count = TRUE,
  include.lowest = FALSE
) %>% as.data.frame() 

microbiome::plot_taxa_prevalence(fish_phy_meso_low_core, "Order", detection = 0) + theme(legend.position = "none")
# ok let's use prevalent ASVs at prevalence > 10% (we have 22 samples so that ASVs have to be in at least 2 samples)
fish_phy_meso_low_core <- microbiome::core(fish_phy_meso_low_core, detection = 0, prevalence = .1)
fish_phy_meso_low_core # reduces to 23 ASVs

```

```{r save-physeq-input-rdata, eval = TRUE, message = FALSE, warning = FALSE}

# Save rarefied (> 50) phyloseq objects by depths -- prevalence-filtered
save(prot_phy_epi_core, met_phy_epi_core, fish_phy_epi_core, file = "./hpc-inputs/curated_physeq_forSpiecEasi_epi_core_network_20250410.RData")

save(prot_phy_meso_up_core, met_phy_meso_up_core, fish_phy_meso_up_core, file = "./hpc-inputs/curated_physeq_forSpiecEasi_meso_up_core_network_20250410.RData")

save(prot_phy_meso_low_core, met_phy_meso_low_core, fish_phy_meso_low_core, file = "./hpc-inputs/curated_physeq_forSpiecEasi_meso_low_core_network_20250410.RData")

```
