---
title: "phyloseq-ctd-eukarya"
author: "Nina Yang, Postdoctoral Investigator, WHOI"
created date: "2023-02-04"
last updated: "2025-05-13"
output: html_document
editor_options: 
  chunk_output_type: console
---

# 1. Background
This document outlines the steps taken to analyze the 18S V9 and 12S MiFish metabarcoding from eDNA samples collected via CTD from the AR43 Cruise (Armstrong, March 2020). 

## 1.1 Set up Environment

### 1.1.1 Install and load packages
```{r install-pkgs, include=FALSE}

if(!require(BiocManager)){install.packages("BiocManager")}
if(!require(devtools)){install.packages("devtools")}
if(!require(conflicted)){install.packages("conflicted")}
if(!require(phyloseq)){install.packages("phyloseq")}
if(!require(phylosmith)){install.packages("phylosmith")}
if(!require(tidyverse)){install.packages("tidyverse")}
if(!require(ggpubr)){install_github("ggpubr")}
if(!require(dplyr)){install_github("dplyr")}
if(!require(vegan)){install_github("vegan")}
if(!require(patchwork)){install.packages("patchwork")}
if(!require(Tjazi)){remotes::install_github("thomazbastiaanssen/Tjazi")}
if(!require(pairwiseAdonis)){install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")}
if(!require(betapart)){install.packages("betapart")}
if(!require(ComplexUpset)){remotes::install_github("krassowski/complex-upset")}

```

```{r load-pkgs, include=FALSE}
# load packages
library("devtools"); packageVersion("devtools")
library("conflicted"); packageVersion("conflicted")
library("tidyverse"); packageVersion("tidyverse")
library("phyloseq"); packageVersion("phyloseq")
library("phylosmith"); packageVersion("phylosmith")
library("vegan"); packageVersion("vegan")
library("patchwork"); packageVersion("patchwork")
library("broom"); packageVersion("patchwork")
library("pairwiseAdonis"); packageVersion("pairwiseAdonis")
library("Tjazi"); packageVersion("Tjazi")
library("betapart"); packageVersion("betapart")
library("ComplexUpset"); packageVersion("ComplexUpset")

```

### 1.1.2 Setup figure

```{r fig-setup, include = FALSE}
# Chunk options
knitr::opts_chunk$set(
 fig.width = 12,
 fig.asp = 0.5,
 out.width = "80%"
)

```

```{r theme_ggplot2, echo = FALSE}
theme_custom_ppt <- function() {
  theme_bw(
    base_size = 14) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent'),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      legend.background = element_rect(color = NA, fill='transparent'), 
      legend.box.background = element_rect(color = NA, fill='transparent'),
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = 'transparent', linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 12)
      )
}
theme_custom_poster <- function() {
  theme_bw(
    base_size = 30) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(), 
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1.5),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = "white", linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 30)
      )
}
theme_custom_ppt_blkbg <- function() {
  theme_bw(
    base_size = 14) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent', color = 'transparent'),
      panel.border = element_rect(linewidth = 1, color = "white"),
      plot.title = element_text(color = "white"),
      # axes
      axis.text.y = element_text(color = "white", size = 14), 
      axis.text.x = element_text(color = "white", size = 14), 
      axis.ticks.x = element_line(color = "white"),
      axis.ticks.y = element_line(color = "white"),
      axis.title.x = element_text(color = "white", size = 16),
      axis.title.y = element_text(color = "white", size = 16),
      # legend
      legend.text = element_text(colour ="white"), 
      legend.position = "right",
      legend.title = element_text(colour = "white", size = 14), 
      legend.background = element_rect(color = NA, fill='transparent'), 
      legend.box.background = element_rect(color = NA, fill='transparent'),
      # faceting
      strip.background.y = element_rect(color = "white", fill = "transparent", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = "transparent", linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 14, color = "white")
      )
}
theme_custom_upset <- function() {
  theme_bw(
    base_size = 10) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent'),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_blank(),
      axis.text.x.bottom = element_blank(),
      axis.ticks.x = element_blank(),
      # legend
      legend.title = element_text(size = 0),
      legend.text = element_text(size = 8),
      legend.key.height = unit(0.25, "mm"),
      legend.key.width = unit(1, "mm"),
      legend.key.size = unit(2, 'mm'),
      legend.margin = margin(t = 0, r = 0, b = 0, l = -1))
}
theme_custom_paper <- function() {
  theme_bw(
    base_size = 10) + 
    theme(
      # plot background
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(fill = 'transparent'), 
      plot.background = element_rect(fill = 'transparent'),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
      # axes
      axis.text.y = element_text(colour = "black"), 
      axis.text.x = element_text(colour = "black"), 
      # legend
      legend.text = element_text(colour ="black"), 
      legend.position = "right",
      legend.title = element_text(colour = "black"), 
      legend.background = element_rect(color = NA, fill='transparent'), 
      legend.box.background = element_rect(color = NA, fill='transparent'),
      # faceting
      strip.background.y = element_rect(color = "black", fill = "#ededed", linewidth = 1, linetype = "solid"), 
      strip.background.x = element_rect(color = "white", fill = 'transparent', linewidth = 1, linetype = "solid"),
      strip.text = element_text(size = 8)
      )
}
       
# Changing the default theme
theme_set(theme_custom_paper())

```


## 1.2 Evaluate classifier

```{r classifier-stats, eval = TRUE, message = FALSE, warning = FALSE}
### make this classifier plot
classifier <- read_csv("classifier_stats.csv")
head(classifier)


classifier$Taxonomy <- fct_relevel(classifier$Taxonomy, "Domain", "Supergroup",
                             "Division", 
                             "Subdivision", 
                             "Class", "Order", "Family", "Genus", "Species")


classifier_plot_prot <- classifier %>% 
  filter(Group == "Protist") %>%
  ggplot(aes(x = Taxonomy, y = Fmeasure, group = 1)) +
  geom_point(color = "#78C18A", alpha = 1, size = 1, stroke = 1) + 
  geom_line(color = "#78C18A", alpha = 1, linewidth = 1) +
  geom_hline(yintercept = 0.95, linetype = "dotted", color = "black") +
  scale_y_continuous(limits = c(0, 1)) +
  labs(
    x = "Taxonomic Rank",
    y = "F-measure",
    title = "Protists"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5)
  ) + theme_custom_paper()

classifier_plot_prot

classifier$Taxonomy <- fct_relevel(classifier$Taxonomy, "Domain", "Kingdom", "Phylum",
                             "Class", "Order", "Family", "Genus", "Species")

classifier_plot_met <- classifier %>% 
  filter(Group == "Invertebrate") %>%
  ggplot(aes(x = Taxonomy, y = Fmeasure, group = 1)) +
  geom_point(color = "#FE869B", alpha = 1, size = 1, stroke = 1) + 
  geom_line(color = "#FE869B", alpha = 1, linewidth = 1) +
  geom_hline(yintercept = 0.95, linetype = "dotted", color = "black") +
  scale_y_continuous(limits = c(0, 1)) +
  labs(
    x = "Taxonomic Rank",
    y = "F-measure",
    title = "Invertebrates"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5)
  ) + theme_custom_paper()

classifier_plot_met  
  

classifier_plot_fish <- classifier %>% 
  filter(Group == "Vertebrate") %>%
  ggplot(aes(x = Taxonomy, y = Fmeasure, group = 1)) +
  geom_point(color = "#1986be", alpha = 1, size = 1, stroke = 1) + 
  geom_line(color = "#1986be", alpha = 1, linewidth = 1) +
  geom_hline(yintercept = 0.95, linetype = "dotted", color = "black") +
  scale_y_continuous(limits = c(0, 1)) +
  labs(
    x = "Taxonomic Rank",
    y = "F-measure",
    title = "Vertebrates"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5)
  ) + theme_custom_paper()

classifier_plot_fish


classifier_plot_prot / classifier_plot_met / classifier_plot_fish &
    theme(plot.title = element_text(size = 10),
        axis.text = element_text(size = 7),
        axis.title = element_text(size = 8))

ggsave("./plots-final/classifier-stats.pdf", height = 6, width = 5, units = "in", dpi=300)


```

## 1.3 Setup Files

### 1.3.1 Protists
```{r import-files-prot, eval = TRUE, message = FALSE, warning = FALSE}

physeq_prot <- readRDS(file = "../02_phyloseq/18S/decontam-protist-ctd-phyloseq_05.rds")
physeq_prot # 9737 taxa
head(sample_data(physeq_prot))
head(tax_table(physeq_prot))

# remove ASVs that are not considered Protists or relevant for this analysis
physeq_prot <- subset_taxa(physeq_prot, is.na(Domain) | Domain !="Eukaryota:nucl")
physeq_prot # 9179 taxa
physeq_prot <- subset_taxa(physeq_prot, is.na(Domain) | Domain !="Unassigned")
physeq_prot # 9666 taxa
physeq_prot <- subset_taxa(physeq_prot, Supergroup != "NA")
physeq_prot # 7127 taxa

# check physeq
ntaxa(physeq_prot)
nsamples(physeq_prot)
sample_names(physeq_prot)[1:5]  
rank_names(physeq_prot)  
sample_variables(physeq_prot)
tax_table(physeq_prot)[90:120, 4]

# let's remove the controls from physeq
physeq_prot_sample <- subset_samples(physeq_prot, Type=="sample")
physeq_prot_sample

# remove controls from metadata as well as other columns
# load metadata
metadata_18s <- read.table("../00_qiime2-outputs/18S/metadata_AR43_update_ctd.txt", sep='\t', header=T, row.names=1, comment = "")
head(metadata_18s)
metadata_18s$sample <- rownames(metadata_18s)

colnames(metadata_18s)
metadata_18s <- metadata_18s %>%
  select(sample, c(2:16))

head(metadata_18s)
tail(metadata_18s)

metadata_18s <- filter(metadata_18s, Type == "sample")
colnames(metadata_18s)
summary(metadata_18s)

physeq_prot_sample # 7127
sample_data(physeq_prot_sample)

# originally wanted to remove rare reads but decided not to do this step so I am keeping the code so that it's easy to remove reads, but basically not removing anything. "physeq_xx_sample" files are the same as "physeq_xx_rr" files. 
physeq_prot_rr <- prune_taxa(taxa_sums(physeq_prot_sample) >= 0, physeq_prot_sample)
physeq_prot_rr # 7127
sample_data(physeq_prot_rr)

```

### 1.3.2 Metazoa
```{r import-files-met, eval = TRUE, message = FALSE, warning = FALSE}

# import met and protist phyloseq objects to combine later on

# the metazoa phyloseq object after decontam
physeq_met <- readRDS(file = "../02_phyloseq/18S/decontam-metazoa-ctd-phyloseq_05.rds")
physeq_met #480

# remove ASVs that do not fall under marine Metazoa
physeq_met <- subset_taxa(physeq_met, is.na(Kingdom) | Kingdom != "Unassigned")
physeq_met #432
physeq_met <- subset_taxa(physeq_met, is.na(Kingdom) | Kingdom != " k__Archaea")
physeq_met #431
physeq_met <- subset_taxa(physeq_met, is.na(Kingdom) | Kingdom != " k__Fungi")
physeq_met #431
physeq_met <- subset_taxa(physeq_met, is.na(Phylum) | Phylum != " p__Protalveolata")
physeq_met #431
physeq_met <- subset_taxa(physeq_met, is.na(Class) | Class != " c__Arachnida")
physeq_met #430
physeq_met <- subset_taxa(physeq_met, is.na(Class) | Class != " c__Insecta")
physeq_met #426
physeq_met <- subset_taxa(physeq_met, is.na(Species) | Species != " s__Homo_sapiens")
physeq_met #426

# let's also remove the remaining vertebrates, which are all "Actinopterygii" and cannot resolve any further. We have 12S data so this will be cleaned out
physeq_met <- subset_taxa(physeq_met, is.na(Phylum) | Phylum != " p__Vertebrata")
physeq_met #404

# check physeq
ntaxa(physeq_met)
nsamples(physeq_met)
sample_names(physeq_met)[1:5]  
rank_names(physeq_met)  
sample_variables(physeq_met)
tax_table(physeq_met)[90:120, 2]

# let's remove the controls from physeq
physeq_met_sample <- subset_samples(physeq_met, Type=="sample")

# remove controls from metadata as well as other columns
# load metadata
metadata_18s <- read.table("../00_qiime2-outputs/18S/metadata_AR43_update_ctd.txt", sep='\t', header=T, row.names=1, comment = "")
head(metadata_18s)
metadata_18s$sample <- rownames(metadata_18s)

colnames(metadata_18s)
metadata_18s <- metadata_18s %>%
  select(sample, c(2:16))

head(metadata_18s)
tail(metadata_18s)

metadata_18s <- filter(metadata_18s, Type == "sample")
colnames(metadata_18s)
summary(metadata_18s)

physeq_met_sample
# originally wanted to remove rare reads but decided not to do this step so I am keeping the code so that it's easy to remove reads, but basically not removing anything. "physeq_xx_sample" files are the same as "physeq_xx_rr" files. 
physeq_met_rr <- prune_taxa(taxa_sums(physeq_met_sample) >= 0, physeq_met_sample)
physeq_met_rr # 404

```

### 1.3.3 Fish
```{r import-files-fish, eval = TRUE, message = FALSE, warning = FALSE}

physeq_fish <- readRDS(file = "../02_phyloseq/12S/decontam-12s-phyloseq_05.rds")
physeq_fish # 859 taxa
head(sample_data(physeq_fish))
head(tax_table(physeq_fish))

physeq_fish <- subset_taxa(physeq_fish, is.na(Class) | Class !="Aves") # this is chicken
ntaxa(physeq_fish) # 858
physeq_fish <- subset_taxa(physeq_fish, is.na(Genus) | Genus !="Fundulus")
ntaxa(physeq_fish) # 857
physeq_fish <- subset_taxa(physeq_fish, is.na(Genus) | Genus !="Centropyge")
ntaxa(physeq_fish) # 856

# check physeq
ntaxa(physeq_fish)
nsamples(physeq_fish)
sample_names(physeq_fish)[1:5]  
rank_names(physeq_fish)  
sample_variables(physeq_fish)
tax_table(physeq_fish)[90:120, 4]

# let's remove the controls from physeq
physeq_fish_sample <- subset_samples(physeq_fish, Type=="sample")
physeq_fish_sample

# remove controls from metadata as well as other columns (metadata imported earlier)
metadata_12s <- read.table("../00_qiime2-outputs/12S/metadata_AR43_12S_merged.txt", sep='\t', header=T, row.names=1, comment = "")
head(metadata_12s)
colnames(metadata_12s)
metadata_12s <- metadata_12s %>%
  select(sample, reads, Type, Cruise, platform, cast, deployment, depth, experiment, diel, rep, depth_group, temp, oxy, sal, fluor)

head(metadata_12s)
colnames(metadata_12s)

metadata_12s <- filter(metadata_12s, Cruise == "AR43")
metadata_12s <- filter(metadata_12s, Type == "sample")
colnames(metadata_12s)
summary(metadata_12s)

physeq_fish_sample # 856

physeq_fish1 <- subset_taxa(physeq_fish_sample, Class != "NA") 
physeq_fish1 #787
check_physeq_fish <- psmelt(physeq_fish1)
actinopterygii <- check_physeq_fish %>% filter(Class == "Actinopterygii" & is.na(Order)) %>% select(OTU) %>% unique() 
dim(actinopterygii) ## 321 ASVs

# steps to blast and assign
# 1. convert unassigned csv to txt file (delete extraneous headers and columns, change extension). 
# 2. I will run a script to extract sequences for the unknown feature IDs based against a reference fasta (prdb-metazoa-ctd-rep-seqs.fasta - created during decontam)
## 2.1. unassigned actinopterygii code: nohup ./actinopterygii_unassigned_script.sh > actinopterygii_unassigned_script_log.txt 2>&1 &
# 3. import for blastn in Geneious Prime
# 4. Retrieve top hit. Keep the top hit that is at least 95% grade (explain grade in GP), is not uncultured or other unknown. Keep top hit based on grade

# filter unassigned metazoa
unass_actinopterygii <- (actinopterygii$OTU) %>% unique() 
view(unass_actinopterygii) # there are 321
write.csv(unass_actinopterygii, "unassigned_vert/actinopterygii_unassigned.csv", row.names = TRUE)

# 123 did not have any results
# 117 did not pass grade at least 95%
# 19 contaminants: bacteria, danio rerio 
# retained 62 that seem to be fish that I will have to rename

## let's remove the contaminants
vert_assigned <- read.csv("./unassigned_vert/geneious_assigned_actinopterygii.csv")
dim(vert_assigned) # 81 x 17 (of which 62 are good and 19 are contaminants)
unique(vert_assigned$Order)
colnames(vert_assigned)
colnames(vert_assigned)[12] <- "OTU" # rename column for downstream analysis

# identify the contaminants
asvs_to_remove <- vert_assigned %>% filter(Order == "Contaminant") %>% select(OTU)  #19 as expected
asvs_to_remove <- asvs_to_remove$OTU
check_physeq_fish2 <- check_physeq_fish[!check_physeq_fish$OTU %in% asvs_to_remove, ]
check_physeq_fish2$OTU %>% unique() # 769 ASVs (as expected)

# let's keep troubleshooting by family to identify potential contaminants and other issues
unique(check_physeq_fish2$Family) # there are NAs here
unknown_family <- check_physeq_fish2 %>% filter(is.na(Family) & !is.na(Order)) %>% select(OTU, Class, Order, Family, Genus, Species) %>% unique()  # the Order = NA are all Class = Actinopterygii that I could not annotate
unknown_family$Order %>% unique()
unknown_family$OTU %>% unique() # 109 ASVs

## 1. start with Cetartiodactyla & Carnivora (Mammals)
unass_carnivora <- unknown_family %>% filter(Order == "Cetartiodactyla" | Order == "Carnivora") %>% select(OTU) %>% unique()
view(unass_carnivora) # there are 38
write.csv(unass_carnivora, "unassigned_vert/carnivora_unassigned.csv", row.names = TRUE)

## 2. then do fish (Perciformes, Argentiniformes)
unass_fish_family <- unknown_family %>% filter(Order == "Perciformes" | Order == "Argentiniformes") %>% select(OTU) %>% unique()
view(unass_fish_family) # there are 71
write.csv(unass_fish_family, "unassigned_vert/fish_family_unassigned.csv", row.names = TRUE)

# all 38 Carnivora are terrestrial land contaminants
# I got 65/71 fish (note that Perciformes was annotated to Beryciformes...), the other 6 are just not well annotated so I will leave it at that...

# identify the contaminants
asvs_to_remove <- unass_carnivora$OTU # 38 as expected
# Filter out ASVs that are in asvs_to_remove
check_physeq_fish3 <- check_physeq_fish2[!check_physeq_fish2$OTU %in% asvs_to_remove, ]
check_physeq_fish3$OTU %>% unique() # 730 ASVs (as expected)

## from experience - the mammals are a little suspicious so I am going to go troubleshoot all of these again especially the family of seals...
check_taxonomy <- check_physeq_fish3 %>% filter(is.na(Species)) %>% select(OTU, Class, Order, Family, Genus, Species) %>% unique() 
unass_carnivora_genus <- check_taxonomy %>% filter(Class == "Mammalia") %>% select(OTU) %>% unique()
view(unass_carnivora_genus) # there are 6
write.csv(unass_carnivora_genus, "unassigned_vert/carnivora_unassigned_genus.csv", row.names = TRUE)

vert_assigned <- read.csv("./unassigned_vert/geneious_assigned_carnivora_genus.csv")
dim(vert_assigned) # 6 x 15
head(vert_assigned)
unique(vert_assigned$Genus)
colnames(vert_assigned)
colnames(vert_assigned)[12] <- "OTU" # rename column for downstream analysis

# identify the contaminants
asvs_to_remove <- vert_assigned %>% filter(Genus == "Contaminant") %>% select(OTU)  #3 as expected
asvs_to_remove <- asvs_to_remove$OTU
check_physeq_fish4 <- check_physeq_fish3[!check_physeq_fish3$OTU %in% asvs_to_remove, ]
check_physeq_fish4$OTU %>% unique() # 727 ASVs (as expected)

ASVs_physeq_fish <- check_physeq_fish4 %>% select(OTU, Abundance, sample) %>% 
  pivot_wider(names_from = sample, values_from = Abundance, values_fill = 0)
head(ASVs_physeq_fish)
ASVs <- as.matrix(ASVs_physeq_fish[,-1])  # exclude OTU column
rownames(ASVs) <- ASVs_physeq_fish$OTU

## updated metadata
head(metadata_12s)
rownames(metadata_12s) <- metadata_12s$sample

## taxonomy
tax_physeq_fish <- check_physeq_fish4 %>% select(OTU, Domain, Phylum, Class, Order, Family, Genus, Species) %>% unique() 
dim(tax_physeq_fish) #727 x 8
head(tax_physeq_fish)
tax_table_fish <- as.matrix(tax_physeq_fish[,-1])
head(tax_table_fish)
rownames(tax_table_fish) <- tax_physeq_fish$OTU
head(tax_table_fish)

physeq_fish_rr <- phyloseq(
  otu_table(ASVs, taxa_are_rows = T),
  tax_table(tax_table_fish),
  sample_data(metadata_12s)
)

physeq_fish_rr # 727 ASVs
#physeq_fish_rr_test <- prune_taxa(taxa_sums(physeq_fish_rr) >= 50, physeq_fish_rr)
#physeq_fish_rr_test # 406

```

### 1.3.4 Set up phyloseq objects and save
```{r setup-phyloseq-save, eval = TRUE, message = FALSE, warning = FALSE}

# rename sample_id for each of these so they are exactly the same
sample_names(physeq_met_rr) <- sample_data(physeq_met_rr)$sample
sample_names(physeq_prot_rr) <- sample_data(physeq_prot_rr)$sample
sample_names(physeq_fish_rr) <- sample_data(physeq_fish_rr)$sample

physeq_fish_rr <- set_sample_order(physeq_fish_rr, "sample")
sample_names(physeq_fish_rr)

physeq_prot_rr <- set_sample_order(physeq_prot_rr, "sample")
sample_names(physeq_prot_rr)

physeq_met_rr <- set_sample_order(physeq_met_rr, "sample")
sample_names(physeq_met_rr)

# Save multiple objects, rr = rare reads removed. However, for this analysis, we did not remove rare reads.
save(physeq_met_rr, physeq_prot_rr, physeq_fish_rr, file = "./phyloseq_rr_fordiversityanalyses.RData")
save(metadata_18s, metadata_12s, file = "./updated_metadata_fordiversityanalyses.RData")

```

```{r load-rdata, eval = TRUE, message = FALSE, warning = FALSE}

# load the data again
load("../03_diversity/phyloseq_rr_fordiversityanalyses.RData", verbose = T)
load("../03_diversity/updated_metadata_fordiversityanalyses.RData", verbose = T)

```


# 2. Alpha Diversity
## 2.1 Rarefaction

Rare reads were not removed. Instead, samples were rarefied based on a sampling depth threshold determined by an iterative, subsampling approach from Pat Schloss (https://journals.asm.org/doi/10.1128/msphere.00355-23). 

### 2.1.1 Setting rarefaction cut-off threshold 

```{r rarefy-prot, eval = TRUE, message = FALSE, warning = FALSE}

# rarefy protists data
physeq_prot_rr
head(sample_data(physeq_prot_rr))

# let's "melt" the dataframe
melt_prot <- psmelt(physeq_prot_rr)
head(melt_prot)
colnames(melt_prot)

# let's get a sense for the distribution of reads across samples for rarefying
prot_sampling_coverage <- melt_prot %>%
  dplyr::select(sample, Abundance) %>%
  group_by(sample) %>%
  dplyr::summarize(n_seqs = sum(Abundance))

head(prot_sampling_coverage)
dim(prot_sampling_coverage) # 72 samples

# density
## curve is nice / smooth, but also this is a bad thing; don't have sense of count on y-axis
prot_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_density()

# let's use histogram
## everything has at least 10,000 reads
prot_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_histogram(binwidth = 500) +
  coord_cartesian(xlim=c(0, 50000))

# jitter plot - position on x is randomized so they don't overlap
# point
prot_sampling_coverage %>% 
  ggplot(aes(x=1, y = n_seqs)) + 
  geom_jitter() + theme_light() +
  scale_y_log10() # a handful of samples below 10,000 

# looks 3 samples have below ~15000

prot_sampling_coverage %>% 
  arrange(n_seqs) %>%
  ggplot(aes(x=1:nrow(.), y = n_seqs)) +
  geom_line() + theme_light() +
  coord_cartesian(xlim = c(0,10), ylim = c(0, 50000))

# cut off at 15,000
prot_sampling_coverage %>% 
  arrange(n_seqs) %>%
  print(n=10)

low_seqs_prot <- melt_prot %>%
  dplyr::select(sample, Abundance, deployment, depth) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  filter(n_seqs < 15000) %>%
  distinct(sample, .keep_all = TRUE) %>%
  print()

colnames(melt_prot)

# finalize data frame for plotting - remove samples with < 15,000 reads
rare_prot <- melt_prot %>%
  dplyr::select(sample, OTU, Abundance) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  filter(n_seqs >= 15000) %>%
  ungroup() %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

```

```{r rarefy-met, eval = TRUE, message = FALSE, warning = FALSE}

physeq_met_rr
head(sample_data(physeq_met_rr))
melt_met <- psmelt(physeq_met_rr)
head(melt_met)
colnames(melt_met)
dim(melt_met)

# After reviewing the unassigned sequences, I had to remove this sequence: 5081a2333021c38dbd6b8334bdcff438 which is unassigned past phylum (blast annotated as "Balaustium murorum" or red velvet mite)
melt_met <- melt_met %>% filter(OTU != "5081a2333021c38dbd6b8334bdcff438")
# need to remove this "rodent". I found this after assigning unassigned sequences 
melt_met <- melt_met %>% filter(OTU != "500f4198862a145397ef700583fd575e") 
dim(melt_met)

met_sampling_coverage <- melt_met %>%
  dplyr::select(sample, Abundance) %>%
  dplyr::group_by(sample) %>%
  summarize(n_seqs = sum(Abundance))

head(met_sampling_coverage)
dim(met_sampling_coverage)

# density
## curve is nice / smooth, but also this is a bad thing; don't have sense of count on y-axis
met_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_density()

# let's use histogram
met_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_histogram(binwidth = 1000) +
  coord_cartesian(xlim=c(0, 10000))

# jitter plot - position on x is randomized so they don't overlap
# point
met_sampling_coverage %>% 
  ggplot(aes(x=1, y = n_seqs)) + 
  geom_jitter() + theme_light() +
  scale_y_log10() 

# looks like 2 samples below 6000

met_sampling_coverage %>% 
  arrange(n_seqs) %>%
  ggplot(aes(x=1:nrow(.), y = n_seqs)) +
  geom_line() + theme_light() +
  coord_cartesian(xlim = c(0,10), ylim = c(0, 10000))

# cut off at 6000
met_sampling_coverage %>% 
  arrange(n_seqs) %>%
  print(n=20)

low_seqs_met <- melt_met %>%
  dplyr::select(sample, Abundance, deployment, depth) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  filter(n_seqs < 6000) %>%
  distinct(sample, .keep_all = TRUE) %>%
  print()

colnames(melt_met)
head(melt_met)

# finalize data frame for plotting
rare_met <- melt_met %>%
  dplyr::select(sample, OTU, Abundance) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  ungroup() %>%
  filter(n_seqs >= 6000) %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

```

```{r rarefy-fish, eval = TRUE, message = FALSE, warning = FALSE}

physeq_fish_rr
head(sample_data(physeq_fish_rr))
melt_fish <- psmelt(physeq_fish_rr)
head(melt_fish)
colnames(melt_fish)

fish_sampling_coverage <- melt_fish %>%
  dplyr::select(sample, Abundance) %>%
  dplyr::group_by(sample) %>% 
  summarize(n_seqs = sum(Abundance))

head(fish_sampling_coverage)
dim(fish_sampling_coverage)

# density
## curve is nice / smooth, but also this is a bad thing; don't have sense of count on y-axis
fish_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_density()

# let's use histogram
fish_sampling_coverage %>% 
  ggplot(aes(x=n_seqs)) + 
  geom_histogram(binwidth = 50) + 
  coord_cartesian(xlim=c(0, 600))

# jitter plot - position on x is randomized so they don't overlap
# point
fish_sampling_coverage %>% 
  ggplot(aes(x=1, y = n_seqs)) + 
  geom_jitter() + theme_light() +
  scale_y_log10()

# seems to be several samples that now have no sequences after the data cleaning steps
fish_sampling_coverage %>% 
  arrange(n_seqs) %>%
  ggplot(aes(x=1:nrow(.), y = n_seqs)) +
  geom_line() + theme_light() +
  geom_point() +
  coord_cartesian(xlim = c(0,20), ylim = c(0, 10000))

# cut off at 3000
fish_sampling_coverage %>% 
  arrange(n_seqs) %>%
  print(n=20)

low_seqs_fish <- melt_fish %>%
  dplyr::select(sample, Abundance, deployment, depth) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  filter(n_seqs < 3000) %>%
  distinct(sample, .keep_all = TRUE) %>%
  print()

# 10 samples less than 3000 reads
rare_fish <- melt_fish %>%
  dplyr::select(sample, OTU, Abundance) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  filter(n_seqs >= 3000) %>%
  ungroup() %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

```

```{r rare-threshold-prot, eval = TRUE, message = FALSE, warning = FALSE}
# remind yourself what the dataframe looked like
head(rare_prot)
dim(rare_prot)

# find the lowest # of reads which will serve as the rarefaction threshold
prot_min_seqs <- rare_prot %>% 
  group_by(sample) %>%
  summarize(n_seqs = sum(Abundance)) %>%
  summarize(min = min(n_seqs)) %>%
  pull(min)

prot_min_seqs

```

```{r rare-threshold-met, eval = TRUE, message = FALSE, warning = FALSE}

head(rare_met)
dim(rare_met)

met_min_seqs <- rare_met %>% 
  group_by(sample) %>%
  summarize(n_seqs = sum(Abundance)) %>%
  summarize(min = min(n_seqs)) %>%
  pull(min)

met_min_seqs

```

```{r rare-threshold-fish, eval = TRUE, message = FALSE, warning = FALSE}

#remind yourself what the dataframe looked like
head(rare_fish)
dim(rare_fish)

fish_min_seqs <- rare_fish %>% 
  group_by(sample) %>%
  summarize(n_seqs = sum(Abundance)) %>%
  summarize(min = min(n_seqs)) %>%
  pull(min)

fish_min_seqs

```

### 2.1.2 Save Data
```{r save-rdata, eval = TRUE, message = FALSE, warning = FALSE}

# save rare & melt files
save(melt_prot, melt_met, melt_fish, rare_prot, rare_met, rare_fish, 
     prot_min_seqs, met_min_seqs, fish_min_seqs, file = "rare_melt_dataframes.RData")

```

```{r load-rdata, eval = TRUE, message = FALSE, warning = FALSE}

# load the data again

load(file = "rare_melt_dataframes.RData", verbose = TRUE)

```

## 2.2 Calculate Alpha Diversity

```{r alpha-div-fxn, eval = TRUE, message = FALSE, warning = FALSE}

# this gets rid of all the warnings when using summarise by dplyr
options(dplyr.summarise.inform = FALSE)
getOption("dplyr.summarise.inform")

# functions to calculate alpha diversity metrics
shannon_vegan <- function(data, sample_size){
  data %>%
    dplyr::select(OTU, sample, Abundance) %>%
    group_by(sample) %>%
    uncount(Abundance) %>%
    sample_n(size = sample_size) %>%
    group_by(sample, OTU) %>%
    summarize(abundance = n()) %>%
    summarize(shannon = vegan::diversity(abundance, index = "shannon"))
}
invsimpson_vegan <- function(data, sample_size){
  data %>%
    dplyr::select(OTU, sample, Abundance) %>%
    group_by(sample) %>%
    uncount(Abundance) %>%
    sample_n(size = sample_size) %>%
    group_by(sample, OTU) %>%
    summarize(abundance = n()) %>%
    summarize(invsimpson = vegan::diversity(abundance, index = "invsimpson"))
}
sobs_vegan <- function(data, sample_size){
  data %>%
    dplyr::select(OTU, sample, Abundance) %>%
    group_by(sample) %>%
    uncount(Abundance) %>%
    sample_n(size = sample_size) %>%
    group_by(sample, OTU) %>%
    summarize(abundance = n()) %>%
    summarize(sobs = vegan::specnumber(abundance))
}

# evenness is calculated from shannon index

```

```{r alpha-rare-prot, eval = TRUE, message = FALSE, warning = FALSE}

# plot different diversity metrics as a gut check
rare_prot %>%
  group_by(sample) %>%
  summarise(sobs = specnumber(Abundance), # this is the "exact" number that will be used for diversity metrics
            shannon = vegan::diversity(Abundance, index = "shannon"),
            evenness = shannon/log(sobs),
            invsimpson = vegan::diversity(Abundance, index = "invsimpson"),
            n = sum(Abundance)) %>%
  pivot_longer(cols = c(sobs, shannon, invsimpson, evenness),
                        names_to = "metric") %>%
  ggplot(aes(x = n, y = value)) + 
  geom_point() + 
  geom_smooth() + 
  facet_wrap(~metric, nrow = 4, scales = "free_y")

# save this for "exact" analysis
prot_div <- rare_prot %>%
  group_by(sample) %>%
  summarize(sobs = specnumber(Abundance),
            shannon = vegan::diversity(Abundance, index = "shannon"),
            invsimpson = vegan::diversity(Abundance, index = "invsimpson"),
            evenness = shannon/log(sobs))

head(prot_div)

set.seed(1000)

prot_shannon <- map_dfr(1:1000, ~shannon_vegan(rare_prot, prot_min_seqs), .id="iters")

# prot_div is "exact"
exact <- prot_div %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))

# the subsampling is empirical
empirical <- prot_shannon %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

prot_mean_sd_shannon <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

prot_mean_sd_shannon

## inverse simpson
prot_invsimpson <- map_dfr(1:1000, ~invsimpson_vegan(rare_prot, prot_min_seqs), .id="iters")

# prot_div is "exact"
exact <- prot_div %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))

# inverse simpson
empirical <- prot_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

prot_mean_sd_invsimp <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

prot_mean_sd_invsimp


## ASV richness
prot_sobs <- map_dfr(1:1000, ~sobs_vegan(rare_prot, prot_min_seqs), .id="iters")
head(prot_sobs)

# prot_div is "exact"
exact <- prot_div %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(exact)

# observed "species"
empirical <- prot_sobs %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(empirical)

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

prot_mean_sd_sobs <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

prot_mean_sd_sobs

#let's join these different dataframes

invsimpson_df <- prot_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson = mean(invsimpson, na.rm = TRUE))

shannon_df <- prot_shannon %>%
  group_by(sample) %>%
  summarize(shannon = mean(shannon, na.rm = TRUE))

obs_df <- prot_sobs %>%
  group_by(sample) %>%
  summarize(obs = mean(sobs, na.rm = TRUE))

prot_list = list(shannon_df, invsimpson_df, obs_df, metadata_18s)
prot_list

prot_alpha_div <- prot_list %>% 
  reduce(inner_join, by='sample') %>%
  dplyr::select(sample, obs, shannon, invsimpson, deployment, diel, depth, depth_group)

# let's add an organism group column
prot_alpha_div$Group <- "Protists"
head(prot_alpha_div)
dim(prot_alpha_div)

```

```{r save-rdata-load-prot, eval = TRUE, message = FALSE, warning = FALSE}

save(prot_invsimpson, prot_shannon, prot_sobs, prot_alpha_div, prot_div, file = "prot_alpha_indices_rarefied.RData")

load("prot_alpha_indices_rarefied.RData", verbose = T)

```

```{r alpha-rare-met, eval = TRUE, message = FALSE, warning = FALSE}

# plot different diversity metrics as a gut check
rare_met %>%
  group_by(sample) %>%
  summarise(sobs = specnumber(Abundance),
            shannon = vegan::diversity(Abundance, index = "shannon"),
            evenness = shannon/log(sobs),
            invsimpson = vegan::diversity(Abundance, index = "invsimpson"),
            n = sum(Abundance)) %>%
  pivot_longer(cols = c(sobs, shannon, invsimpson, evenness),
                        names_to = "metric") %>%
  ggplot(aes(x = n, y = value)) + 
  geom_point() + 
  geom_smooth() + 
  facet_wrap(~metric, nrow = 4, scales = "free_y")

# save this for "exact" analysis
met_div <- rare_met %>%
  group_by(sample) %>%
  summarize(sobs = specnumber(Abundance),
            shannon = vegan::diversity(Abundance, index = "shannon"),
            invsimpson = vegan::diversity(Abundance, index = "invsimpson"))

head(met_div)

set.seed(1000)
# I will eventually want to calculate those stats!
met_shannon <- map_dfr(1:1000, ~shannon_vegan(rare_met, met_min_seqs), .id="iters")

# met_div is "exact"
exact <- met_div %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))

# the subsampling is empirical
empirical <- met_shannon %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()


#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

met_mean_sd_shannon <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

met_mean_sd_shannon

## inverse simpson
met_invsimpson <- map_dfr(1:1000, ~invsimpson_vegan(rare_met, met_min_seqs), .id="iters")

# met_div is "exact"
exact <- met_div %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))

# inverse simpson
empirical <- met_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()


#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

met_mean_sd_invsimp <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

met_mean_sd_invsimp

## ASV richness
met_sobs <- map_dfr(1:1000, ~sobs_vegan(rare_met, met_min_seqs), .id="iters")
head(met_sobs)

# met_div is "exact"
exact <- met_div %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(exact)

# observed "species"
empirical <- met_sobs %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(empirical)

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

met_mean_sd_sobs <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

met_mean_sd_sobs

#let's join these different dataframes

invsimpson_df <- met_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson = mean(invsimpson, na.rm = TRUE))

shannon_df <- met_shannon %>%
  group_by(sample) %>%
  summarize(shannon = mean(shannon, na.rm = TRUE))

obs_df <- met_sobs %>%
  group_by(sample) %>%
  summarize(obs = mean(sobs, na.rm = TRUE))

met_list = list(shannon_df,invsimpson_df, obs_df, metadata_18s)

met_alpha_div <- met_list %>% 
  reduce(inner_join, by='sample') %>%
  dplyr::select(sample, obs, shannon, invsimpson, deployment, diel, depth, depth_group)

# let's add an organism group column
met_alpha_div$Group <- "Metazoa"

head(met_alpha_div)
summary(met_alpha_div)

```

```{r save-rdata-load-met, eval = TRUE, message = FALSE, warning = FALSE}

save(met_invsimpson, met_shannon, met_sobs, met_alpha_div, met_div, file = "met_alpha_indices_rarefied.RData")

load("met_alpha_indices_rarefied.RData", verbose = T)

```

```{r alpha-rare-fish, eval = TRUE, message = FALSE, warning = FALSE}

### fish
head(rare_fish)
dim(rare_fish)

# plot different diversity metrics as a gut check
rare_fish %>%
  group_by(sample) %>%
  summarise(sobs = specnumber(Abundance),
            shannon = diversity(Abundance, index = "shannon"),
            simpson = diversity(Abundance, index = "simpson"),
            invsimpson = diversity(Abundance, index = "invsimpson"),
            n = sum(Abundance)) %>%
  pivot_longer(cols = c(sobs, shannon, simpson, invsimpson),
                        names_to = "metric") %>%
  ggplot(aes(x = n, y = value)) + 
  geom_point() + 
  geom_smooth() + 
  facet_wrap(~metric, nrow = 4, scales = "free_y")

# save this for "exact" analysis
fish_div <- rare_fish %>%
  group_by(sample) %>%
  summarize(sobs = specnumber(Abundance),
            shannon = diversity(Abundance, index = "shannon"),
            simpson = diversity(Abundance, index = "simpson"),
            invsimpson = diversity(Abundance, index = "invsimpson"))

head(fish_div)

set.seed(1000)

# I will eventually want to calculate those stats!
fish_shannon <- map_dfr(1:1000, ~shannon_vegan(rare_fish, fish_min_seqs), .id="iters")

# fish_div is "exact"
exact <- fish_div %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))

# the subsampling is empirical
empirical <- fish_shannon %>%
  group_by(sample) %>%
  summarize(shannon_mean = mean(shannon, na.rm = TRUE))
head(empirical)

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()


#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

# 4 samples are at 0

fish_mean_sd_shannon <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "shannon_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error, na.rm = TRUE), sd = sd(error, na.rm = TRUE))

fish_mean_sd_shannon

## inverse simpson

fish_invsimpson <- map_dfr(1:1000, ~invsimpson_vegan(rare_fish, fish_min_seqs), .id="iters")

# fish_div is "exact"
exact <- fish_div %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))

# inverse simpson
empirical <- fish_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson_mean = mean(invsimpson, na.rm = TRUE))


# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

fish_mean_sd_invsimp <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "invsimpson_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

fish_mean_sd_invsimp

## ASV richness
fish_sobs <- map_dfr(1:1000, ~sobs_vegan(rare_fish, fish_min_seqs), .id="iters")
head(fish_sobs)

# fish_div is "exact"
exact <- fish_div %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(exact)

# observed "species"
empirical <- fish_sobs %>%
  group_by(sample) %>%
  summarize(sobs_mean = mean(sobs, na.rm = TRUE))
head(empirical)

# let's compare
# plot a line
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  ggplot(aes(x = exact, y = empirical, group = sample)) +
  geom_abline(intercept = 0, slope = 1, color = "darkgreen") + 
  geom_point() + theme_light()

#error
bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  ggplot(aes(x=error)) + 
  geom_density() + theme_light()

fish_mean_sd_sobs <- bind_rows(empirical = empirical, exact = exact, .id = "approach") %>%
  pivot_wider(names_from = "approach", values_from = "sobs_mean") %>%
  mutate(error = 100*(exact - empirical)/exact) %>%
  summarize(mean = mean(error), sd = sd(error))

fish_mean_sd_sobs

#let's join these different dataframes
invsimpson_df <- fish_invsimpson %>%
  group_by(sample) %>%
  summarize(invsimpson = mean(invsimpson, na.rm = TRUE))

shannon_df <- fish_shannon %>%
  group_by(sample) %>%
  summarize(shannon = mean(shannon, na.rm = TRUE))

obs_df <- fish_sobs %>%
  group_by(sample) %>%
  summarize(obs = mean(sobs, na.rm = TRUE))

fish_list = list(shannon_df, invsimpson_df, obs_df, metadata_12s)

fish_alpha_div <- fish_list %>% 
  reduce(inner_join, by='sample') %>%
  dplyr::select(sample, obs, shannon, invsimpson, deployment, diel, depth, depth_group)

# let's add an organism group column
fish_alpha_div$Group <- "Fish"
head(fish_alpha_div)

```

```{r save-rdata-load-fish, eval = TRUE, message = FALSE, warning = FALSE}

save(fish_invsimpson, fish_shannon, fish_sobs, fish_alpha_div, fish_div, file = "fish_alpha_indices_rarefied.RData")

load("fish_alpha_indices_rarefied.RData", verbose = T)

```

```{r save-rdata-rarefaction-stats, eval = TRUE, message = FALSE, warning = FALSE}

save(prot_mean_sd_sobs, prot_mean_sd_shannon, prot_mean_sd_invsimp, met_mean_sd_sobs, met_mean_sd_shannon, met_mean_sd_invsimp, fish_mean_sd_sobs, fish_mean_sd_shannon, fish_mean_sd_invsimp, file = "./alleuk_mean_sd_rarefaction_stats.RData")

load("./alleuk_mean_sd_rarefaction_stats.RData", verbose = T)

```

```{r alpha-plots-setup-save, eval = TRUE, message = FALSE, warning = FALSE}

# let's merge these
alpha_div_indice <- rbind(prot_alpha_div, met_alpha_div, fish_alpha_div)
head(alpha_div_indice)

save(alpha_div_indice, prot_alpha_div, met_alpha_div, fish_alpha_div, file = "alpha_div_indice_forplotting.RData")

load(file = "alpha_div_indice_forplotting.RData", verbose = T)

```

## 2.3 Plot Species richness

```{r asv-scan-setup, eval = TRUE, message = FALSE, warning = FALSE}

# this function plots relative abundance of ASVs
make_taxabar_asv_relabun_collapse <- function(df) {
  df %>%
    group_by(depth) %>%
    mutate(sum_obs = sum(sobs)) %>%
    mutate(RelativeAbundance = (sobs / sum_obs) * 100, na.rm = TRUE) %>%
    mutate(depth = as.factor(depth)) %>%
    ungroup() %>%
    group_by(depth, TAXA) %>%
    summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(depth, desc(sum_RelativeAbundance)) %>%
    mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>%
    ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.78, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
#    facet_wrap(. ~ deployment) + 
    scale_fill_manual(values = taxa_colors,
                      limits = taxa_sorted) +
    theme_custom_paper()# Use a manually sorted order for the legend
}

# group colors
group_colors <- c("#78C18A", "#FE869B", "#1986be")

# calculate relative abundance
met_relabun <- melt_met %>%
  dplyr::select(sample, OTU, Abundance, everything(), -raw_reads) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  ungroup() %>%
  filter(n_seqs >= 6000) %>%
  ungroup() %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

prot_relabun <- melt_prot %>%
  dplyr::select(sample, OTU, Abundance, everything(), -raw_reads) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  ungroup() %>%
  filter(n_seqs >= 15000) %>%
  ungroup() %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

fish_relabun <- melt_fish %>%
  dplyr::select(sample, OTU, Abundance, -reads, everything()) %>%
  group_by(sample) %>%
  mutate(n_seqs = sum(Abundance)) %>%
  ungroup() %>%
  filter(n_seqs >= 3000) %>%
  ungroup() %>%
  group_by(OTU) %>%
  filter(Abundance > 0) %>%
  ungroup() %>%
  print()

# let's plot prot vs met to get relative proportion of the sequences
prot_check <- prot_relabun %>% select(sample, depth, n_seqs)
prot_check$taxa <- "Protists"
head(prot_check)

met_check <- met_relabun %>% select(sample, depth, n_seqs)
met_check$taxa <- "Invertebrates"
head(met_check)

euk_merge <- rbind(prot_check, met_check) %>% unique()

head(euk_merge)

euk_merge_calc <- euk_merge %>% group_by(sample) %>%
  mutate(sample_seqs = sum(n_seqs)) %>%
  mutate(prop_seqs = n_seqs/sample_seqs * 100)
head(euk_merge_calc)

euk_merge_calc$taxa <- fct_relevel(euk_merge_calc$taxa, "Invertebrates", after = Inf)

euk_plot <- ggplot(euk_merge_calc, aes(y = fct_rev(as.factor(depth)), x = prop_seqs)) +
  geom_boxplot(aes(group = as.factor(depth), fill = taxa), alpha = 1, color ="black", size = 0.25, outlier.size = 0.1) +
  geom_jitter(aes(fill = taxa), color = "black", stroke = 0.25, size = 0.5, alpha = 0.66, shape = 21) + 
  stat_summary(fun=mean, geom="point", shape=23, size=1, fill = "white", color = "black") +
  facet_grid(~taxa) +
  labs(title = NULL,
       y = "Depth (m)",
       x = "Relative Abundance (%)") +
  scale_fill_manual(values = group_colors, guide = "none") + theme_custom_paper() + 
  theme(axis.text = element_text(size = 6),
        axis.title = element_text(size = 7),
        panel.border = element_rect(linewidth = 0.5))
  
euk_plot

ggsave("plots-final/prot_met_18S_sequence_split.pdf", height = 60, width = 80, units = "mm", dpi=300)

euk_merge_calc_stats <- euk_merge_calc %>% select(depth, taxa, prop_seqs) %>%
  group_by(depth, taxa) %>%
  mutate(avg_seqs = mean(prop_seqs)) %>%
  select(-c(sample, prop_seqs)) %>% unique() %>%
  arrange((depth))

```

```{r alphadiv-plots, eval = TRUE, message = FALSE, warning = FALSE}

alpha_div_indice_df <- alpha_div_indice %>% 
  mutate(taxa = case_when(
  Group == "Protists" ~ "Protists (18S)",
    Group == "Metazoa" ~ "Invertebrates (18S)",
    Group == "Fish" ~ "Vertebrates (12S)",
    TRUE ~ Group)) %>%
  mutate(evenness = shannon / log(obs))

head(alpha_div_indice_df)
tail(alpha_div_indice_df)
unique(alpha_div_indice_df$taxa)

alpha_div_indice_df$taxa <- fct_relevel(alpha_div_indice_df$taxa, "Invertebrates (18S)", after = 1)

# let's make a new dataframe
obs_df <- alpha_div_indice_df %>% select(sample, obs, depth, taxa) %>%
  mutate(alpha_div = obs) %>%
  mutate(indice = "ASV Richness") %>%
  select(-obs)

sha_df <- alpha_div_indice_df %>% select(sample, shannon, depth, taxa) %>%
  mutate(alpha_div = shannon) %>%
  mutate(indice = "Shannon") %>%
  select(-shannon)

even_df <- alpha_div_indice_df %>% select(sample, evenness, depth, taxa) %>%
  mutate(alpha_div = evenness) %>%
  mutate(indice = "Evenness") %>%
  filter(!is.na(evenness)) %>%
  select(-evenness)

alpha_obs <- ggplot(obs_df, aes(y = fct_rev(as.factor(depth)), x = alpha_div)) +
  geom_boxplot(aes(group = as.factor(depth), fill = taxa), alpha = 1, color ="black", size = 0.25, outlier.size = 0.1) +
  stat_summary(fun=mean, geom="point", shape=23, size=1, fill = "white", color = "black") +
  facet_grid(indice ~ taxa, scale = "free_x") +
  labs(title = "Richness",
       y = "Depth (m)",
       x = NULL) +
  scale_fill_manual(values = group_colors, guide = "none") + theme_custom_paper() +  
  theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  )
   
alpha_obs

alpha_sha <- ggplot(sha_df, aes(y = fct_rev(as.factor(depth)), x = alpha_div)) +
  geom_boxplot(aes(group = as.factor(depth), fill = taxa), alpha = 1, color ="black", size = 0.25, outlier.size = 0.1) +
  stat_summary(fun=mean, geom="point", shape=23, size=1, fill = "white", color = "black") +
  facet_grid(indice ~ taxa, scale = "free_x") +
  labs(title = "Shannon",
       y = "Depth (m)",
       x = NULL) +
  scale_fill_manual(values = group_colors, guide = "none") + theme_custom_paper() +
    theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  )
   
   
alpha_sha

alpha_even <- ggplot(even_df, aes(y = fct_rev(as.factor(depth)), x = alpha_div)) +
  geom_boxplot(aes(group = as.factor(depth), fill = taxa), alpha = 1, color ="black", size = 0.25, outlier.size = 0.1) +
  stat_summary(fun=mean, geom="point", shape=23, size=1, fill = "white", color = "black") +
  facet_grid(indice ~ taxa, scale = "free_x") +
  labs(title = "Evenness",
       y = "Depth (m)",
       x = NULL) +
  scale_fill_manual(values = group_colors, guide = "none") + theme_custom_paper() +
    theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  )
   
   
alpha_even

alpha_obs / alpha_sha /alpha_even + plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 8),
        plot.margin = unit(c(t = 0.1, r = 0.05, b = 0.1, l = 0.05), "lines"),
        plot.title = element_text(size = 7),
        plot.background = element_blank(),
        panel.spacing = unit(3, "mm"), 
        panel.border = element_rect(color = "black", linewidth = 0.5),
        axis.title.y = element_text(size = 6), 
        axis.text.x = element_text(size = 6), 
        axis.text.y = element_text(size = 6))


ggsave("plots-final/alpha_div_indice_panel_pub.pdf", height = 130, width = 80, units = "mm", dpi=300)

```

```{r prot-asv-taxa-scan, eval = TRUE, message = FALSE, warning = FALSE}

# I want to plot the ASVs split out into major groups by Kingdom - Animalia

head(prot_relabun)
unique(prot_relabun$Division)

# let's group ASVs into main Supergroup & Division/Subdivision taxa
pr2_rename_taxa <- function(df) {
    # Add a column
    df$TAXA <- "Unassigned-Eukaryote"
    #df$TAXA[df$Supergroup == "TSAR" & is.na(df$Division)] = "TSAR-Unassigned"
    df$TAXA[df$Supergroup == "TSAR"] = "TSAR-Other"
    df$TAXA[df$Division == "Telonemia"] = "TSAR-Telonemia"
    df$TAXA[df$Division == "Alveolata"] = "Alveolata-Other"
    df$TAXA[df$Subdivision == "Ciliophora"] = "Alveolata-Ciliates"
    df$TAXA[df$Subdivision == "Dinoflagellata"] = "Alveolata-Dinoflagellates"
    df$TAXA[df$Class == "Syndiniales"] = "Alveolata-Syndiniales"
    df$TAXA[df$Subdivision == "Apicomplexa"] = "Alveolata-Apicomplexa"
    df$TAXA[df$Supergroup == "Cryptista"] = "Cryptista-Other"
    df$TAXA[df$Subdivision == "Kathablepharida"] = "Cryptista-Kathablepharida"
    df$TAXA[df$Division == "Cryptophyta"] = "Cryptophyta"
    df$TAXA[df$Supergroup == "Haptista"] = "Haptista-Other"
    df$TAXA[df$Division == "Centroplasthelida"] = "Haptista-Centroplasthelida"
    df$TAXA[df$Division == "Haptophyta"] = "Haptophyta"
    df$TAXA[df$Division == "Breviatea"] = "Obazoa-Breviatea"
    df$TAXA[df$Division == "Opisthokonta"] = "Obazoa-Opisthokonta"
    df$TAXA[df$Division == "Apusomonada"] = "Obazoa-Apusomonada"
    df$TAXA[df$Subdivision == "Fungi"] = "Obazoa-Fungi"
    df$TAXA[df$Subdivision == "Choanoflagellata"] = "Obazoa-Choanoflagellata"
    df$TAXA[df$Division == "Ancyromonadida"] = "Eukaryota_X-Ancyromonadida"
    df$TAXA[df$Division == "Nibbleridia"] = "Provora-Nibbleridia"
    mast <- unique(filter(df, grepl("MAST", Family)) %>% select(Family))
    mast_list <- as.character(mast$Family)
    df$TAXA[df$Division == "Stramenopiles"] = "Stramenopiles-Other"
    df$TAXA[df$Class == "Mediophyceae"] = "Stramenopiles-Diatoms"
    df$TAXA[df$Class == "Coscinodiscophyceae"] = "Stramenopiles-Diatoms"
    df$TAXA[df$Class == "Bacillariophyceae"] = "Stramenopiles-Diatoms"
    df$TAXA[df$Family %in% mast_list] = "Stramenopiles-MAST"
    df$TAXA[df$Supergroup == "Archaeplastida"] = "Archaeplastida-Other"
    df$TAXA[df$Division == "Picozoa"] = "Archaeplastida-Picozoa"
    df$TAXA[df$Division == "Streptophyta"] = "Archaeplastida-Streptophyta"
    df$TAXA[df$Division == "Streptophyta"] = "Archaeplastida-Streptophyta"
    df$TAXA[df$Division == "Rhodophyta"] = "Rhodophyta"
    df$TAXA[df$Division == "Chlorophyta"] = "Chlorophyta"
    df$TAXA[df$Division == "Rhizaria"] = "Rhizaria-Other"
    df$TAXA[df$Subdivision == "Cercozoa"] = "Rhizaria-Cercozoa"
    df$TAXA[df$Subdivision == "Radiolaria"] = "Rhizaria-Radiolaria"
    df$TAXA[df$Supergroup == "Excavata"] = "Excavata"
    df$TAXA[df$Supergroup == "Amoebozoa"] = "Amoebozoa"
    df$TAXA[df$Supergroup == "Apusozoa"] = "Apusozoa"
    return(df)
}

## rename eDNA dataframe
prot_taxa <- pr2_rename_taxa(prot_relabun)
head(prot_taxa)
unique(prot_taxa$TAXA)

prot_taxa_abun_check <- prot_taxa %>% select(TAXA, Abundance) %>%
  mutate(sumabun = sum(Abundance)) %>%
  group_by(TAXA) %>%
  mutate(relabun = Abundance / sumabun) %>%
  summarise(sum_RelativeAbundance = sum(relabun) * 100, na.rm = TRUE, .groups = "drop")

head(prot_taxa_abun_check)

prot_low_abun_taxa <- prot_taxa_abun_check %>% filter(sum_RelativeAbundance < 1) 
head(prot_low_abun_taxa)
prot_low_abun_taxa = prot_low_abun_taxa$TAXA
prot_low_abun_taxa

# let's group low abundance into "Other Protists"
prot_taxa2 <- prot_taxa %>% mutate(TAXA = case_when(
    TAXA %in% prot_low_abun_taxa ~ "Other Protists",
    TRUE ~ TAXA))

head(prot_taxa2)
unique(prot_taxa2$TAXA)

# Filter out rows where Abundance is greater than 0 and convert 'depth' to a factor
prot_asv <- 
  prot_taxa2 %>%
  filter(Abundance > 0) %>% # Keep only rows with Abundance > 0
  mutate(depth = as.factor(depth)) %>% # Convert 'depth' to a factor
  
  # Group by 'depth', 'OTU', and 'TAXA', then summarize the Abundance
  group_by(depth, OTU, TAXA) %>%
  summarise(Abund = sum(Abundance)) %>% # Sum the Abundance for each group
  ungroup() %>% # Ungroup the data
  # Group by 'depth' and 'TAXA', then calculate species number (sobs)
  group_by(depth, TAXA) %>%
  mutate(sobs = vegan::specnumber(Abund)) %>% # Calculate species number
  
  # Select columns excluding 'OTU' and 'Abund'
  select(-c(OTU, Abund)) %>% # Remove 'OTU' and 'Abund' columns
  unique() %>% # Remove duplicate rows
  ungroup() %>% # Ungroup the data
  
  # Group by 'TAXA' and calculate the sum of sobs for each TAXA
  group_by(TAXA) %>%
  mutate(sum_obs = sum(sobs)) %>% # Sum of sobs for each TAXA
  
  # Arrange by 'depth' and descending 'sum_obs', then remove duplicates
  arrange(depth, desc(sum_obs)) %>% # Arrange by depth and sum_obs
  
  # Reorder 'TAXA' factor based on 'sum_obs' values in ascending order
  mutate(TAXA = fct_reorder(TAXA, sum_obs, .desc = FALSE)) # Reorder TAXA 

head(prot_asv)
    
prot_asv$TAXA <- fct_relevel(prot_asv$TAXA, "Other Protists", after = Inf)
taxa_sorted <- c("Alveolata-Apicomplexa", 
                 "Alveolata-Ciliates",
                 "Alveolata-Dinoflagellates", 
                 # "Alveolata-Other", 
                 "Alveolata-Syndiniales", 
                 "Chlorophyta", 
                 #  "Cryptophyta", 
                 "Excavata",
                 "Haptophyta", #"Rhizaria-Cercozoa", 
                 # "Rhizaria-Other", 
                # "Obazoa-Opisthokonta",
                 "Rhizaria-Radiolaria", # "Rhodophyta",
                 "Stramenopiles-Diatoms", "Stramenopiles-MAST", 
                 "Stramenopiles-Other", 
                 "Other Protists")

taxa_colors <- c("#a1c2fd", "#56C4F0", "#399dff",
                 #"#0b62fb", 
                 "#0344b7",
                 "#9CDC46",#"#dafa0f",
                 "#efdf48", "#ff9aa5",
                 #"#ceceff", "#a7a7ff",
               #  "#ff9d3a",
                 "#7474ff",
                 #"#ff9aa5",
                 "#009a4d","#1ec38c", 
                 "#90eebf",
                 #"#82A4E3", 
                 "#d9d9d9")

prot_asv_plot <- prot_asv %>%
  ggplot(aes(y=fct_rev(depth), x=sobs)) +
  geom_col(aes(group = depth, fill = TAXA), color = "black", width = 0.78, linewidth = 0.25, alpha=1) +
  labs(title=NULL,
       y="Depth (m)",
       x="Observed ASVs - Protists") +
  scale_fill_manual(values = taxa_colors,
                   limits = taxa_sorted) + 
  guides(fill = guide_legend(title = "Protists"),
         color = "none") + theme_custom_paper() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10))

prot_asv_plot

#ggsave("./plots-final/prot_obs_taxa_barplot.pdf", height = 5, width = 5, units = "in", dpi=300)

prot_relabun_asv_plot <- make_taxabar_asv_relabun_collapse(prot_asv)
prot_relabun_asv_plot + 
  guides(fill = guide_legend(title = "Protists"))

#ggsave("./plots-final/prot_obs_taxa_relabun.pdf", height = 5, width = 5, units = "in", dpi=300)

head(prot_asv)

# let's identify the taxa that make up the largest fraction of ASV diversity. This is termed "top taxa" in the code. 
# This pipeline effectively identifies and ranks the top 5 taxa by relative abundance within each depth.
# 
# top_prot <- prot_asv %>% group_by(depth) %>% # group by depth
#   mutate(sum_obs = sum(sobs)) %>% # get the total number of ASVs for each depth in a new column called "sum_obs"
#   mutate(RelativeAbundance = (sobs / sum_obs) * 100, na.rm = TRUE) %>% # calculate the relative abundance of ASVs out of the total # of ASVs at each depth and create a new column. 
#   mutate(depth = as.factor(depth)) %>% # change depth to a factor 
#   ungroup() %>% # ungroup 
#   group_by(depth, TAXA) %>% # regroup by depth and TAXA
#   summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>% # summarize to get the sum of the relative abundance of each taxa at each depth
#   mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>% # Reorders the levels of the TAXA factor based on the sum_RelativeAbundance, maintaining the order of appearance if ties occur.
#   ungroup() %>% # ungroup 
#   group_by(depth) %>% # regroup by depth only 
#   slice_max(order_by = sum_RelativeAbundance, n = 5, with_ties = FALSE) %>% # Selects the top 5 rows in each depth group based on sum_RelativeAbundance, excluding ties beyond the top 5.
#   ungroup() # ungroup
# 
# top_prot %>% group_by(depth) %>% mutate(sum = sum(sum_RelativeAbundance)) %>% select(-TAXA, -sum_RelativeAbundance) %>% unique()
# 
# top_prot_pub <- top_prot %>% ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
#     geom_col(aes(fill = TAXA), color = "black", width = 0.78, linewidth = 0.25) +
#     labs(title = NULL,
#          y = "Depth (m)",
#          x = "Relative Proportion (%)") +
#   scale_x_continuous(limits = c(0, 100)) + 
#     scale_fill_manual(values = taxa_colors,
#                       limits = taxa_sorted) +
#     theme_custom_paper()# Use a manually sorted order for the legend
# 
# top_prot_pub

```

```{r met-asv-taxa-scan1, eval = TRUE, message = FALSE, warning = FALSE}

# clean taxa labeling for metazoa
head(met_relabun)
dim(met_relabun) # 28140 x 31
met_relabun <- met_relabun %>%
  mutate(Domain = gsub("^d__", "", Domain)) %>%
  mutate(Kingdom = gsub("^ k__", "", Kingdom)) %>%
  mutate(Phylum = gsub("^ p__", "", Phylum)) %>%
  mutate(Class = gsub("^ c__", "", Class)) %>%
  mutate(Order = gsub("^ o__", "", Order)) %>%
  mutate(Family = gsub("^ f__", "", Family)) %>%
  mutate(Genus = gsub("^ g__", "", Genus)) %>%
  mutate(Species = gsub("^ s__", "", Species))

head(met_relabun)

# summarize the cleaned dataframe by Kingdom Phylum, and Class, which will help define groupings for renaming and plotting
summary_df <- met_relabun %>% 
  group_by(Kingdom, Phylum, Class, Order, Genus, Species) %>%
  summarise(count = n())

print(summary_df)

# note that for the species "Anthopleura elegantissima", all other taxonomic levels are labeled "Animalia" so I renamed this "TAXA" as Cnidaria-Actiniaria which was then lumped under Cnidaria-Other.
# Fix the taxonomy for aggregating anemone
met_relabun[met_relabun$Species == "Anthopleura_elegantissima", "Phylum"] <- "Cnidaria"
met_relabun[met_relabun$Species == "Anthopleura_elegantissima", "Class"] <- "Hexacorallia"
met_relabun[met_relabun$Species == "Anthopleura_elegantissima", "Order"] <- "Actiniaria"
met_relabun[met_relabun$Phylum == "Nematozoa", "Phylum"] <- "Nematoda"

# Fix the taxonomy for copepod "Poecilostomatoida" which is actually under Cyclopoida
met_relabun[met_relabun$Species == "Sapphirina_darwinii", "Order"] <- "Cyclopoida" 

summary_df <- met_relabun %>% 
  group_by(Kingdom, Phylum, Class, Order, Genus, Species) %>%
  summarise(count = n())

print(summary_df)

silva_met_rename_order <- function(df) {
  df$TAXA <- "Unassigned-Eukaryote"
  df$TAXA[df$Phylum == "Annelida"] = "Annelida"
  df$TAXA[df$Phylum == "uncultured"] = "Unassigned Metazoa"
  df$TAXA[df$Phylum == "Unassigned"] = "Unassigned Metazoa"
  df$TAXA[df$Phylum == "Arthropoda"] = "Arthropoda-Other"
  df$TAXA[df$Order == "Calanoida"] = "Arthropoda-Calanoida"
  df$TAXA[df$Order == "Cyclopoida"] = "Arthropoda-Cyclopoida"
  df$TAXA[df$Order == "Collembola"] = "Arthropoda-Collembola"
  df$TAXA[df$Order == "Eucarida"] = "Arthropoda-Eucarida"
  df$TAXA[df$Order == "Peracarida"] = "Arthropoda-Peracarida"
  df$TAXA[df$Order == "Poecilostomatoida"] = "Arthropoda-Poecilostomatoida"
  df$TAXA[df$Order == "Harpacticoida"] = "Arthropoda-Harpacticoida"
  df$TAXA[df$Order == "Sessilia"] = "Arthropoda-Sessilia"
  df$TAXA[df$Order == "Halocyprida"] = "Arthropoda-Halocyprida"
  df$TAXA[df$Order == "Amphipoda"] = "Arthropoda-Amphipoda"
  df$TAXA[df$Order == "Decapoda"] = "Arthropoda-Decapoda"
  df$TAXA[df$Order == "Mormonilloida"] = "Arthropoda-Mormonilloida"
  df$TAXA[df$Order == "Halocyprida"] = "Arthropoda-Halocyprida"
  df$TAXA[df$Phylum == "Bryozoa"] = "Bryozoa"
  df$TAXA[df$Phylum == "Chaetognatha"] = "Chaetognatha"
  df$TAXA[df$Phylum == "Cnidaria"] = "Cnidaria-Other"
  df$TAXA[df$Order == "Actiniaria"] = "Cnidaria-Actiniaria"
  df$TAXA[df$Order == "Ceriantharia"] = "Cnidaria-Ceriantharia"
  df$TAXA[df$Order == "Anthoathecata"] = "Cnidaria-Anthoathecata"
  df$TAXA[df$Order == "Leptothecata"] = "Cnidaria-Leptothecata"
  df$TAXA[df$Order == "Narcomedusae"] = "Cnidaria-Narcomedusae"
  df$TAXA[df$Order == "Siphonophorae"] = "Cnidaria-Siphonophorae"
  df$TAXA[df$Order == "Trachylinae"] = "Cnidaria-Trachylinae"
  df$TAXA[df$Order == "Trachymedusae"] = "Cnidaria-Trachymedusae"
  df$TAXA[df$Order == "Coronatae"] = "Cnidaria-Coronatae"
  df$TAXA[df$Order == "Semaeostomeae"] = "Cnidaria-Semaeostomeae"
  df$TAXA[df$Phylum == "Ctenophora"] = "Ctenophora-Other"
  df$TAXA[df$Order == "Cydippida"] = "Ctenophora-Cydippida"
  df$TAXA[df$Order == "Platyctenida"] = "Ctenophora-Platyctenida"
  df$TAXA[df$Phylum == "Echinodermata"] = "Echinodermata"
  df$TAXA[df$Phylum == "Gastrotricha"] = "Gastrotricha"
  df$TAXA[df$Phylum == "Hemichordata"] = "Hemichordata"
  df$TAXA[df$Phylum == "Mollusca"] = "Mollusca"
  df$TAXA[df$Phylum == "Nematoda"] = "Nematoda"
  df$TAXA[df$Phylum == "Nemertea"] = "Nemertea"
  df$TAXA[df$Phylum == "Platyhelminthes"] = "Platyhelminthes"
  df$TAXA[df$Phylum == "Rotifera"] = "Rotifera"
  df$TAXA[df$Phylum == "Porifera"] = "Porifera"
  df$TAXA[df$Phylum == "Tunicata"] = "Tunicata-Other"
  df$TAXA[df$Order == "Salpida"] = "Tunicata-Salpida"
  df$TAXA[df$Order == "Doliolida"] = "Tunicata-Doliolida"
  df$TAXA[df$Order == "Copelata"] = "Tunicata-Copelata"
  return(df)
}

met_taxa_order <- silva_met_rename_order(met_relabun)

unique(met_taxa_order$TAXA)

met_taxa_check <- met_taxa_order %>%
  filter(Abundance > 0) %>%
  group_by(Order,TAXA) %>%
  summarise(sum = sum(Abundance)) %>%
  arrange(by = sum) %>%
  print(n=70)

```

```{r assign-unknown-met, eval = TRUE, message = FALSE, warning = FALSE}
# There are a lot of unassigned sequences 

# filter unassigned metazoa
unass_met <- filter(met_taxa_order, TAXA == "Unassigned Metazoa") 
unass_met <- unique(unass_met$OTU)
view(unass_met) # there are 98
#write.csv(unass_met, "unassigned_met/met_unassigned.csv", row.names = TRUE)

# filter unassigned arthropoda
unass_cop <- filter(met_taxa_order, Phylum == "Arthropoda" & Order == "Unassigned") 
unass_cop <- unique(unass_cop$OTU)
view(unass_cop) # there are 44
#write.csv(unass_cop, "./unassigned_met/met_copepod_unassigned.csv", row.names = TRUE)

# steps to blast and assign
# 1. convert unassigned csv to txt file (delete extraneous headers and columns, change extension). 
# 2. I will run a script to extract sequences for the unknown feature IDs based against a reference fasta (prdb-metazoa-ctd-rep-seqs.fasta - created during decontam)
## 2.1. unassigned metazoan code: nohup ./met_unassigned_script.sh > met_unassigned_script_log.txt 2>&1 &
## 2.2. unassigned copepod code: nohup ./cop_unassigned_script.sh > cop_unassigned_script_log.txt 2>&1 &
# 3. import for blastn in Geneious Prime
# 4. Retrieve top hit. Keep the top hit that is at least 95% grade (explain grade in GP), is not uncultured or other unknown. Keep top hit based on grade
## 4.1 - for those where the top hit is uncultured or unknown, I will look at other top hits to determine.
### 4.1.1. I have 18 of these that I will now redo
### 4.1.2. unassigned retry code: nohup ./unassigned_retry_script.sh > unassigned_retry_script_log.txt 2>&1 &
### 4.1.3. blasting these with 10 hits to see if I get anything (I needed to blast with 100 hits for some of these to get some sort of ID)

taxa_assigned_met <- read.csv("./unassigned_met/geneious_assigned_all.csv")
head(taxa_assigned_met)
dim(taxa_assigned_met) # 67 x 14
unique(taxa_assigned_met$Order)
colnames(taxa_assigned_met)
colnames(taxa_assigned_met)[12] <- "OTU" # rename column for downstream analysis

otu1 <- filter(taxa_assigned_met, Order == "Siphonophorae")
otu1 <- otu1$OTU
otu2 <- filter(taxa_assigned_met, Order == "Calanoida")
otu2 <- otu2$OTU
otu3 <- filter(taxa_assigned_met, Order == "Copelata")
otu3 <- otu3$OTU
otu4 <- filter(taxa_assigned_met, Order == "Enoplida")
otu4 <- otu4$OTU
otu5 <- filter(taxa_assigned_met, Order == "Leptothecata")
otu5 <- otu5$OTU
otu6 <- filter(taxa_assigned_met, Order == "Ploima")
otu6 <- otu6$OTU
otu7 <- filter(taxa_assigned_met, Order == "Spionida")
otu7 <- otu7$OTU
otu8 <- filter(taxa_assigned_met, Order == "Salpida")
otu8 <- otu8$OTU
otu9 <- filter(taxa_assigned_met, Order == "Doliolida")
otu9 <- otu9$OTU
otu10 <- filter(taxa_assigned_met, Order == "Phyllodocida")
otu10 <- otu10$OTU
otu11 <- filter(taxa_assigned_met, Order == "Mormonilloida") 
otu11 <- otu11$OTU
otu12 <- filter(taxa_assigned_met, Order == "Cyclopoida")
otu12 <- otu12$OTU
otu13 <- filter(taxa_assigned_met, Order == "Harpacticoida")
otu13 <- otu13$OTU
otu14 <- filter(taxa_assigned_met, Order == "Scalpellomorpha")
otu14 <- otu14$OTU
otu15 <- filter(taxa_assigned_met, Order == "Decapoda")
otu15 <- otu15$OTU
otu16 <- filter(taxa_assigned_met, Order == "Amphipoda")
otu16 <- otu16$OTU
otu17 <- filter(taxa_assigned_met, Order == "Neogastropoda")
otu17 <- otu17$OTU

met_unknown_rename <- function(df){
  df %>% 
  mutate(Phylum = case_when(
    OTU %in% otu1 ~ "Cnidaria",
    OTU %in% otu2 ~ "Arthropoda",
    OTU %in% otu3 ~ "Tunicata",
    OTU %in% otu4 ~ "Nematoda",
    OTU %in% otu5 ~ "Cnidaria",
    OTU %in% otu6 ~ "Rotifera",
    OTU %in% otu7 ~ "Annelida",
    OTU %in% otu8 ~ "Tunicata",
    OTU %in% otu9 ~ "Tunicata",
    OTU %in% otu10 ~ "Annelida",
    OTU %in% otu11 ~ "Arthropoda",
    OTU %in% otu12 ~ "Arthropoda",
    OTU %in% otu13 ~ "Arthropoda",
    OTU %in% otu14 ~ "Arthropoda",
    OTU %in% otu15 ~ "Arthropoda",
    OTU %in% otu16 ~ "Arthropoda",
    OTU %in% otu17 ~ "Mollusca",
    TRUE ~ Phylum)) %>%
  mutate(Class = case_when(
    OTU %in% otu1 ~ "Hydrozoa",
    OTU %in% otu2 ~ "Copepoda",
    OTU %in% otu3 ~ "Appendicularia",
    OTU %in% otu4 ~ "Enoplea",
    OTU %in% otu5 ~ "Hydrozoa",
    OTU %in% otu6 ~ "Eurotatoria",
    OTU %in% otu7 ~ "Polychaeta",
    OTU %in% otu8 ~ "Thaliacea",
    OTU %in% otu9 ~ "Thaliacea",
    OTU %in% otu10 ~ "Polychaeta",
    OTU %in% otu11 ~ "Copepoda",
    OTU %in% otu12 ~ "Copepoda",
    OTU %in% otu13 ~ "Copepoda",
    OTU %in% otu14 ~ "Thecostraca",
    OTU %in% otu15 ~ "Malacostraca",
    OTU %in% otu16 ~ "Malacostraca",
    OTU %in% otu17 ~ "Gastropoda",
    TRUE ~ Class)) %>%
  mutate(Order = case_when(
    OTU %in% otu1 ~ "Siphonophorae",
    OTU %in% otu2 ~ "Calanoida",
    OTU %in% otu3 ~ "Copelata",
    OTU %in% otu4 ~ "Enoplida",
    OTU %in% otu5 ~ "Leptothecata",
    OTU %in% otu6 ~ "Ploima",
    OTU %in% otu7 ~ "Spionida",
    OTU %in% otu8 ~ "Salpida",
    OTU %in% otu9 ~ "Doliolida",
    OTU %in% otu10 ~ "Phyllodocida",
    OTU %in% otu11 ~ "Mormonilloida",
    OTU %in% otu12 ~ "Cyclopoida",
    OTU %in% otu13 ~ "Harpacticoida",
    OTU %in% otu14 ~ "Scalpellomorpha",
    OTU %in% otu15 ~ "Decapoda",
    OTU %in% otu16 ~ "Amphipoda",
    OTU %in% otu17 ~ "Neogastropoda",
    TRUE ~ Order))
}

```

```{r met-asv-taxa-scan2, eval = TRUE, message = FALSE, warning = FALSE}

met_relabun_rename <- met_unknown_rename(met_relabun)
unique(met_relabun_rename$Order) %>% sort()

met_taxa_order2 <- silva_met_rename_order(met_relabun_rename)
unique(met_taxa_order2$TAXA) %>% sort()

## met taxa scan for TAXA

df_summarized_met <- met_taxa_order2 %>%
  filter(Abundance > 0) %>%
  mutate(depth = as.factor(depth)) %>%
  group_by(depth, rep, deployment, TAXA) %>%
  summarise(n_OTU = n_distinct(OTU))

# check data
head(df_summarized_met) 

met_taxa_abun_check <- met_taxa_order2 %>% select(TAXA, Abundance) %>%
  mutate(sumabun = sum(Abundance)) %>%
  group_by(TAXA) %>%
  mutate(relabun = Abundance / sumabun) %>%
  summarise(sum_RelativeAbundance = sum(relabun) * 100, na.rm = TRUE, .groups = "drop")

head(met_taxa_abun_check)

met_low_abun_taxa <- met_taxa_abun_check %>% filter(sum_RelativeAbundance < 1) 
head(met_low_abun_taxa)
met_low_abun_taxa = met_low_abun_taxa$TAXA
met_low_abun_taxa

met_taxa_order3 <- met_taxa_order2 %>% mutate(TAXA = case_when(
    TAXA %in% met_low_abun_taxa ~ "Other Invertebrates",
    TRUE ~ TAXA))

head(met_taxa_order3)

met_asv <- 
  met_taxa_order3 %>%
  filter(Abundance > 0) %>% # Keep only rows with Abundance > 0
  mutate(depth = as.factor(depth)) %>% # Convert 'depth' to a factor
  
  # Group by 'depth', 'OTU', and 'TAXA', then summarize the Abundance
  group_by(depth, OTU, TAXA) %>%
  summarise(Abund = sum(Abundance)) %>% # Sum the Abundance for each group
  ungroup() %>% # Ungroup the data
  
  # Group by 'depth' and 'TAXA', then calculate species number (sobs)
  group_by(depth, TAXA) %>%
  mutate(sobs = vegan::specnumber(Abund)) %>% # Calculate species number
  
  # Select columns excluding 'OTU' and 'Abund'
  select(-c(OTU, Abund)) %>% # Remove 'OTU' and 'Abund' columns
  unique() %>%
  ungroup() %>% # Ungroup the data
  
  # Group by 'TAXA' and calculate the sum of sobs for each TAXA
  group_by(TAXA) %>%
  mutate(sum_obs = sum(sobs)) %>% # Sum of sobs for each TAXA
  
  # Arrange by 'depth' and descending 'sum_obs', then remove duplicates
  arrange(depth, desc(sum_obs)) %>% # Arrange by depth and sum_obs
  
  # Reorder 'TAXA' factor based on 'sum_obs' values in ascending order
  mutate(TAXA = fct_reorder(TAXA, sum_obs, .desc = FALSE)) # Reorder TAXA


taxa_sorted <- c("Arthropoda-Calanoida", "Arthropoda-Cyclopoida",
                             #"Arthropoda-Other", 
                             "Cnidaria-Siphonophorae",
                          #   "Cnidaria-Other", 
                             "Ctenophora-Cydippida", 
                         #    "Ctenophora-Other",
                             "Tunicata-Copelata", "Tunicata-Salpida", 
                 #"Unassigned Metazoa", 
                 "Other Invertebrates")

taxa_colors <- c("#fc996c", "#ffce5c",
                 #"#fdfb9b", 
                 "#047de9", 
                 #"#0344b7", 
                 "#fa9db8", 
                 #"#fc5d7b",
                 "#96fac6","#1ec38c", 
                  #"#82A4E3", 
                 "#d9d9d9")

met_asv_plot <- met_asv %>%
  ggplot(aes(y=fct_rev(depth), x=sobs)) +
  geom_col(aes(group = depth, fill = TAXA), color = "black", width = 0.78, linewidth = 0.25, alpha=1) +
  labs(title=NULL,
       y="Depth (m)",
       x="Observed ASVs") +
  scale_fill_manual(values = taxa_colors,
                   limits = taxa_sorted) + 
  guides(fill = guide_legend(title = "Invertebrates"),
         color = "none") + theme_custom_paper() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10))

met_asv_plot

#ggsave("./plots-final/met_obs_taxa_barplot.pdf", height = 5, width = 5, units = "in", dpi=300)

met_relabun_asv_plot <- make_taxabar_asv_relabun_collapse(met_asv)
met_relabun_asv_plot

#ggsave("./plots-final/met_obs_taxa_relabun_panel.pdf", height = 5, width = 5, units = "in", dpi=300)
# 
# head(met_asv)
# top_met <- met_asv %>% group_by(depth) %>%
#   mutate(sum_obs = sum(sobs)) %>%
#   mutate(RelativeAbundance = (sobs / sum_obs) * 100, na.rm = TRUE) %>%
#   mutate(depth = as.factor(depth)) %>%
#   ungroup() %>%
#   group_by(depth, TAXA) %>%
#   summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>%
# #  arrange(depth, desc(sum_RelativeAbundance)) %>%
#   mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>%
#   ungroup() %>%
#   group_by(depth) %>%
#   slice_max(order_by = sum_RelativeAbundance, n = 5, with_ties = FALSE) %>%
#   ungroup()
# 
# top_met %>% group_by(depth) %>% mutate(sum = sum(sum_RelativeAbundance)) %>% select(-TAXA, -sum_RelativeAbundance) %>% unique()
# 
# 
# top_met_pub <- top_met %>% ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
#     geom_col(aes(fill = TAXA), color = "black", width = 0.78, linewidth = 0.25) +
#     labs(title = NULL,
#          y = "Depth (m)",
#          x = "Relative Proportion (%)") +
#   scale_x_continuous(limits = c(0, 100)) + 
#     scale_fill_manual(values = taxa_colors,
#                       limits = taxa_sorted) +
#     theme_custom_paper()# Use a manually sorted order for the legend
# 
# top_met_pub

```

```{r fish-asv-taxa-scan, eval = TRUE, message = FALSE, warning = FALSE}

# fix formatting
head(fish_relabun)
dim(fish_relabun) # 992 x 28

# summarize the cleaned dataframe by Kingdom Phylum, and Class, which will help define groupings for renaming and plotting
summary_df <- fish_relabun %>% 
  group_by(Class, Order, Family, Genus) %>%
  summarise(count = n())

unique(fish_relabun$Order)

# loading tables from the unknown sequences identified earlier
load_lookup_table <- function(filepath, key_col = "OTU") {
  df <- read.csv(filepath)
  colnames(df)[12] <- "OTU"
  df <- df %>% select(OTU, Species, Genus, Family, Order)
  return(df)
}
actinopterygii_assigned <- load_lookup_table("./unassigned_vert/geneious_assigned_actinopterygii.csv")
fish_family_assigned <- load_lookup_table("./unassigned_vert/geneious_assigned_fish_family.csv")
carnivora_genus_assigned <- load_lookup_table("./unassigned_vert/geneious_assigned_carnivora_genus.csv")

rename_by_lookup <- function(df, lookup_df) {
  df %>%
    left_join(lookup_df, by = "OTU", suffix = c("", ".lookup")) %>%
    mutate(
      Species = coalesce(Species.lookup, Species),
      Genus = coalesce(Genus.lookup, Genus),
      Family = coalesce(Family.lookup, Family),
      Order = coalesce(Order.lookup, Order)
    ) %>%
    select(-ends_with(".lookup"))
}
fish_relabun1 <- rename_by_lookup(fish_relabun, actinopterygii_assigned)
fish_relabun %>% select(OTU, Class, Order) %>% filter(Class == "Actinopterygii" & is.na(Order)) %>% unique() %>% count() # 279 unknown
fish_relabun1 %>% select(OTU, Class, Order) %>% filter(Class == "Actinopterygii" & is.na(Order)) %>% unique() %>% count() # 222 unknown ## annotated 57 of these (less than 62, but some were removed with rarefaction)

fish_relabun2 <- rename_by_lookup(fish_relabun1, fish_family_assigned)
fish_relabun1 %>% select(OTU, Class, Order, Family) %>% 
  filter(Order %in% c("Perciformes", "Argentiniformes") & is.na(Family)) %>% 
  unique() %>% count() # 64 unknown
fish_relabun2 %>% select(OTU, Class, Order, Family) %>% 
  filter(Order %in% c("Perciformes", "Argentiniformes") & is.na(Family)) %>% 
  unique() %>% count() # 6 unknown # annotated 58

fish_relabun3 <- rename_by_lookup(fish_relabun2, carnivora_genus_assigned)
fish_relabun2 %>% select(OTU, Order, Family, Genus) %>% filter(Order == "Cetartiodactyla" & is.na(Genus)) %>% unique() %>% count() # 2 unknown
fish_relabun3 %>% select(OTU, Order, Family, Genus) %>% filter(Order == "Cetartiodactyla" & is.na(Genus)) %>% unique() %>% count() # 0 unknown # 2 annotated

## 345 were unknown in this category I think
## annotated 117 of these

rename_taxa_fish_order <- function(df){
  df_out <- 
    df %>%
    mutate(TAXA = case_when(
      is.na(Order) & Class == "Actinopterygii" ~ "Unassigned Actinopterygii",
      is.na(Family) & Order == "Perciformes" ~ "Unassigned Perciformes",
      TRUE ~ Family)) %>%
    mutate(ORDER = Order)
  
}

fish_taxa_order <- rename_taxa_fish_order(fish_relabun3)
head(fish_taxa_order)
unique(fish_taxa_order$TAXA)

# going to rename the Cetartiodactyla based on blast reporting from QIIME
df_summarized_fish <- fish_taxa_order %>%
  filter(Abundance > 0) %>%
  mutate(depth = as.factor(depth)) %>%
  group_by(sample, depth, TAXA) %>%
  summarise(n_OTU = n_distinct(OTU)) 

# check data
head(df_summarized_fish)
print(df_summarized_fish)

taxa_sorted <- sort(unique(df_summarized_fish$TAXA))
print(taxa_sorted)

df_summarized_fish <- fish_taxa_order %>%
  filter(Abundance > 0) %>%
  mutate(depth = as.factor(depth)) %>%
  group_by(sample, depth, TAXA) %>%
  summarise(n_OTU = n_distinct(OTU)) 

# check data
head(df_summarized_fish)
print(df_summarized_fish)

unique(df_summarized_fish$TAXA)

# using this chunk of code to calculate what % of the reads from Mammals are in our overall dataset (to show that it is a very small proportion)
fish_taxa_order %>%
  filter(Abundance > 0) %>%
  group_by(depth, OTU, TAXA) %>%
  summarise(Abund = sum(Abundance)) %>% 
  ungroup() %>%
  mutate(TotAbun = sum(Abund)) %>%
  filter(TAXA == "Delphinidae" | TAXA == "Balaenopteridae") %>%
  mutate(relAbun = Abund / TotAbun * 100) %>%
  ungroup() %>%
  summarize(sumRelAbun = sum(relAbun))
  
head(fish_taxa_order)
fish_taxa_order %>% select(Species) %>% unique() %>% print(n=100)


fish_asv <- 
  fish_taxa_order %>%
  filter(Abundance > 0) %>% # Keep only rows with Abundance > 0
  mutate(depth = as.factor(depth)) %>% # Convert 'depth' to a factor
  
  # Group by 'depth', 'OTU', and 'TAXA', then summarize the Abundance
  group_by(depth, OTU, TAXA) %>%
  summarise(Abund = sum(Abundance)) %>% # Sum the Abundance for each group
  ungroup() %>% # Ungroup the data
  
  # Group by 'depth' and 'TAXA', then calculate species number (sobs)
  group_by(depth, TAXA) %>%
  mutate(sobs = vegan::specnumber(Abund)) %>% # Calculate species number
  
  # Select columns excluding 'OTU' and 'Abund'
  select(-c(OTU, Abund)) %>% # Remove 'OTU' and 'Abund' columns
  unique() %>%
  ungroup() %>% # Ungroup the data
  
  # Group by 'TAXA' and calculate the sum of sobs for each TAXA
  group_by(TAXA) %>%
  mutate(sum_obs = sum(sobs)) %>% # Sum of sobs for each TAXA
  
  # Arrange by 'depth' and descending 'sum_obs', then remove duplicates
  arrange(depth, desc(sum_obs)) %>% # Arrange by depth and sum_obs
  
  # Reorder 'TAXA' factor based on 'sum_obs' values in ascending order
  mutate(TAXA = fct_reorder(TAXA, sum_obs, .desc = FALSE)) # Reorder TAXA

taxa_colors <- c("#f4c6e5", "#e886c8", "#DC469F",
                 "#fcada4", "#fa5340",
                 "#fa8016",
                 "#9CDC46", 
                 "#56C4F0", "#0c62fb", 
                 "#ffe9a1", "#fac116", 
                 "#fb7285", "#73eaff",
                 "#a1f1d6", "#37d49f", "#116b4c", "#97d7c1","#5db843","#93e77b" ,
                 "#c31e56", "#1ea9c3",
                 "#d5d2f8", "#948ced", "#564ae3", "#2b1ec3",
                 "#910096",
                 "#cddcf7", "#82A4E3")


fish_asv$TAXA <- fct_relevel(fish_asv$TAXA, "Derichthyidae", "Nemichthyidae", "Serrivomeridae",
                             "Bathylagidae", "Microstomatidae",
                             "Alepisauridae", 
                             "Melamphaidae", 
                             "Balaenopteridae", "Delphinidae", 
                             "Moridae", "Phycidae",
                             "Myctophidae", 
                             "Alepocephalidae", 
                             "Ammodytidae", "Caristiidae", "Coryphaenidae", "Scombridae", "Stromateidae", "Zoarcidae",
                             "Eurypharyngidae",
                             "Gempylidae", 
                             "Gonostomatidae", "Phosichthyidae", "Sternoptychidae", "Stomiidae", "Molidae",
                             "Unassigned Perciformes", "Unassigned Actinopterygii"
                             )



taxa_sorted <- c("Derichthyidae", "Nemichthyidae", "Serrivomeridae",
                             "Bathylagidae", "Microstomatidae",
                             "Alepisauridae", 
                             "Melamphaidae", 
                             "Balaenopteridae", "Delphinidae", 
                             "Moridae", "Phycidae",
                             "Myctophidae", 
                             "Alepocephalidae", 
                             "Ammodytidae", "Caristiidae", "Coryphaenidae", "Scombridae", "Stromateidae", "Zoarcidae",
                             "Eurypharyngidae",
                             "Gempylidae", 
                             "Gonostomatidae", "Phosichthyidae", "Sternoptychidae", "Stomiidae", "Molidae", "Unassigned Perciformes", "Unassigned Actinopterygii"
                 )

fish_asv_plot <- fish_asv %>%
  ggplot(aes(y=fct_rev(depth), x=sobs)) +
  geom_col(aes(group = depth, fill = TAXA), color = "black", width = 0.8, linewidth = 0.5, alpha=1) +
  labs(title=NULL,
       y="Depth (m)",
       x="Observed ASVs") +
  scale_fill_manual(values = taxa_colors,
                   limits = taxa_sorted) + 
  guides(fill = guide_legend(title = "Vertebrates"),
         color = "none") + theme_custom_paper() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10))

fish_asv_plot

#ggsave("./plots-final/fish_obs_taxa_family_barplot.pdf", height = 5, width = 6, units = "in", dpi=300)

# I don't want to plot something with 26 groups especially since most groups have few ASVs. Let's combine.
fish_taxa_abun_check <- fish_taxa_order %>% select(TAXA, Abundance) %>%
  mutate(sumabun = sum(Abundance)) %>%
  group_by(TAXA) %>%
  mutate(relabun = Abundance / sumabun) %>%
  summarise(sum_RelativeAbundance = sum(relabun) * 100, na.rm = TRUE, .groups = "drop")

head(fish_taxa_abun_check)

# filter out low relative abundance (< 1%) for plotting purposes
fish_low_abun_taxa <- fish_taxa_abun_check %>% filter(sum_RelativeAbundance < 1)
head(fish_low_abun_taxa)
fish_low_abun_taxa = fish_low_abun_taxa$TAXA
fish_low_abun_taxa

fish_taxa_order2 <- fish_taxa_order %>% mutate(TAXA = case_when(
    TAXA %in% fish_low_abun_taxa ~ "Other Vertebrates",
    TRUE ~ TAXA))

fish_asv2 <- fish_taxa_order2 %>%
  filter(Abundance > 0) %>% # Keep only rows with Abundance > 0
  mutate(depth = as.factor(depth)) %>% # Convert 'depth' to a factor
  
  # Group by 'depth', 'OTU', and 'TAXA', then summarize the Abundance
  group_by(depth, OTU, TAXA) %>%
  summarise(Abund = sum(Abundance)) %>% # Sum the Abundance for each group
  ungroup() %>% # Ungroup the data
  
  # Group by 'depth' and 'TAXA', then calculate species number (sobs)
  group_by(depth, TAXA) %>%
  mutate(sobs = vegan::specnumber(Abund)) %>% # Calculate species number
  
  # Select columns excluding 'OTU' and 'Abund'
  select(-c(OTU, Abund)) %>% # Remove 'OTU' and 'Abund' columns
  unique() %>%
  ungroup() %>% # Ungroup the data
  
  # Group by 'TAXA' and calculate the sum of sobs for each TAXA
  group_by(TAXA) %>%
  mutate(sum_obs = sum(sobs)) %>% # Sum of sobs for each TAXA
  
  # Arrange by 'depth' and descending 'sum_obs', then remove duplicates
  arrange(depth, desc(sum_obs)) %>% # Arrange by depth and sum_obs
  
  # Reorder 'TAXA' factor based on 'sum_obs' values in ascending order
  mutate(TAXA = fct_reorder(TAXA, sum_obs, .desc = FALSE)) # Reorder TAXA
  
unique(fish_asv2$TAXA)
head(fish_asv2, n = 30)

taxa_colors <- c("#9CDC46",
                 "#d5d2f8", "#9790ee", 
                "#a1f1d6", "#f5eb8e",
                  "#56C4F0", "#367eff", "#034dd1","#fb7285", "#d9d9d9")

fish_asv2$TAXA <- fct_relevel(fish_asv2$TAXA, 
                              "Alepocephalidae", 
                              "Nemichthyidae", "Serrivomeridae",
                              "Melamphaidae", 
                              "Myctophidae", 
                              "Gonostomatidae", "Sternoptychidae", "Stomiidae", 
                              "Molidae",
                              "Other Vertebrates"
                              )

taxa_sorted <- c("Alepocephalidae", 
                 "Nemichthyidae", "Serrivomeridae",
                 "Melamphaidae", 
                 "Myctophidae", 
                 "Gonostomatidae", "Sternoptychidae", "Stomiidae", 
                 "Molidae",
                 "Other Vertebrates"
                 )


fish_relabun_asv_plot <- make_taxabar_asv_relabun_collapse(fish_asv2)
fish_relabun_asv_plot

#ggsave("./plots-final/fish_obs_taxa_family_relabun.pdf", height = 5, width = 5, units = "in", dpi=300)

fish_asv_plot <- fish_asv2 %>% select(depth, TAXA, sobs) %>%
  unique() %>%
  ggplot(aes(y=fct_rev(depth), x=sobs)) +
  geom_col(aes(group = depth, fill = TAXA), color = "black", width = 0.78, linewidth = 0.25, alpha=1) +
  labs(title=NULL,
       y="Depth (m)",
       x="Observed ASVs") +
  scale_fill_manual(values = taxa_colors,
                   limits = taxa_sorted) + 
  guides(fill = guide_legend(title = "Vertebrates"),
         color = "none") + theme_custom_paper() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10))

fish_asv_plot

#ggsave("./plots-final/fish_obs_taxa_family_barplot_simplified.pdf", height = 5, width = 5, units = "in", dpi=300)
# 
# head(fish_asv2)
# 
# top_fish <- fish_asv2 %>% group_by(depth) %>%
#   mutate(sum_obs = sum(sobs)) %>%
#   mutate(RelativeAbundance = (sobs / sum_obs) * 100, na.rm = TRUE) %>%
#   mutate(depth = as.factor(depth)) %>%
#   ungroup() %>%
#   group_by(depth, TAXA) %>%
#   summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>%
# #  arrange(depth, desc(sum_RelativeAbundance)) %>%
#   mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>%
#   ungroup() %>%
#   group_by(depth) %>%
#   slice_max(order_by = sum_RelativeAbundance, n = 5, with_ties = FALSE) %>%
#   ungroup()
# 
# top_fish
# 
# top_fish %>% group_by(depth) %>% mutate(sum = sum(sum_RelativeAbundance)) %>% select(-TAXA, -sum_RelativeAbundance) %>% unique()
# 
# top_fish_pub <- top_fish %>% ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
#     geom_col(aes(fill = TAXA), color = "black", width = 0.78, linewidth = 0.25) +
#     labs(title = NULL,
#          y = "Depth (m)",
#          x = "Relative Proportion (%)") +
#   scale_x_continuous(limits = c(0, 100)) + 
#     scale_fill_manual(values = taxa_colors,
#                       limits = taxa_sorted) +
#     theme_custom_paper()# Use a manually sorted order for the legend
# 
# top_fish_pub

```

```{r save-load-asv-files, eval = TRUE, message = FALSE, warning = FALSE}

# save files to avoid redoing work
save(prot_asv, met_asv, fish_asv, fish_asv2, file = "./asv_tables_alltaxa_for_plotting.RData")

load(file = "./asv_tables_alltaxa_for_plotting.RData", verbose = T)

```

```{r combined-asv-proportion, eval = TRUE, message = FALSE, warning = FALSE}

## Relative Proportion (%)
prot_relabun_asv_pub <- prot_relabun_asv_plot + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Protists") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

prot_relabun_asv_pub

met_relabun_asv_pub <- met_relabun_asv_plot + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Invertebrates") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))
met_relabun_asv_pub

fish_relabun_asv_pub <- fish_relabun_asv_plot + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Vertebrates") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

fish_relabun_asv_pub

(prot_relabun_asv_pub + met_relabun_asv_pub + fish_relabun_asv_pub) + plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect') & 
  theme(plot.margin = unit(c(t = 1, r = 1, b = 1, l = 1), "lines"),
        plot.title = element_text(size = 8),
        plot.tag = element_text(size = 8),
        panel.spacing = unit(5, "mm"),
        legend.text = element_text(size = 6, margin = margin(l = 2)),
        legend.position = "bottom", 
        legend.spacing.x = unit(7, "mm"), 
        legend.margin = margin(t = -4, r = 0, b = 0, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 6),
        axis.title = element_text(size = 8))

ggsave("./plots-final/alleuk_asv_relabun_barplot.pdf", height = 100, width = 180, units = "mm", dpi=300)

```

```{r combined-asv-observed, eval = TRUE, message = FALSE, warning = FALSE}

prot_asv_pub <- prot_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Protists") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

prot_asv_pub

met_asv_pub <- met_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Invertebrates") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))
met_asv_pub

fish_asv_pub <- fish_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Vertebrates") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

fish_asv_pub

(prot_asv_pub + met_asv_pub + fish_asv_pub) + plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect') & 
  theme(plot.margin = unit(c(t = 1, r = 1, b = 1, l = 1), "lines"),
        plot.title = element_text(size = 8),
        plot.tag = element_text(size = 8),
        panel.spacing = unit(5, "mm"),
        legend.text = element_text(size = 6, margin = margin(l = 2)),
        legend.position = "bottom", 
        legend.spacing.x = unit(7, "mm"), 
        legend.margin = margin(t = -4, r = 0, b = 0, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 6),
        axis.title = element_text(size = 8))

ggsave("./plots-final/alleuk_obs_asv_barplot.pdf", height = 100, width = 180, units = "mm", dpi=300)

```

```{r asv-figure-for-pub eval = TRUE, message = FALSE, warning = FALSE}

# these have no legend so we can combine all the asv plots into one
prot_asv_pub <- prot_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Protists") + 
  theme(legend.position = "none") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

prot_asv_pub

met_asv_pub <- met_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Invertebrates") + 
  theme(legend.position = "none") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))
met_asv_pub

fish_asv_pub <- fish_asv_plot + labs(x = "Observed ASVs",
       y = "Depth (m)", 
       title = "Vertebrates") + 
  theme(legend.position = "none") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2))

fish_asv_pub


(prot_asv_pub + met_asv_pub + fish_asv_pub) / (prot_relabun_asv_pub + met_relabun_asv_pub + fish_relabun_asv_pub) + 
    plot_layout(guides = 'collect') + plot_annotation(tag_levels = "A", tag_suffix = '.')  & 
  theme(plot.margin = unit(c(t = 1, r = 1, b = 1, l = 1), "lines"),
        plot.title = element_text(size = 8),
        plot.tag = element_text(size = 8),
        panel.spacing = unit(3, "mm"),
        legend.text = element_text(size = 6, margin = margin(l = 2)),
        legend.position = "bottom", 
        legend.spacing.x = unit(7, "mm"), 
        legend.margin = margin(t = -4, r = 0, b = 0, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text.x = element_text(size = 6),
        axis.text.y = element_text(size = 6),
        axis.title = element_text(size = 8))

ggsave("./plots-final/alleuk_asv_obs_relabun_barplot_pub.pdf", height = 180, width = 180, units = "mm", dpi=300)

```

```{r alpha-stats-anova, eval = TRUE, message = FALSE, warning = FALSE}

# Protist
head(alpha_div_indice_df)
prot_stats <- alpha_div_indice_df %>% 
  filter(Group == "Protists") %>%
  as.data.frame()

head(prot_stats)
class(prot_stats)
dim(prot_stats)

# extract Protist data out

# set up dataframe
rownames(prot_stats) <- prot_stats$sample
prot_stats$depth <- as.factor(prot_stats$depth)
prot_stats$deployment <- as.factor(prot_stats$deployment)
head(prot_stats)

# Observed
res_aov <- aov(obs ~ depth, data = prot_stats)
summary(res_aov)
hist(res_aov$residuals)

prot_aov_stats <- tidy(res_aov)
prot_aov_stats
prot_aov_stats$taxa <- "protists"
prot_aov_stats$stats <- "obs"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
prot_tukey_sig <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
prot_tukey_sig
prot_tukey_sig$taxa <- "protists"
prot_tukey_sig$stats <- "obs"
prot_tukey_sig

# Shannon
res_aov <- aov(shannon ~ depth, data = prot_stats)
summary(res_aov)
hist(res_aov$residuals)
prot_aov_stats_sha <- tidy(res_aov)
prot_aov_stats_sha
prot_aov_stats_sha$taxa <- "protists"
prot_aov_stats_sha$stats <- "shannon"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
prot_tukey_sig_sha <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
prot_tukey_sig_sha
prot_tukey_sig_sha$taxa <- "protists"
prot_tukey_sig_sha$stats <- "shannon"
prot_tukey_sig_sha

# Evenness
res_aov <- aov(evenness ~ depth, data = prot_stats)
summary(res_aov)
hist(res_aov$residuals)

prot_aov_stats_even <- tidy(res_aov)
prot_aov_stats_even
prot_aov_stats_even$taxa <- "protists"
prot_aov_stats_even$stats <- "evenness"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
prot_tukey_sig_even <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
prot_tukey_sig_even
prot_tukey_sig_even$taxa <- "protists"
prot_tukey_sig_even$stats <- "evenness"
prot_tukey_sig_even


# Metazoa
head(alpha_div_indice_df)
met_stats <- alpha_div_indice_df %>% 
  filter(Group == "Metazoa") %>%
  as.data.frame()

head(met_stats)
class(met_stats)
dim(met_stats)

# extract Metazoa data out

# set up dataframe
rownames(met_stats) <- met_stats$sample
met_stats$depth <- as.factor(met_stats$depth)
met_stats$deployment <- as.factor(met_stats$deployment)
head(met_stats)

# Observed
res_aov <- aov(obs ~ depth, data = met_stats)
summary(res_aov)
hist(res_aov$residuals)

met_aov_stats <- tidy(res_aov)
met_aov_stats
met_aov_stats$taxa <- "metazoa"
met_aov_stats$stats <- "obs"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
met_tukey_sig <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
met_tukey_sig
met_tukey_sig$taxa <- "metazoa"
met_tukey_sig$stats <- "obs"
met_tukey_sig

# Shannon
res_aov <- aov(shannon ~ depth, data = met_stats)
summary(res_aov)
hist(res_aov$residuals)
met_aov_stats_sha <- tidy(res_aov)
met_aov_stats_sha
met_aov_stats_sha$taxa <- "metazoa"
met_aov_stats_sha$stats <- "shannon"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
met_tukey_sig_sha <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
met_tukey_sig_sha
met_tukey_sig_sha$taxa <- "metazoa"
met_tukey_sig_sha$stats <- "shannon"
met_tukey_sig_sha

# Evenness
res_aov <- aov(evenness ~ depth, data = met_stats)
summary(res_aov)
hist(res_aov$residuals)

met_aov_stats_even <- tidy(res_aov)
met_aov_stats_even
met_aov_stats_even$taxa <- "metazoa"
met_aov_stats_even$stats <- "evenness"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
met_tukey_sig_even <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
met_tukey_sig_even
met_tukey_sig_even$taxa <- "metazoa"
met_tukey_sig_even$stats <- "evenness"
met_tukey_sig_even


# Fish
head(alpha_div_indice_df)
fish_stats <- alpha_div_indice_df %>% 
  filter(Group == "Fish") %>%
  as.data.frame()

head(fish_stats)
class(fish_stats)
dim(fish_stats)

# extract Fish data out

# set up dataframe
rownames(fish_stats) <- fish_stats$sample
fish_stats$depth <- as.factor(fish_stats$depth)
fish_stats$deployment <- as.factor(fish_stats$deployment)
head(fish_stats)

# Observed
res_aov <- aov(obs ~ depth, data = fish_stats)
summary(res_aov)
hist(res_aov$residuals)

fish_aov_stats <- tidy(res_aov)
fish_aov_stats
fish_aov_stats$taxa <- "fish"
fish_aov_stats$stats <- "obs"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
fish_tukey_sig <- tukey %>% tidy() %>% filter(adj.p.value < 0.1)
fish_tukey_sig
fish_tukey_sig$taxa <- "fish"
fish_tukey_sig
fish_tukey_sig$stats <- "obs"

# Shannon
res_aov <- aov(shannon ~ depth, data = fish_stats)
summary(res_aov)
hist(res_aov$residuals)
fish_aov_stats_sha <- tidy(res_aov)
fish_aov_stats_sha
fish_aov_stats_sha$taxa <- "fish"
fish_aov_stats_sha$stats <- "shannon"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
fish_tukey_sig_sha <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
fish_tukey_sig_sha
fish_tukey_sig_sha$taxa <- "fish"
fish_tukey_sig_sha$stats <- "shannon"
fish_tukey_sig_sha

# Evenness
res_aov <- aov(evenness ~ depth, data = fish_stats)
summary(res_aov)
hist(res_aov$residuals)

fish_aov_stats_even <- tidy(res_aov)
fish_aov_stats_even
fish_aov_stats_even$taxa <- "fish"
fish_aov_stats_even$stats <- "evenness"

tukey <- TukeyHSD(x = res_aov, ordered = TRUE)
tukey
fish_tukey_sig_even <- tukey %>% tidy() %>% filter(adj.p.value < 0.05)
fish_tukey_sig_even
fish_tukey_sig_even$taxa <- "fish"
fish_tukey_sig_even$stats <- "evenness"
fish_tukey_sig_even

# let's combine these into a table to export 
met_aov <- rbind(met_aov_stats, met_aov_stats_sha, met_aov_stats_even)
met_tukey <- rbind(met_tukey_sig, met_tukey_sig_sha, met_tukey_sig_even)

prot_aov <- rbind(prot_aov_stats, prot_aov_stats_sha, prot_aov_stats_even)
prot_tukey <- rbind(prot_tukey_sig, prot_tukey_sig_sha, prot_tukey_sig_even)

fish_aov <- rbind(fish_aov_stats, fish_aov_stats_sha, fish_aov_stats_even)
fish_tukey <- rbind(fish_tukey_sig, fish_tukey_sig_sha, fish_tukey_sig_even)


write.csv(met_aov, file = "./stats/aov_stats_metazoa.csv")
write.csv(met_tukey, file = "./stats/aov_tukey_metazoa.csv")

write.csv(prot_aov, file = "./stats/aov_stats_protist.csv")
write.csv(prot_tukey, file = "./stats/aov_tukey_protist.csv")

write.csv(fish_aov, file = "./stats/aov_stats_fish.csv")
write.csv(fish_tukey, file = "./stats/aov_tukey_fish.csv")

```

# 3. Beta Diversity 
## 3.1 Relative Abundance
```{r relabun-setup, eval = TRUE, message = FALSe, warning = FALSE}

# functions
make_taxabar_relabun_agg <- function(df) {
  df %>%
    group_by(depth) %>%
    mutate(RelativeAbundance = (Abundance / sum(Abundance)) * 100, na.rm = TRUE) %>%
    mutate(depth = as.factor(depth)) %>%
    ungroup() %>%
    group_by(depth, TAXA) %>%
    summarise(sum_RelativeAbundance = sum(RelativeAbundance), na.rm = TRUE, .groups = "drop") %>%
    arrange(depth, desc(sum_RelativeAbundance)) %>%
    mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>%
    ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + theme_custom_paper()
}

make_taxabar_relabun_deploy <- function(df) {
  df %>%
    group_by(depth, deployment) %>%
    mutate(RelativeAbundance = (Abundance / sum(Abundance)) * 100, na.rm = TRUE) %>%
    mutate(depth = as.factor(depth)) %>%
    ungroup() %>%
    group_by(depth, deployment, TAXA) %>%
    summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(depth, desc(sum_RelativeAbundance)) %>%
    mutate(TAXA = fct_reorder(TAXA, sum_RelativeAbundance, .desc = FALSE)) %>%
    ggplot(aes(y = fct_rev(depth), x = sum_RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.8, linewidth = 0.5) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    facet_wrap(. ~ deployment) + 
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) +
    theme_custom_paper()# Use a manually sorted order for the legend
}

```

```{r relabun-plots-agg, eval = TRUE, message = FALSE, warning = FALSE}

unique(prot_taxa2$TAXA)

prot_taxa2$TAXA <- fct_relevel(prot_taxa2$TAXA, "Other Protists", after = Inf)
taxa_sorted <- c("Alveolata-Apicomplexa", 
                 "Alveolata-Ciliates",
                 "Alveolata-Dinoflagellates", 
                 # "Alveolata-Other", 
                 "Alveolata-Syndiniales", 
                 "Chlorophyta", 
                 #  "Cryptophyta", 
                 "Excavata",
                 "Haptophyta", #"Rhizaria-Cercozoa", 
                 # "Rhizaria-Other", 
                 #"Obazoa-Opisthokonta",
                 "Rhizaria-Radiolaria", # "Rhodophyta",
                 "Stramenopiles-Diatoms", "Stramenopiles-MAST", "Stramenopiles-Other", "Other Protists")

taxa_colors <- c("#a1c2fd", "#56C4F0", "#399dff",
                 #"#0b62fb", 
                 "#0344b7",
                 "#9CDC46",#"#dafa0f",
                 "#efdf48", "#ff9aa5",
                 #"#ceceff", "#a7a7ff",
                 #"#ff9d3a",
                 "#7474ff",
                 #"#ff9aa5",
                 "#009a4d","#1ec38c", "#90eebf",#"#82A4E3", 
                 "#d9d9d9")

head(prot_taxa2)
dim(prot_taxa2)
unique(prot_taxa2$TAXA) # 12 taxa

prot_taxa2$TAXA <- fct_relevel(prot_taxa2$TAXA, "Other Protists", after = Inf)

prot_relabun_agg <- make_taxabar_relabun_agg(prot_taxa2) + 
  theme(legend.position = "bottom") +
      labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Protists (18S)") + 
  guides(fill = guide_legend(title = NULL,
                             ncol = 2,
                             override.aes = list(stroke = 1)))
    
prot_relabun_agg 

#ggsave("./plots-final/prot_relabun_agg_barplot.pdf", height = 5, width = 3, units = "in", dpi=300)

# just to see how relative abundance differs between casts
prot_relabun_deploy <- make_taxabar_relabun_deploy(prot_taxa2) + 
  theme(legend.position = "right") +
      labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Protists") + 
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))
    
prot_relabun_deploy 

length(unique(prot_taxa2$OTU)) #7046

#ggsave("./plots-final/prot_relabun_deploy_barplot.pdf", height = 4, width = 9, units = "in", dpi=300)

# Invertebrates

taxa_sorted <- c("Arthropoda-Calanoida", "Arthropoda-Cyclopoida",
                             #"Arthropoda-Other", 
                             "Cnidaria-Siphonophorae",
                          #   "Cnidaria-Other", 
                             "Ctenophora-Cydippida", 
                         #    "Ctenophora-Other",
                             "Tunicata-Copelata", "Tunicata-Salpida", 
                 #"Unassigned Metazoa", 
                 "Other Invertebrates")

taxa_colors <- c("#fc996c", "#ffce5c",
                 #"#fdfb9b", 
                 "#047de9", 
                 #"#0344b7", 
                 "#fa9db8", 
                 #"#fc5d7b",
                 "#96fac6","#1ec38c", 
                  #"#82A4E3", 
                 "#d9d9d9")

met_relabun_agg <- make_taxabar_relabun_agg(met_taxa_order3) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Invertebrates (18S)") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 2,
                             override.aes = list(stroke = 1)))
met_relabun_agg

#ggsave("./plots-final/met_relabun_agg_barplot.pdf", height = 5, width = 4, units = "in", dpi=300)

met_relabun_deploy <- make_taxabar_relabun_deploy(met_taxa_order3) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Invertebrates") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))
met_relabun_deploy

#ggsave("./plots-final/met_relabun_deploy_barplot.pdf", height = 4, width = 9, units = "in", dpi=300)

length(unique(met_taxa_order3$OTU)) # 399

# let's assess the unassigned portions
unassigned_check_met <- filter(met_taxa_order, TAXA == "Arthropoda-Other" | TAXA == "Unassigned Metazoa") %>% select(sample, OTU, Abundance, depth, Phylum, Order, TAXA)
dim(unassigned_check_met)
unassigned_check_met <- unassigned_check_met %>% filter(Order %in% c("Unassigned", "uncultured"))
dim(unassigned_check_met)
head(unassigned_check_met)
tail(unassigned_check_met)
unassigned_check_met$Order %>% unique()
unassigned_check_met$OTU %>% unique() # 142 ASVs are unassigned (as expected)

# let's figure out what % of sequences in the entire dataset is "unassigned" and what the dominant ASVs are
sum_seqs <- met_taxa_order %>% select(n_seqs) %>% # n_seqs is the total number of sequences for each sample 
  unique() %>%
  summarize(sum_seqs = sum(n_seqs)) %>% # note that sum_seqs is the total # of sequneces by depth
  print()

sum_abun_unassigned <- unassigned_check_met %>%
  select(OTU, sample, Abundance) %>% 
  summarize(SumAbundance = sum(Abundance)) %>% 
  print()

print((sum_abun_unassigned$SumAbundance / sum_seqs$sum_seqs) * 100) # this is the total % of the dataset that was unassigned 18.42325%

sum_abun_assigned_met <- taxa_assigned_met %>% select(OTU, Order) %>%
  left_join(unassigned_check_met, by = "OTU") %>%
  select(OTU, sample, Abundance) %>% 
  summarize(SumAbundance = sum(Abundance)) %>% 
  print()

print((sum_abun_assigned_met$SumAbundance / sum_seqs$sum_seqs) * 100) # this is the total % of the dataset that got assigned 18.07%

otu_assigned_check <- taxa_assigned_met %>% select(OTU, Order) %>%
  left_join(unassigned_check_met, by = "OTU") %>%
  select(OTU, Order = Order.x, Abundance) %>%
  group_by(OTU) %>%
  mutate(SumAbundance = sum(Abundance)) %>%
  select(-Abundance) %>% unique() %>%
  mutate(RelativeAbundance = (SumAbundance / sum_seqs) * 100, na.rm = TRUE)

# sum_seqs
depth_seqs <- met_taxa_order %>% select(depth, n_seqs) %>%
  unique() %>%
  group_by(depth) %>%
  summarize(sum_depth_seqs = sum(n_seqs)) # note that sum_seqs is the total # of seqneces by depth

head(depth_seqs)

#let's see what % of sequences were recovered
# Set the option to disable scientific notation
options(scipen = 999)

depth_assigned_check <- taxa_assigned_met %>%
  select(OTU, Organism, Order) %>%
  left_join(unassigned_check_met, by = "OTU") %>%
  left_join(depth_seqs, by = "depth") %>%
  select(OTU, Order = Order.x, Abundance, depth, sum_depth_seqs) %>%
  group_by(OTU, depth) %>%
  mutate(SumAbundance = sum(Abundance, na.rm = TRUE)) %>%
  reframe(
    Order = first(Order),
    RelativeAbundance = (first(SumAbundance) / first(sum_seqs)) * 100
  )

head(depth_assigned_check)
dim(depth_assigned_check)

total_added <- depth_assigned_check %>% 
  group_by(depth) %>%
  summarize(totaladded = sum(RelativeAbundance))
total_added

total_added_bytaxa <- depth_assigned_check %>% 
  group_by(depth, Order) %>%
  summarize(taxa_added = sum(RelativeAbundance))
total_added_bytaxa

# for plotting:
head(unassigned_check_met)
head(depth_assigned_check)

taxa_assigned_plot <- unassigned_check_met %>%
  select(OTU, depth, Order, TAXA, Abundance) %>%
  left_join(depth_assigned_check, by = c("OTU", "depth")) %>%
  select(-RelativeAbundance)

taxa_assigned_plot1 <- taxa_assigned_plot %>%
    mutate(Phylum = case_when(
      Order.y == "Amphipoda" ~ "Arthropoda",
      Order.y == "Calanoida" ~ "Arthropoda",
      Order.y == "Copelata" ~ "Tunicata",
      Order.y == "Cyclopoida" ~ "Arthropoda",
      Order.y == "Decapoda" ~ "Arthropoda",
      Order.y == "Doliolida" ~ "Tunicata",
      Order.y == "Enoplida" ~ "Nematoda",
      Order.y == "Harpacticoida" ~ "Arthropoda",
      Order.y == "Leptothecata" ~ "Cnidaria",
      Order.y == "Mormonilloida" ~ "Arthropoda",
      Order.y == "Phyllodocida" ~ "Annelida",
      Order.y == "Ploima" ~ "Rotifera",
      Order.y == "Salpida" ~ "Tunicata",
      Order.y == "Scalpellomorpha" ~ "Arthropoda",
      Order.y == "Siphonophorae" ~ "Cnidaria",
      Order.y == "Spionida" ~ "Annelida",
      Order.y == "Neogastropoda" ~ "Mollusca",
      TRUE ~ TAXA)) %>%
  select(OTU, depth, Phylum, Order = Order.y, Abundance)

taxa_assigned_plot1 %>% select(OTU, Phylum) %>% unique() # 142 (as expected)
taxa_assigned_plot1 %>% select(Phylum) %>% unique() # 142 (as expected)

taxa_assigned_plot2 <- silva_met_rename_order(taxa_assigned_plot1) %>%
  mutate(TAXA = case_when(
    Phylum == "Unassigned Metazoa" ~ "Unassigned Metazoa",
    Phylum == "Arthropoda-Other" ~ "Arthropoda-Unassigned",
    Order == "Scalpellomorpha" ~ "Arthropoda-Scalpellomorpha",
    TRUE ~ TAXA))

taxa_assigned_plot3 <- taxa_assigned_plot2 %>%
  group_by(depth, TAXA) %>%
  summarize(SumAbundance = sum(Abundance)) %>%
  left_join(depth_seqs, by = "depth") %>%
  mutate(RelativeAbundance = (SumAbundance / sum_depth_seqs) * 100)

taxa_assigned_plot3$TAXA %>% unique() %>% sort()

# let's assign other inverts
other_arthropoda <- c("Arthropoda-Amphipoda", "Arthropoda-Decapoda", "Arthropoda-Harpacticoida", "Arthropoda-Mormonilloida", "Arthropoda-Scalpellomorpha")
other_invert <- c("Annelida", "Mollusca", "Nematoda", "Rotifera")

taxa_assigned_plot3 <- taxa_assigned_plot3 %>% 
  mutate(TAXA = case_when(
    TAXA %in% other_arthropoda ~ "Arthropoda-Other",
    TAXA %in% other_invert ~ "Other Invertebrates",
    TRUE ~ TAXA))

taxa_assigned_plot3$TAXA %>% unique()

taxa_sorted <- c("Arthropoda-Calanoida", "Arthropoda-Cyclopoida","Arthropoda-Other", "Arthropoda-Unassigned",
                 "Cnidaria-Leptothecata", "Cnidaria-Siphonophorae",
                 "Tunicata-Copelata", "Tunicata-Doliolida", "Tunicata-Salpida",
                 "Other Invertebrates", "Unassigned Metazoa"
                 )

taxa_colors <- c("#fc996c", "#ffce5c","#fdfb9b", "#FDFBD4",
                 "#0344b7", "#047de9",
                 "#96fac6","#2E6F40", "#1ec38c",
                 "#d9d9d9", "#82A4E3"
                 
                 )

assigned_met <- taxa_assigned_plot3 %>%
    ggplot(aes(y = fct_rev(as.factor(depth)), x = RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + 
   scale_x_continuous(limits = c(0, 100)) + theme_custom_paper()

assigned_met

assigned_met <- assigned_met + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Invertebrates - After Assignment") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))

assigned_met

head(unassigned_check_met)
taxa_unassigned_plot <- unassigned_check_met %>%
  mutate(TAXA = case_when(
    TAXA == "Arthropoda-Other" ~ "Arthropoda-Unassigned",
    TRUE ~ TAXA)) %>%
  select(OTU, Abundance, depth, TAXA) %>%
  group_by(depth, TAXA) %>%
  summarize(SumAbundance = sum(Abundance)) %>%
  left_join(depth_seqs, by = "depth") %>%
  mutate(RelativeAbundance = (SumAbundance / sum_depth_seqs) * 100)

taxa_sorted <- c("Arthropoda-Unassigned", "Unassigned Metazoa")

taxa_colors <- c("#FDFBD4", "#82A4E3")

unassigned_met <- taxa_unassigned_plot %>%
    ggplot(aes(y = fct_rev(as.factor(depth)), x = RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + 
   scale_x_continuous(limits = c(0, 100)) + theme_custom_paper()

unassigned_met

unassigned_met <- unassigned_met + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Invertebrates - Unassigned") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))

unassigned_met


unassigned_met + assigned_met &
#  plot_layout(guides = 'collect') &
  theme(plot.margin = unit(c(t = 0.5, r = 0.5, b = 0.5, l = 0.5), "lines"),
        panel.spacing = unit(5, "mm"),
        panel.background = element_rect(linewidth = 0.5),
        plot.title = element_text(size = 10),
        plot.tag = element_text(size = 8),
        legend.text = element_text(size = 8, margin = margin(l = 2)),
        legend.position = "right", 
        legend.spacing.x = unit(4, "mm"), 
        legend.margin = margin(t = 0, r = 0, b = 50, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 8))

ggsave("./plots-final/met_unassigned_relabun_ponel_barplot_v2.pdf", height = 3, width = 7.5, units = "in", dpi=300)

head(fish_taxa_order2)

unique(fish_taxa_order2$TAXA)
dim(fish_taxa_order2)
unique(fish_taxa_order2$TAXA) # 10 taxa

fish_taxa_order2$TAXA <- fct_relevel(fish_taxa_order2$TAXA, 
                                     "Alepocephalidae",
                                     "Nemichthyidae", "Serrivomeridae",
                                     "Melamphaidae", 
                                     "Myctophidae",
                                     "Gonostomatidae", "Sternoptychidae", 
                                     "Stomiidae", 
                                     "Molidae", 
                                     "Other Vertebrates")

taxa_sorted <- c("Alepocephalidae",
                 "Nemichthyidae", "Serrivomeridae",
                 "Melamphaidae", 
                 "Myctophidae",
                 "Gonostomatidae", "Sternoptychidae", 
                 "Stomiidae", 
                 "Molidae", 
                 "Other Vertebrates")


taxa_colors <- c("#9CDC46",
                 "#d5d2f8", "#9790ee", 
                 "#a1f1d6",
                 "#f5eb8e",
                 "#56C4F0", "#367eff", "#034dd1","#fb7285", "#d9d9d9")



fish_relabun_agg <- make_taxabar_relabun_agg(fish_taxa_order2) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Vertebrates (12S)") + 
  theme(legend.position = "bottom") +
  guides(fill = guide_legend(title = NULL,
                             nrow = 5,
                             override.aes = list(stroke = 1)))

fish_relabun_agg

#ggsave("./plots-final/fish_relabun_agg_barplot.pdf", height = 5, width = 3, units = "in", dpi=300)

fish_relabun_deploy <- make_taxabar_relabun_deploy(fish_taxa_order2) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Vertebrates") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))

fish_relabun_deploy
#ggsave("./plots-final/fish_relabun_deploy_barplot.pdf", height = 4, width = 9, units = "in", dpi=300)

length(unique(fish_taxa_order2$OTU)) # 667

# fish_taxa_order is assigned
# fish_relabun is unassigned

# let's assess the unassigned portions
unassigned_check1 <- fish_relabun %>%
  filter(Class == "Actinopterygii") %>%
  filter(is.na(Order))
unique(unassigned_check1$OTU) # 279 (as expected)

unassigned_check2 <- fish_relabun %>% filter(Order %in% c("Perciformes", "Argentiniformes") & is.na(Family))
unique(unassigned_check2$OTU) # 64 (as expected)

unassigned_check3 <- fish_relabun %>% filter(Order == "Cetartiodactyla" & is.na(Genus))
unique(unassigned_check3$OTU) # 2 (as expected)

unassigned_check_vert <- rbind(unassigned_check1, unassigned_check2, unassigned_check3) %>%   
  select(sample, OTU, depth, Abundance, Class, Order, Family, Genus, Species)
dim(unassigned_check_vert) # 505 x 9
head(unassigned_check_vert)
tail(unassigned_check_vert)
unassigned_check_vert$OTU %>% unique() # 345 ASVs are unassigned as expected

# let's figure out what % of sequences in the entire dataset is "unassigned" and what the dominant ASVs are
sum_seqs <- fish_relabun %>% select(n_seqs) %>% # n_seqs is the total number of sequences for each sample 
  unique() %>%
  summarize(sum_seqs = sum(n_seqs)) %>% # note that sum_seqs is the total # of sequneces by depth
  print()

sum_abun_unassigned <- unassigned_check_vert %>%
  select(OTU, sample, Abundance) %>% 
  summarize(SumAbundance = sum(Abundance)) %>% 
  print()

print((sum_abun_unassigned$SumAbundance / sum_seqs$sum_seqs) * 100) # this is the total % of the dataset that was unassigned 12.88%

taxa_assigned_vert <- rbind(actinopterygii_assigned, fish_family_assigned, carnivora_genus_assigned)
head(taxa_assigned_vert)
taxa_assigned_vert$OTU %>% unique() # 152 - there are 22 of these that are not relevant because they are contaminants 
taxa_assigned_vert <- taxa_assigned_vert %>% filter(Order != "Contaminant") %>% filter(Genus!= "Contaminant")
dim(taxa_assigned_vert) # 130 x 2 assigned ASVs (unique) - removed the 22 contaminants which are not in consideration (the "assigned" blast files were generated before I finished decontaminating)
taxa_assigned_vert$OTU %>% unique() # 130 as expected -- I think these are ASVs that were assigned but not all are part of the dataset given rarefaction

sum_abun_assigned <- taxa_assigned_vert %>% select(OTU, Order) %>% #I want a list of assigned sequences so I can compare it against the unassigned sequences
  left_join(unassigned_check_vert, by = "OTU") %>%
  select(OTU, sample, Abundance) %>% 
  filter(!is.na(sample)) %>%
  summarize(SumAbundance = sum(Abundance)) %>% 
  print()

print((sum_abun_assigned$SumAbundance / sum_seqs$sum_seqs) * 100) # this is the total % of the dataset that got assigned 11.93%

otu_assigned_check <- taxa_assigned_vert %>% select(OTU, Family) %>%
  left_join(unassigned_check_vert, by = "OTU") %>%
  select(OTU, Order = Order, Abundance) %>%
  group_by(OTU) %>%
  mutate(SumAbundance = sum(Abundance)) %>%
  select(-Abundance) %>% unique() %>%
  mutate(RelativeAbundance = (SumAbundance / sum_seqs) * 100, na.rm = TRUE) %>%
  filter(!is.na(SumAbundance)) # 117 x 5

otu_assigned_check$OTU %>% unique() # 117 unique ASVs (out of the 345 unassigned) as expected...

# sum_seqs
depth_seqs <- fish_taxa_order %>% select(depth, n_seqs) %>%
  unique() %>%
  group_by(depth) %>%
  summarize(sum_depth_seqs = sum(n_seqs)) # note that sum_seqs is the total # of seqneces by depth

head(depth_seqs)

#let's see what % of sequences were recovered
# Set the option to disable scientific notation
options(scipen = 999)

# let's filter taxa_assigned by the 117 ASVs that we actually want

asvs_to_keep <- otu_assigned_check$OTU %>% unique()
taxa_assigned_filtered <- taxa_assigned_vert %>%
  filter(OTU %in% asvs_to_keep)

depth_assigned_check <- taxa_assigned_filtered %>%
  select(OTU, Order, Family, Genus, Species) %>%
  left_join(unassigned_check_vert, by = "OTU") %>%
  left_join(depth_seqs, by = "depth") %>%
  select(OTU, Order = Order.x, Family = Family.x, Genus = Genus.x, Species = Species.x, Abundance, depth, sum_depth_seqs) %>%
  group_by(OTU, depth) %>%
  mutate(SumAbundance = sum(Abundance, na.rm = TRUE)) %>%
  reframe(
    Order = first(Order),
    Family = first(Family),
    Genus = first(Genus),
    Species = first(Species),
    RelativeAbundance = (first(SumAbundance) / first(sum_depth_seqs)) * 100
  )

head(depth_assigned_check)
dim(depth_assigned_check)

total_added <- depth_assigned_check %>% 
  group_by(depth) %>%
  summarize(totaladded = sum(RelativeAbundance))
total_added

total_added_bytaxa <- depth_assigned_check %>% 
  group_by(depth, Family) %>%
  summarize(taxa_added = sum(RelativeAbundance))
total_added_bytaxa

# for plotting:
head(unassigned_check_vert)
head(depth_assigned_check)

taxa_assigned_plot <- unassigned_check_vert %>%
  select(OTU, depth, Class, Order, Family, Abundance) %>%
  left_join(depth_assigned_check, by = c("OTU", "depth"))
head(taxa_assigned_plot)
tail(taxa_assigned_plot)

taxa_assigned_plot %>% select(OTU) %>% unique() # 345 ASVs (as expected)
taxa_assigned_plot1 <- taxa_assigned_plot %>%
  mutate(Order = case_when(
    is.na(Order.y) & Class == "Actinopterygii" ~ "Unassigned Actinopterygii",
    TRUE ~ Order.y)) %>%
  mutate(Family = case_when(
    is.na(Family.y) & Order.x == "Perciformes" ~ "Unassigned Perciformes",
    TRUE ~ Family.y)) %>%
  mutate(TAXA = case_when(
    is.na(Order.y) & Class == "Actinopterygii" & is.na(Family) ~ "Unassigned Actinopterygii",
    TRUE ~ Family)) %>%
  select(OTU, depth, Abundance, Class, Order, Family, Genus, Species, TAXA)

taxa_assigned_plot1$TAXA %>% unique() # taxa

taxa_assigned_plot2 <- taxa_assigned_plot1 %>%
  group_by(depth, TAXA) %>%
  summarize(SumAbundance = sum(Abundance)) %>%
  left_join(depth_seqs, by = "depth") %>%
  mutate(RelativeAbundance = (SumAbundance / sum_depth_seqs) * 100)

taxa_assigned_plot2$TAXA %>% unique() %>% sort()
head(taxa_assigned_plot2)

taxa_sorted <- c("Howellidae", #Acropomatiformes
                 "Bathylagidae", #Argentiniformes
                 "Melamphaidae", # Beryciformes
                 "Myctophidae", # Myctophiformes
                 "Gempylidae", # Scombriformes
                 "Gonostomatidae", "Sternoptychidae", # Stomiiformes
                 "Delphinidae", # Cetartiodactyla
                 "Unassigned Perciformes", "Unassigned Actinopterygii")
                            

taxa_colors <- c("#948ced",
                 "#fcada4",
                 "#a1f1d6",
                 "#f5eb8e",
                 "#116b4c",
                 "#56C4F0", 
                 "#367eff",
                 "#FF746C",
                 "#cddcf7", "#82A4E3")

assigned_fish <- taxa_assigned_plot2 %>%
    ggplot(aes(y = fct_rev(as.factor(depth)), x = RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + 
   scale_x_continuous(limits = c(0, 100)) + theme_custom_paper()

assigned_fish

assigned_fish <- assigned_fish + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Vertebrates - After Assignment") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))

assigned_fish

head(unassigned_check_vert)
taxa_unassigned_plot <- unassigned_check_vert %>%
  mutate(TAXA = case_when(
    Class == "Actinopterygii" & is.na(Order) ~ "Unassigned Actinopterygii",
    Order %in% c("Perciformes") & is.na(Family) ~ "Unassigned Perciformes",
    Order %in% c("Argentiniformes") & is.na(Family) ~ "Unassigned Argentiniformes",
    Order %in% c("Cetartiodactyla") & is.na(Genus) ~ "Unassigned Cetartiodactyla",
    TRUE ~ Class)) %>%
  select(OTU, Abundance, depth, TAXA) %>%
  group_by(depth, TAXA) %>%
  summarize(SumAbundance = sum(Abundance)) %>%
  left_join(depth_seqs, by = "depth") %>%
  mutate(RelativeAbundance = (SumAbundance / sum_depth_seqs) * 100)

taxa_unassigned_plot$TAXA %>% unique() %>% sort()

taxa_sorted <- c("Unassigned Argentiniformes",
                 "Unassigned Cetartiodactyla",
                 "Unassigned Perciformes", 
                 "Unassigned Actinopterygii")
                            

taxa_colors <- c("#fcada4",
                 "#FF746C",
                 "#cddcf7", "#82A4E3")


unassigned_fish <- taxa_unassigned_plot %>%
    ggplot(aes(y = fct_rev(as.factor(depth)), x = RelativeAbundance)) +
    geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + 
   scale_x_continuous(limits = c(0, 100)) + theme_custom_paper()

unassigned_fish

unassigned_fish <- unassigned_fish + labs(x = "Relative Abundance (%)",
       y = "Depth (m)", 
       title = "Vertebrates - Unassigned") + 
  theme(legend.position = "right") +
  guides(fill = guide_legend(title = NULL,
                             ncol = 1,
                             override.aes = list(stroke = 1)))

unassigned_fish

unassigned_fish + assigned_fish &
  #plot_layout(guides = 'collect') &
  theme(plot.margin = unit(c(t = 0.5, r = 0.5, b = 0.5, l = 0.5), "lines"),
        panel.spacing = unit(5, "mm"),
        panel.background = element_rect(linewidth = 0.5),
        plot.title = element_text(size = 10),
        plot.tag = element_text(size = 8),
        legend.text = element_text(size = 8, margin = margin(l = 2)),
        legend.position = "right", 
        legend.spacing.x = unit(4, "mm"), 
        legend.margin = margin(t = 0, r = 0, b = 50, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 8))

ggsave("./plots-final/vert_unassigned_relabun_ponel_barplot_v2.pdf", height = 3, width = 7.5, units = "in", dpi=300)

(unassigned_met + assigned_met) / (unassigned_fish + assigned_fish) &
  plot_annotation(tag_levels = 'A') & 
  #plot_layout(guides = 'collect') &
  theme(plot.margin = unit(c(t = 0.5, r = 0.5, b = 0.5, l = 0.5), "lines"),
        panel.spacing = unit(5, "mm"),
        panel.background = element_rect(linewidth = 0.5),
        plot.title = element_text(size = 10),
        plot.tag = element_text(size = 8),
        legend.text = element_text(size = 8, margin = margin(l = 2)),
        legend.position = "right", 
        legend.spacing.x = unit(4, "mm"), 
        legend.margin = margin(t = 0, r = 0, b = 50, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 8))

ggsave("./plots-final/all_metazoan_unassigned_relabun_ponel_barplot_v2.pdf", height = 5.5, width = 7.5, units = "in", dpi=300)


# combined

panel2 <- prot_relabun_agg + met_relabun_agg + fish_relabun_agg + 
#  plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect') &
  theme(plot.margin = unit(c(t = 0.5, r = 1.5, b = 0.5, l = 1), "lines"),
        panel.spacing = unit(10, "mm"),
        panel.background = element_rect(linewidth = 0.5),
        plot.background = element_blank(),
        plot.title = element_text(size = 10, hjust = 0.5),
        plot.tag = element_text(size = 8),
        legend.text = element_text(size = 6, margin = margin(l = 2)),
        legend.position = "bottom", 
        legend.spacing.x = unit(4, "mm"), 
        legend.margin = margin(t = -4, r = 0, b = 0, l = 0), 
        legend.key.size = unit(2, 'mm'), 
        legend.key.height = unit(1, "mm"),
        legend.key.spacing.x = unit(1, "mm"),
        legend.key.spacing.y = unit(0.25, "mm"),
        axis.text = element_text(size = 6),
        axis.title = element_text(size = 8))

panel2

ggsave("./plots-final/alleuk_relabun_barplot_pub.pdf", height = 90, width = 180, units = "mm", dpi=300)

# 
# panel3 <- prot_relabun_deploy / met_relabun_deploy / fish_relabun_deploy + 
#   plot_annotation(tag_levels = "A", tag_suffix = '.') + 
#   plot_layout(guides = 'collect') &
#   theme(plot.margin = unit(c(t = 0.25, r = 0.25, b = 0.25, l = 0.25), "lines"),
#         panel.spacing = unit(6, "mm"),
#         panel.background = element_rect(linewidth = 0.5),
#         plot.title = element_text(size = 8),
#         plot.tag = element_text(size = 8),
#         legend.text = element_text(size = 6, margin = margin(l = 1)),
#         legend.position = "right", 
#         legend.margin = margin(t = 10, r = 20, b = 50, l = 0), 
#         legend.key.size = unit(2.5, 'mm'), 
#         legend.key.height = unit(2.5, "mm"),
#         legend.key.spacing.x = unit(1, "mm"),
#         legend.key.spacing.y = unit(0.5, "mm"),
#         axis.text = element_text(size = 6),
#         axis.title = element_text(size = 8))
# 
# panel3
# 
# #ggsave("./plots-final/alleuk_relabun_barplot_pub_deploy.pdf", height = 200, width = 180, units = "mm", dpi=300)


# let's calculate ASV count and relative abundance for all taxa, but this is especially for the "Other" groups

# Summarize the data
prot_taxa_asvs <- prot_taxa %>%
  group_by(TAXA) %>%
  summarise(
    Unique_OTUs = n_distinct(OTU))

# going to join this with the prot_taxa_abun_check

prot_taxa_summary <- left_join(prot_taxa_asvs, prot_taxa_abun_check, by = "TAXA") %>%
  filter(sum_RelativeAbundance < 1)

write.csv(prot_taxa_summary, file = "prot_taxa_summary_lowabund.csv")

# Summarize the data
met_taxa_asvs <- met_taxa_order2 %>%
  group_by(TAXA) %>%
  summarise(
    Unique_OTUs = n_distinct(OTU))

# going to join this with the met_taxa_abun_check

met_taxa_summary <- left_join(met_taxa_asvs, met_taxa_abun_check, by = "TAXA") %>%
  filter(sum_RelativeAbundance < 1)

write.csv(met_taxa_summary, file = "met_taxa_summary_lowabund.csv")

# Summarize the data
fish_taxa_asvs <- fish_taxa_order %>%
  group_by(TAXA) %>%
  summarise(
    Unique_OTUs = n_distinct(OTU))

# going to join this with the fish_taxa_abun_check

fish_taxa_summary <- left_join(fish_taxa_asvs, fish_taxa_abun_check, by = "TAXA") %>%
  filter(sum_RelativeAbundance < 1)

write.csv(fish_taxa_summary, file = "fish_taxa_summary_lowabund.csv")


```

```{r prot-top-asv-evenness, eval = TRUE, message = FALSE, warning = FALSE}

# Based on alpha diversity, there is an interesting dip in diversity indices at depths where there is high number of observed ASVs. It's likely that this is impacted by evenness e.g. some ASVs are relatively abundant compared to all other ASVs.

head(prot_taxa2)

# let's get the unique # of ASVs here
length(unique(prot_taxa2$OTU)) # 7046

head(prot_taxa2)

top_prot_df <- prot_taxa2 %>%
  group_by(depth) %>%
  mutate(RelativeAbundance = (Abundance / sum(Abundance)) * 100, na.rm = TRUE) %>%
  mutate(depth = as.factor(depth)) %>%
  ungroup() %>%
  group_by(depth, OTU) %>%
  mutate(sum_RelativeAbundance = sum(RelativeAbundance), na.rm = TRUE, .groups = "drop") %>%
  select(OTU, depth, sum_RelativeAbundance, TAXA) %>% unique()

taxa_interest <- c("Alveolata-Apicomplexa", "Alveolata-Dinoflagellates", "Alveolata-Syndiniales", "Chlorophyta", "Rhizaria-Radiolaria", "Stramenopiles-Diatoms", "Stramenopiles-MAST", "Stramenopiles-Other")

taxa_colors <- c("#a1c2fd", "#399dff", "#0344b7",
                 "#9CDC46",
                 "#7474ff",
                 "#009a4d","#1ec38c", "#90eebf", "transparent")

prot_dotplot_top <- top_prot_df %>% filter(TAXA %in% taxa_interest) %>%
  filter(sum_RelativeAbundance > 0.1) %>%
  mutate(TAXA_new = case_when(
    sum_RelativeAbundance < 1 ~ "Low Abundance < 1%",
    TRUE ~ TAXA
  ))

taxa_sorted <- c("Alveolata-Apicomplexa", "Alveolata-Dinoflagellates", "Alveolata-Syndiniales", "Chlorophyta", "Rhizaria-Radiolaria", "Stramenopiles-Diatoms", "Stramenopiles-MAST", "Stramenopiles-Other","Low Abundance < 1%")

prot_dotplot_top <- 
  prot_dotplot_top %>% arrange(depth, TAXA_new, sum_RelativeAbundance)
head(prot_dotplot_top)  

dotplot <- ggplot(prot_dotplot_top, aes(y = fct_rev(depth), (OTU))) + 
  geom_jitter(aes(size = sum_RelativeAbundance, color = TAXA, fill = TAXA_new, group = TAXA_new), shape = 21, alpha = 0.75, height = 0.2, stroke = 0.5) +
  labs(title = NULL,
       x = "ASV",
       y = "Depth (m)") + 
  scale_color_manual(values = taxa_colors,
                    limits = taxa_sorted) +
  scale_fill_manual(values = taxa_colors,
                    limits = taxa_sorted) +
  guides(color = "none",
    size = guide_legend(title = "Relative Abundance"),
         fill = guide_legend(title="Taxa",
                             order = 1,
                             override.aes = list(shape = 21, 
                                                 alpha = 1,
                                                 stroke = 0.5))) + 
  theme_custom_paper() + 
  theme(panel.background = element_rect(linewidth = 0.5),
        panel.grid.major = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.ticks.x = element_blank())

dotplot

#ggsave("./plots-final/jitter_dotplot_topASVs.pdf", height = 80, width = 180, units = "mm", dpi=300)

head(prot_dotplot_top)

stats <- top_prot_df %>% 
  filter(sum_RelativeAbundance >= 1) %>%
  group_by(depth, TAXA) %>% 
  mutate(n = n()) %>% select(-OTU, - sum_RelativeAbundance) %>% unique() %>%
  arrange(depth, TAXA) %>% print(n = 30)

stats1 <- top_prot_df %>% 
  filter(sum_RelativeAbundance >= 1) %>% filter(TAXA != "Other Protists") %>%
  group_by(depth, TAXA) %>% 
  mutate(part_abun = sum(sum_RelativeAbundance)) %>%
  mutate(n = n()) %>% select(-OTU, - sum_RelativeAbundance) %>% unique() %>%
  arrange(depth, TAXA) %>% print(n = 30)

stats1 %>% group_by(depth) %>%
  mutate(sum = sum(part_abun)) %>% 
  mutate(sum_n = sum(n)) %>%
  select(depth, sum, sum_n) %>% unique()

unique(stats1$TAXA)

taxa_sorted <- c("Alveolata-Apicomplexa", "Alveolata-Dinoflagellates", "Alveolata-Syndiniales", "Chlorophyta", "Rhizaria-Radiolaria", "Stramenopiles-Diatoms", "Stramenopiles-MAST", "Stramenopiles-Other")

taxa_colors <- c("#a1c2fd", "#399dff", "#0344b7",
                 "#9CDC46",
                 "#7474ff",
                 "#009a4d","#1ec38c", "#90eebf")


barplot1 <- stats1 %>% arrange(depth, desc(n)) %>%
  mutate(TAXA = fct_reorder(TAXA, n, .desc = FALSE)) %>%
  ggplot(aes(y = fct_rev(depth), x = n)) +
  geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Observed ASVs") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + theme_custom_paper() + 
  theme(legend.position = "none",
        panel.background = element_rect(linewidth = 0.5),
        panel.grid.major = element_blank(),
        axis.text.x = element_text(size = 8),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 8))

barplot1

barplot2 <- stats1 %>% arrange(depth, desc(part_abun)) %>%
  mutate(TAXA = fct_reorder(TAXA, part_abun, .desc = FALSE)) %>%
  ggplot(aes(y = fct_rev(depth), x = part_abun)) +
  geom_col(aes(fill = TAXA), color = "black", width = 0.75, linewidth = 0.25) +
    labs(title = NULL,
         y = "Depth (m)",
         x = "Relative Abundance (%)") +
    scale_fill_manual(values = taxa_colors,
                        limits = taxa_sorted) + theme_custom_paper() + 
  theme(legend.position = "none",
        panel.background = element_rect(linewidth = 0.5),
        panel.grid.major = element_blank(),
        axis.text.x = element_text(size = 8),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 8))

barplot2

dotplot / (barplot1 + barplot2) + plot_layout(guides = "collect") + plot_annotation(tag_levels = "A", tag_suffix = '.') &
    theme(
      plot.margin = unit(c(t = 0.5, r = 1.5, b = 0.5, l = 1.5), "lines"),
      panel.background = element_rect(linewidth = 0.5, fill = "transparent"),
      plot.background = element_rect(fill = "transparent"),
      panel.spacing = unit(10, "mm"),
      panel.grid.major = element_blank(),
      plot.tag = element_text(size = 8),
      legend.title = element_text(size = 8),
      legend.text = element_text(size = 8, margin = margin(l = 2)),
      legend.key.size = unit(3, 'mm'), 
      legend.key.height = unit(3, "mm"),
      axis.text.y = element_text(size = 10),
      axis.title = element_text(size = 10))

ggsave("./plots-final/evenness_prot_panel_pub.pdf", height = 120, width = 180, units = "mm", dpi=300)


```

## 3.2 Ordination
```{r pca, eval = TRUE, message = FALSE, warning = FALSE}

depth_colors <- c("#C3E9CE", "#60CEAC", "#3497A9",  "#395D9C", "#413D7B", "#2E1E3C")

# check dataframes
head(rare_prot)
dim(rare_prot)

length(unique(rare_prot$sample))

rare_prot_dist <- rare_prot %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>% column_to_rownames(var = "sample") %>% data.frame()

# check dataframes
head(rare_met)
dim(rare_met)

length(unique(rare_met$sample))

rare_met_dist <- rare_met %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>% column_to_rownames(var = "sample") %>% data.frame()

head(rare_fish)
dim(rare_fish)

length(unique(rare_fish$sample))

rare_fish_dist <- rare_fish %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>% column_to_rownames(var = "sample") %>% data.frame()


head(rare_prot_dist)
rare_prot_dist_clr <- clr_lite(rare_prot_dist, samples_are = "rows", method = "const", replicates = 1) #We are imputing zeroes using the 'const' method 
#Essentially, we replace zeroes with 65% of the next lowest value - see Lubbe et al 2021. 

head(rare_prot_dist_clr)
pca_prot_clr <- rda(rare_prot_dist_clr)

dist_prot_clr <- vegdist(rare_prot_dist_clr, method = "euclidean")

head(summary(pca_prot_clr))

unconstrained_eig <- pca_prot_clr$CA$eig/pca_prot_clr$tot.chi*100

expl_var <- unconstrained_eig
barplot(expl_var[1:20], col = rep('black', length (unconstrained_eig)),
         las = 2, ylab = '% variation')

# set env paraproters
rare_prot_env <- rare_prot %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>%  
  inner_join(., metadata_18s, by = "sample") %>%
  select(sample, depth, deployn, temp, oxy, sal, fluor) %>% 
  data.frame()
head(rare_prot_env)
dim(rare_prot_env)

prot.scrs <- as.data.frame(scores(pca_prot_clr, choices = c(1:5), display = "sites"))
prot.scrs
prot.scrs <- cbind(prot.scrs, Depth = rare_prot_env$depth)
prot.scrs <- cbind(prot.scrs, Deployment = rare_prot_env$deployn)

prot_pca_plot <- 
  ggplot(data = prot.scrs, aes(x = PC1, y = PC2)) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), size = 2, stroke = 0.5, alpha = 0.9) + 
  ylab(paste0('PC2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PC1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
    guides(fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1,
                                                 size = 2)),
         shape = guide_legend(title = "CTD Cast")) +
  theme_custom_paper()

prot_pca_plot

#ggsave("./plots-final/beta_diversity_pca_prot.pdf", height = 60, width = 80, units = "mm", dpi=300)

rare_met_dist_clr <- clr_lite(rare_met_dist, samples_are = "rows", method = "const", replicates = 1)
head(rare_met_dist_clr)
pca_met_clr <- rda(rare_met_dist_clr)

dist_met_clr <- vegdist(rare_met_dist_clr, method = "euclidean")

head(summary(pca_met_clr))

unconstrained_eig <- pca_met_clr$CA$eig/pca_met_clr$tot.chi*100

expl_var <- unconstrained_eig
barplot(expl_var[1:20], col = rep('black', length(unconstrained_eig)),
         las = 2, ylab = '% variation')

# set env parameters
rare_met_env <- rare_met %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>%  
  inner_join(., metadata_18s, by = "sample") %>%
  select(depth, deployn, temp, oxy, sal, fluor) %>% 
  data.frame()
head(rare_met_env)
dim(rare_met_env)

met.scrs <- as.data.frame(scores(pca_met_clr, choices = c(1:5), display = "sites"))
met.scrs
met.scrs <- cbind(met.scrs, Depth = rare_met_env$depth)
met.scrs <- cbind(met.scrs, Deployment = rare_met_env$deployn)

met_pca_plot <- 
  ggplot(data = met.scrs, aes(x = PC1, y = PC2)) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), size = 2, stroke = 0.5, alpha = 0.9) + 
  ylab(paste0('PC2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PC1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
    guides(fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1,
                                                 size = 2)),
         shape = guide_legend(title = "CTD Cast")) +
  theme_custom_paper()

met_pca_plot

#ggsave("./plots-final/beta_diversity_pca_met.pdf", height = 60, width = 80, units = "mm", dpi=300)


rare_fish_dist_clr <- clr_lite(rare_fish_dist, samples_are = "rows", method = "const", replicates = 1)
head(rare_fish_dist_clr)
pca_fish_clr <- rda(rare_fish_dist_clr)

dist_fish_clr <- vegdist(rare_fish_dist_clr, method = "euclidean")

head(summary(pca_fish_clr))

unconstrained_eig <- pca_fish_clr$CA$eig/pca_fish_clr$tot.chi*100

expl_var <- unconstrained_eig
barplot (expl_var[1:20], col = rep('black', length (unconstrained_eig)),
         las = 2, ylab = '% variation')

# set env parameters
rare_fish_env <- rare_fish %>%
  select(sample, OTU, Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "Abundance", values_fill = 0) %>%  
  inner_join(., metadata_18s, by = "sample") %>%
  select(depth, deployn, temp, oxy, sal, fluor) %>% 
  data.frame()
head(rare_fish_env)
dim(rare_fish_env)

fish.scrs <- as.data.frame(scores(pca_fish_clr, choices = c(1:5), display = "sites"))
fish.scrs
fish.scrs <- cbind(fish.scrs, Depth = rare_fish_env$depth)
fish.scrs <- cbind(fish.scrs, Deployment = rare_fish_env$deployn)

head(fish.scrs)

fish_pca_plot <- 
  ggplot(data = fish.scrs, aes(x = PC1, y = PC2)) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), size = 2, stroke = 0.5, alpha = 0.9) + 
  ylab(paste0('PC2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PC1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
    guides(fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1,
                                                 size = 2)),
         shape = guide_legend(title = "CTD Cast")) + theme_custom_paper()

fish_pca_plot

fish_pca_plot_revised <- 
  ggplot(data = fish.scrs[-12,], aes(x = PC1, y = PC2)) +
  #ggplot(data = fish.scrs, aes(x = PC1, y = PC2)) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), size = 2, stroke = 0.5, alpha = 0.9) + 
  ylab(paste0('PC2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PC1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
    guides(fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1,
                                                 size = 2)),
         shape = guide_legend(title = "CTD Cast")) + theme_custom_paper()

fish_pca_plot_revised

#ggsave("./plots-final/beta_diversity_pca_fish_revised.pdf", height = 60, width = 80, units = "mm", dpi=300)

prot_pca_plot <- prot_pca_plot + ggtitle("Protists (18S)")
met_pca_plot <- met_pca_plot + ggtitle("Invertebrates (18S)")
fish_pca_plot <- fish_pca_plot + ggtitle("Vertebrates (12S)")
fish_pca_plot_revised <- fish_pca_plot_revised + ggtitle("Vertebrates (12S)")

panel <- prot_pca_plot + met_pca_plot + fish_pca_plot + 
#  plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect') & 
    theme(panel.border = element_rect(linewidth = 0.75),
        legend.position = "right", 
        legend.spacing.y = unit(0, "mm"), 
        legend.margin = margin(3, 3, 3, 3), 
        legend.key.size = unit(3, 'mm'),
        plot.title = element_text(size = 10, hjust = 0.5),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 7),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 7))

panel

ggsave("./plots-final/beta_diversity_pca_all.pdf", height = 62, width = 180, units = "mm", dpi=300)

panel <- prot_pca_plot + met_pca_plot + fish_pca_plot_revised + 
#  plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect') & 
  theme(panel.border = element_rect(linewidth = 0.75),
        legend.position = "right", 
        legend.spacing.y = unit(0, "mm"), 
        legend.margin = margin(3, 3, 3, 3), 
        legend.key.size = unit(3, 'mm'),
        plot.title = element_text(size = 10, hjust = 0.5),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 7),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 7))

panel

ggsave("./plots-final/beta_diversity_pca_all_revised.pdf", height = 62, width = 180, units = "mm", dpi=300)


```

## 3.3 PERMANOVA

```{r beta-stats-euc-adonis, eval = TRUE, message = FALSE, warning = FALSE}

## metazoans
set.seed(100)

# Adonis test

# note: The R-square value is the important statistic for interpreting the effect size. It is the percent variation that is explained by the variable and the p value indicates significance.

# extract rownames
sample_name <- as.data.frame(unique(rare_prot$sample))
colnames(sample_name) <- "sample"
head(sample_name)
class(sample_name)

prot_sd <- left_join(sample_name, metadata_18s, by = "sample")
head(prot_sd)
dim(prot_sd)

adonis_prot <- adonis2(dist_prot_clr ~ depth, data = prot_sd, permutations = 999)
adonis_prot

posthoc_prot <- pairwise.adonis2(dist_prot_clr ~ depth, data = prot_sd, nperm = 999)
posthoc_prot

adonis_prot <- adonis2(dist_prot_clr ~ deployment, data = prot_sd, permutations = 999)
adonis_prot

posthoc_prot <- pairwise.adonis2(dist_prot_clr ~ deployment, data = prot_sd, nperm = 999)
posthoc_prot

# extract rownames
sample_name <- as.data.frame(unique(rare_met$sample))
colnames(sample_name) <- "sample"
head(sample_name)
class(sample_name)
head(metadata_18s)

met_sd <- left_join(sample_name, metadata_18s, by = "sample")
head(met_sd)
dim(met_sd)

adonis_met <- adonis2(dist_met_clr ~ depth, data = met_sd, permutations = 999)
adonis_met
posthoc_met <- pairwise.adonis2(dist_met_clr ~ depth, data = met_sd, nperm = 999)
posthoc_met
adonis_met <- adonis2(dist_met_clr ~ deployment, data = met_sd, permutations = 999)
adonis_met
posthoc_met <- pairwise.adonis2(dist_met_clr ~ deployment, data = met_sd, nperm = 999)
posthoc_met

# extract rownames
sample_name <- as.data.frame(unique(rare_fish$sample))
colnames(sample_name) <- "sample"
head(sample_name)
class(sample_name)

fish_sd <- left_join(sample_name, metadata_12s, by = "sample")
head(fish_sd)
dim(fish_sd)

adonis_fish <- adonis2(dist_fish_clr ~ depth * deployment, data = fish_sd, permutations = 999)
adonis_fish
adonis_fish <- adonis2(dist_fish_clr ~ depth, data = fish_sd, permutations = 999)
adonis_fish
adonis_fish <- adonis2(dist_fish_clr ~ deployment, data = fish_sd, permutations = 999)
adonis_fish

posthoc_fish <- pairwise.adonis2(dist_fish_clr ~ depth*deployment, data = fish_sd, nperm = 999)
posthoc_fish
posthoc_fish <- pairwise.adonis2(dist_fish_clr ~ depth, data = fish_sd, nperm = 999)
posthoc_fish
posthoc_fish <- pairwise.adonis2(dist_fish_clr ~ deployment, data = fish_sd, nperm = 999)
posthoc_fish
##
```

```{r beta-stats-euc-betadispr, eval = TRUE, message = FALSE, warning = FALSE}

# In ecological studies, you might use betadisper to determine whether the variability in species composition is similar across different habitat types before running PERMANOVA to compare the centroids of these habitats.

# when results are significant, at least one group (depth) has significantly different dispersion compared to the others, which suggests that depth affects the variability of community composition. However, plotting these will show that group centroids are far apart (e.g. for protists and metazoans) and we can be more confident in our permanova results for group centroids.

#Protists 
# depth
bd_prot <- betadisper(dist_prot_clr, prot_sd$depth, type = "median")
anova(bd_prot)
permutest(bd_prot, pairwise = TRUE)

plot(bd_prot, hull = FALSE, ellipse = TRUE)

bd_prot_scrs <- as.data.frame(scores(bd_prot, choices = c(1:5), display = "sites"))
bd_prot_scrs
bd_prot_scrs <- cbind(bd_prot_scrs, Depth = rare_prot_env$depth)
bd_prot_scrs <- cbind(bd_prot_scrs, Deployment = rare_prot_env$deployn)

prot_bd_plot <- 
  ggplot(data = bd_prot_scrs, aes(x = PCoA1, y = PCoA2, color = as.factor(Depth), fill = as.factor(Depth))) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), color = "black", size = 4, stroke = 1, alpha = 0.9) + 
  stat_ellipse(type = "t") + 
  ylab(paste0('PCoA2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PCoA1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
  scale_color_manual(values = depth_colors) + 
  guides(color = "none",
         fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1)),
         shape = guide_legend(title = "CTD Cast")) +
  theme_custom_paper()

prot_bd_plot

boxplot(bd_prot)

#Metazoa 
# depth
bd_met <- betadisper(dist_met_clr, met_sd$depth, type = "median")
anova(bd_met)
permutest(bd_met, pairwise = TRUE)

plot(bd_met, hull = FALSE, ellipse = TRUE)

bd_met_scrs <- as.data.frame(scores(bd_met, choices = c(1:5), display = "sites"))
bd_met_scrs
bd_met_scrs <- cbind(bd_met_scrs, Depth = rare_met_env$depth)
bd_met_scrs <- cbind(bd_met_scrs, Deployment = rare_met_env$deployn)

met_bd_plot <- 
  ggplot(data = bd_met_scrs, aes(x = PCoA1, y = PCoA2, color = as.factor(Depth))) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), color = "black", size = 4, stroke = 1, alpha = 0.9) + 
  ylab(paste0('PCoA2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PCoA1 ', round(unconstrained_eig[1],2),'%')) +
  labs(title = NULL) + 
  stat_ellipse(type = "t") + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_color_manual(values = depth_colors) + 
  scale_fill_manual(values = depth_colors) + 
  guides(color = "none",
         fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1)),
         shape = guide_legend(title = "CTD Cast")) +
  theme_custom_paper()

met_bd_plot

#Fish 
# depth
bd_fish <- betadisper(dist_fish_clr, fish_sd$depth, type = "median")
anova(bd_fish)
permutest(bd_fish, pairwise = TRUE)

plot(bd_fish, hull = FALSE, ellipse = TRUE)

bd_fish_scrs <- as.data.frame(scores(bd_fish, choices = c(1:5), display = "sites"))
bd_fish_scrs
bd_fish_scrs <- cbind(bd_fish_scrs, Depth = rare_fish_env$depth)
bd_fish_scrs <- cbind(bd_fish_scrs, Deployment = rare_fish_env$deployn)

fish_bd_plot <- 
  ggplot(data = bd_fish_scrs, aes(x = PCoA1, y = PCoA2, color = as.factor(Depth))) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, color = "#525252", linetype = "dotted") +
  geom_point(aes(fill = as.factor(Depth), shape = as.factor(Deployment)), color = "black", size = 4, stroke = 1, alpha = 0.9) + 
  ylab(paste0('PCoA2 ', round(unconstrained_eig[2],2),'%')) +
  xlab(paste0('PCoA1 ', round(unconstrained_eig[1],2),'%')) +
  stat_ellipse(type = "t") + 
  labs(title = NULL) + 
  scale_shape_manual(values = c(21, 23, 24)) + 
  scale_fill_manual(values = depth_colors) + 
  scale_color_manual(values = depth_colors) + 
  guides(color = "none",
         fill = guide_legend(title="Depth (m)",
                             override.aes = list(shape = 21,
                                                 alpha = 1)),
         shape = guide_legend(title = "CTD Cast")) +
  theme_custom_paper()

fish_bd_plot

```

## 3.4 Euler
```{r beta-stats-euler, eval = TRUE, message = FALSE, warning = FALSE}

prot_taxa2$TAXA <- fct_relevel(prot_taxa2$TAXA, "Other Protists", after = Inf)
taxa_sorted <- c("Alveolata-Apicomplexa", 
                 "Alveolata-Ciliates",
                 "Alveolata-Dinoflagellates", 
                 # "Alveolata-Other", 
                 "Alveolata-Syndiniales", 
                 "Chlorophyta", 
                 #  "Cryptophyta", 
                 "Excavata",
                 "Haptophyta", #"Rhizaria-Cercozoa", 
                 # "Rhizaria-Other", 
                 #"Obazoa-Opisthokonta",
                 "Rhizaria-Radiolaria", # "Rhodophyta",
                 "Stramenopiles-Diatoms", "Stramenopiles-MAST", "Stramenopiles-Other", "Other Protists")


rare_prot_euler <- prot_taxa2
unique(rare_prot_euler$TAXA)

# let's annotate this with taxa from prot_taxa2
rare_prot_euler <- rare_prot_euler %>% select(depth, OTU, Abundance) %>%
  group_by(depth, OTU) %>%
  summarize(total = sum(Abundance)) %>%
  ungroup() %>%
  pivot_wider(names_from = "depth", values_from = "total", values_fill = 0) %>% column_to_rownames(var = "OTU") %>% data.frame() %>%
  mutate_all(~ ifelse(. > 0, 1, .))

head(rare_prot_euler)
# let's add back columns
rare_prot_euler$OTU <- rownames(rare_prot_euler)
head(rare_prot_euler)

head(prot_taxa2)
prot_taxonomy <- prot_taxa2 %>% select(OTU, TAXA) %>% unique()
dim(prot_taxonomy)
head(prot_taxonomy)

# let's combine these
rare_prot_euler <- left_join(rare_prot_euler, prot_taxonomy, by = "OTU") %>% column_to_rownames(var = "OTU") %>% data.frame()
head(rare_prot_euler)

colnames(rare_prot_euler) <- c("10m", "100m", "300m", "500m", "800m", "1000m", "Taxa")
rare_prot_euler <- rare_prot_euler %>% select("1000m", "800m", "500m", "300m", "100m", "10m", "Taxa")
head(rare_prot_euler)
depths <- colnames(rare_prot_euler)[1:6]
depths

taxa_colors <- c("#a1c2fd", "#56C4F0", "#399dff",
                 #"#0b62fb", 
                 "#0344b7",
                 "#9CDC46",#"#dafa0f",
                 "#efdf48", "#ff9aa5",
                 #"#ceceff", "#a7a7ff",
                 #"#ff9d3a",
                 "#7474ff",
                 #"#ff9aa5",
                 "#009a4d","#1ec38c", "#90eebf",#"#82A4E3", 
                 "#d9d9d9")

prot_upset <- ComplexUpset::upset(rare_prot_euler, depths, 
                                  stripes = "white",
                    sort_sets = "FALSE",
                    min_size = 50,
                    name = "", 
                    height_ratio = 0.7,
                    set_sizes = FALSE,
                    wrap = TRUE,
                    matrix=(intersection_matrix(
                      geom=geom_point(
                        size=1),
                      segment=geom_segment(
                        linewidth=0.25),
                      outline_color=list(
                        active='black',
                        inactive='grey70'))),
                    base_annotations=list(
                      "Intersection Size" = intersection_size(counts = FALSE,
                                                              mapping = aes(fill = Taxa), 
                                                              color = "black", linewidth = 0.1) + 
                        scale_fill_manual(values = taxa_colors) + 
                        ylab("# of ASVs") + xlab(NULL)),
                    themes = list(default = theme_custom_upset(), 
                                  intersections_matrix = theme_custom_upset()))

prot_upset

#ggsave("./plots-final/prot_upset.pdf", plot = prot_upset, height = 80, width = 180, units = "mm", dpi=300)


met_taxa_order3$TAXA <- fct_relevel(met_taxa_order3$TAXA, "Arthropoda-Calanoida", "Arthropoda-Cyclopoida",
                                    #"Arthropoda-Other",
                             "Cnidaria-Siphonophorae",
                             #"Cnidaria-Other", 
                             "Ctenophora-Cydippida", 
                             "Tunicata-Copelata", 
                             "Tunicata-Salpida", 
                             #"Unassigned Metazoa", 
                             "Other Invertebrates")

taxa_sorted <- c("Arthropoda-Calanoida", "Arthropoda-Cyclopoida", 
                 #"Arthropoda-Other",
                             "Cnidaria-Siphonophorae",
                             #"Cnidaria-Other", 
                             "Ctenophora-Cydippida", 
                             "Tunicata-Copelata", 
                             "Tunicata-Salpida", 
                 #"Unassigned Metazoa", 
                 "Other Invertebrates")

rare_met_euler <- met_taxa_order3
unique(rare_met_euler$TAXA)

# let's annotate this with taxa from met_taxa2
rare_met_euler <- rare_met_euler %>% select(depth, OTU, Abundance) %>%
  group_by(depth, OTU) %>%
  summarize(total = sum(Abundance)) %>%
  ungroup() %>%
  pivot_wider(names_from = "depth", values_from = "total", values_fill = 0) %>% column_to_rownames(var = "OTU") %>% data.frame() %>%
  mutate_all(~ ifelse(. > 0, 1, .))

head(rare_met_euler)
# let's add back columns
rare_met_euler$OTU <- rownames(rare_met_euler)
head(rare_met_euler)

head(met_taxa_order3)
met_taxonomy <- met_taxa_order3 %>% select(OTU, TAXA) %>% unique()
dim(met_taxonomy)
head(met_taxonomy)

# let's combine these
rare_met_euler <- left_join(rare_met_euler, met_taxonomy, by = "OTU") %>% column_to_rownames(var = "OTU") %>% data.frame()
head(rare_met_euler)

colnames(rare_met_euler) <- c("10m", "100m", "300m", "500m", "800m", "1000m", "Taxa")
rare_met_euler <- rare_met_euler %>% select("1000m", "800m", "500m", "300m", "100m", "10m", "Taxa")
head(rare_met_euler)
depths <- colnames(rare_met_euler)[1:6]
depths

taxa_colors <- c("#fc996c", "#ffce5c","#fdfb9b", 
                 "#047de9", "#0344b7", 
                 "#fa9db8",
                 "#96fac6","#26a33a",  "#82A4E3",
                 "#d9d9d9")


taxa_colors <- c("#fc996c", "#ffce5c",
                 #"#fdfb9b", 
                 "#047de9", 
                 #"#0344b7", 
                 "#fa9db8", 
                 #"#fc5d7b",
                 "#96fac6","#1ec38c", 
                  #"#82A4E3", 
                 "#d9d9d9")


met_upset <- ComplexUpset::upset(rare_met_euler, depths, 
                                  stripes = "white",
                    sort_sets = "FALSE",
                    min_size = 3,
                    name = "", 
                    height_ratio = 0.7,
                    set_sizes = FALSE,
                    wrap = TRUE,
                    matrix=(intersection_matrix(
                      geom=geom_point(
                        size=1),
                      segment=geom_segment(
                        linewidth=0.25),
                      outline_color=list(
                        active='black',
                        inactive='grey70'))),
                    base_annotations=list(
                      "Intersection Size" = intersection_size(counts = FALSE,
                                                              mapping = aes(fill = Taxa), 
                                                              color = "black", linewidth = 0.1) + 
                        scale_fill_manual(values = taxa_colors,
                                          limits = taxa_sorted) + 
                        ylab("# of ASVs") + xlab(NULL)),
                    themes = list(default = theme_custom_upset(), 
                                  intersections_matrix = theme_custom_upset()))

met_upset

#ggsave("./plots-final/met_upset.pdf", plot = met_upset, height = 80, width = 180, units = "mm", dpi=300)


fish_taxa_order3 <- fish_taxa_order2 %>%
  mutate(TAXA = case_when(
    TAXA == "Other Vertebrates" ~ "Other Vertebrates         ",
    TRUE ~ TAXA))


fish_taxa_order3$TAXA <- fct_relevel(fish_taxa_order3$TAXA, 
                                     "Alepocephalidae", 
                                     "Nemichthyidae", "Serrivomeridae",
                                     "Melamphaidae", 
                                     "Myctophidae", 
                                     "Gonostomatidae", "Sternoptychidae", "Stomiidae", 
                                     "Molidae", "Other Vertebrates         ")


taxa_sorted <- c("Alepocephalidae", 
                 "Nemichthyidae", "Serrivomeridae",
                 "Melamphaidae", 
                 "Myctophidae", 
                 "Gonostomatidae", "Sternoptychidae", "Stomiidae", 
                 "Molidae", 
                 "Other Vertebrates         ")

rare_fish_euler <- fish_taxa_order3
unique(rare_fish_euler$TAXA)

# let's annotate this with taxa from fish_taxa2
rare_fish_euler <- rare_fish_euler %>% select(depth, OTU, Abundance) %>%
  group_by(depth, OTU) %>%
  summarize(total = sum(Abundance)) %>%
  ungroup() %>%
  pivot_wider(names_from = "depth", values_from = "total", values_fill = 0) %>% column_to_rownames(var = "OTU") %>% data.frame() %>%
  mutate_all(~ ifelse(. > 0, 1, .))

head(rare_fish_euler)
# let's add back columns
rare_fish_euler$OTU <- rownames(rare_fish_euler)
head(rare_fish_euler)

head(fish_taxa_order3)
fish_taxonomy <- fish_taxa_order3 %>% select(OTU, TAXA) %>% unique()
dim(fish_taxonomy)
head(fish_taxonomy)

# let's combine these
rare_fish_euler <- left_join(rare_fish_euler, fish_taxonomy, by = "OTU") %>% column_to_rownames(var = "OTU") %>% data.frame()
head(rare_fish_euler)

colnames(rare_fish_euler) <- c("10m", "100m", "300m", "500m", "800m", "1000m", "Taxa")
rare_fish_euler <- rare_fish_euler %>% select("1000m", "800m", "500m", "300m", "100m", "10m", "Taxa")
head(rare_fish_euler)
depths <- colnames(rare_fish_euler)[1:6]
depths

taxa_colors <- c("#9CDC46", "#d5d2f8", "#9790ee", 
                "#a1f1d6","#f5eb8e",
                  "#56C4F0", "#0c62fb", "#034dd1","#fb7285", "#d9d9d9")


fish_upset <- ComplexUpset::upset(rare_fish_euler, depths, 
                                  stripes = "white",
                    sort_sets = "FALSE",
                    min_size = 1,
                    name = "", 
                    height_ratio = 0.7,
                    set_sizes = FALSE,
                    wrap = TRUE,
                    matrix=(intersection_matrix(
                      geom=geom_point(
                        size=1),
                      segment=geom_segment(
                        linewidth=0.25),
                      outline_color=list(
                        active='black',
                        inactive='grey70'))),
                    base_annotations=list(
                      "Intersection Size" = intersection_size(counts = FALSE,
                                                              mapping = aes(fill = Taxa), 
                                                              color = "black", linewidth = 0.1) + 
                        scale_fill_manual(values = taxa_colors) + 
                        ylab("# of ASVs") + xlab(NULL)),
                    themes = list(default = theme_custom_upset(), 
                                  intersections_matrix = theme_custom_upset()))
fish_upset

#ggsave("./plots-final/fish_upset.pdf", plot = fish_upset, height = 30, width = 80, units = "mm", dpi=300)


## Combined
prot_upset / met_upset / fish_upset +
  plot_annotation(tag_levels = "A", tag_suffix = '.') + 
  plot_layout(guides = 'collect')
  
ggsave("./plots-final/all_upset.pdf", height = 190, width = 180, units = "mm", dpi=300)

```

```{r save-euler-inputs, eval = TRUE, message = FALSE, warning = FALSE}

save(rare_prot_euler, rare_met_euler, rare_fish_euler, file = "./euler_inputs.RData")

load(file = "./euler_inputs.RData", verbose = T)

```

```{r save-euler-analyses, eval = TRUE, message = FALSE, warning = FALSE}

# clean up dataframes
head(rare_prot_euler)
colnames(rare_prot_euler)
prot_euler_df <- rare_prot_euler
prot_euler_df$ASV <- rownames(prot_euler_df)
rownames(prot_euler_df) <- NULL
prot_euler_df <- prot_euler_df %>% select(ASV, Taxa, "10m", "100m", "300m", "500m", "800m", "1000m")
head(prot_euler_df)

# clean up dataframes
head(rare_met_euler)
colnames(rare_met_euler)
met_euler_df <- rare_met_euler
met_euler_df$ASV <- rownames(met_euler_df)
rownames(met_euler_df) <- NULL
met_euler_df <- met_euler_df %>% select(ASV, Taxa, "10m", "100m", "300m", "500m", "800m", "1000m")
head(met_euler_df)

# clean up dataframes
head(rare_fish_euler)
colnames(rare_fish_euler)
fish_euler_df <- rare_fish_euler
fish_euler_df$ASV <- rownames(fish_euler_df)
rownames(fish_euler_df) <- NULL
fish_euler_df <- fish_euler_df %>% select(ASV, Taxa, "10m", "100m", "300m", "500m", "800m", "1000m")
head(fish_euler_df)

# let's extract the ASVs we think are performing DVM
head(prot_euler_df)
# let's fix column names
colnames(prot_euler_df)[c(1, 3:8)] <- c("OTU", "d10", "d100", "d300", "d500", "d800", "d1000")
prot_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
prot_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
prot_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 1 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
prot_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 1 & d800 == 0 & d1000 == 0) %>% count()
prot_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 1 & d1000 == 0) %>% count()
prot_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 1) %>% count()

prot_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1) %>% count()

355 + 752 + 801 + 750 + 657 + 673 # 3988 unique ASVs
prot_euler_df %>% count() # 7046
3988 / 7046 * 100 # 56.6%%
7046 - 3988 # 3058 shared (including cosmopolitan)
3058 / 7046 * 100 #43.4% (shared including cosmopolitan) 
(3058-140) / 7046 * 100 # 41.4% are shared (140 is cosmo)
140 / 7046 * 100 # 2.0% (cosmopolitan)

# extract ASV & TAXA
#dvm_prot_asv <- dvm_prot_df %>% select(OTU, Taxa)
#head(dvm_prot_asv)

# let's extract the ASVs we think are performing DVM
head(met_euler_df)
# let's fix column names
colnames(met_euler_df)[c(1, 3:8)] <- c("OTU", "d10", "d100", "d300", "d500", "d800", "d1000")
met_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
met_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
met_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 1 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
met_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 1 & d800 == 0 & d1000 == 0) %>% count()
met_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 1 & d1000 == 0) %>% count()
met_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 1) %>% count()

met_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1) %>% count()

19 + 30 + 37 + 34 + 55 + 98 # 273 unique ASVs
met_euler_df %>% count() # 399
273 / 399 * 100 # 68.4%
399 - 273 # 126 shared
126 / 399 * 100 #31.6%
(126-14) / 399 * 100 # 28.1% are shared
14 / 399 * 100 # 3.5%

dvm_met_df1 <- met_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_met_df2 <- met_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_met_df <- rbind(dvm_met_df1, dvm_met_df2) %>% unique()
# extract ASV & TAXA
dvm_met_asv <- dvm_met_df %>% select(OTU, Taxa)
head(dvm_met_asv)

# create relative abundance plot
head(met_taxa_order3)
dvm_met_relabun <- left_join(dvm_met_asv, met_taxa_order3, by = "OTU")
head(dvm_met_relabun)

cast_colors <- c("#3399ff", "#ffb400", "#9080ff")
diel_colors <- c("#FFC857", "#4B3F72")  

dvm_met_relabun <- dvm_met_relabun %>%
  filter(Order != "Unassigned") %>% select(OTU, depth, diel, TAXA, Abundance)
  
dvm_met_taxa <- dvm_met_relabun %>% group_by(depth, OTU) %>%
  mutate(RelativeAbundance = (Abundance / sum(Abundance)) * 100, na.rm = TRUE) %>%
  ungroup() %>%
#  group_by(depth, deployment, OTU) %>%
  group_by(depth, diel, OTU) %>%
  summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "keep") %>%
  ungroup()

# let's get a list of OTUs & TAXA
met_asv_taxa <- met_taxa_order3 %>% select(OTU, TAXA, Abundance) %>% as.data.frame() %>%
  group_by(OTU) %>%
  mutate(total_Abundance = sum(Abundance)) %>%
  select(-Abundance) %>% 
  distinct() %>%
  mutate(read_abundance = paste0("n = ", total_Abundance))
  
head(met_asv_taxa)
dvm_met_plot_df <- left_join(dvm_met_taxa, met_asv_taxa, by = "OTU")
head(dvm_met_plot_df)

dvm_met_plot <- dvm_met_plot_df %>% 
  filter(TAXA != "Other Invertebrates") %>%
  ggplot(aes(y = depth, x = sum_RelativeAbundance)) +
  geom_point(aes(fill = diel, color = diel), alpha = 0.66, shape = 21, size = 1.5, stroke = 0.25) + 
  geom_path(aes(group = diel, color = diel), alpha = 0.66, linewidth = 1) +
  scale_y_reverse() + 
  scale_color_manual(values = diel_colors) + 
  scale_fill_manual(values = diel_colors) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)",
       title = NULL) +
  facet_wrap(~ TAXA + OTU + read_abundance, nrow = 5) + 
  theme_custom_paper() +
  theme(plot.title = element_text(size = 8),
        plot.margin = unit(c(t = 0, r = 1, b = -1, l = 1), "mm"),
        plot.title.position = "panel",
        plot.background = element_rect(linewidth = 0.5),
        panel.background = element_rect(linewidth = 0.5),
        panel.spacing.x = unit(2, "mm"),
        panel.spacing.y = unit(1, "mm"),
        legend.position = "bottom", 
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 10, margin = margin(l = 2)),
        legend.key.size = unit(4, 'mm'), 
        strip.text.x = element_text(size = 7),
        strip.background = element_rect(color = NULL),
        strip.clip = "off",
        strip.text = element_text(margin=margin(t = 1.5, b= 1.5)),
        strip.background.x = element_blank(),
        axis.text = element_text(hjust = 1, vjust = 1, size = 6),
        axis.title = element_text(size = 10)) + 
  guides(fill = "none",
         color = guide_legend(title = "Diel")) 

dvm_met_plot

ggsave("./plots-final/alleuk_met_dvm_pub.pdf", height = 9, width = 7.5, units = "in", dpi=300)

# let's extract the ASVs we think are performing DVM
head(fish_euler_df)
# let's fix column names
colnames(fish_euler_df)[c(1, 3:8)] <- c("OTU", "d10", "d100", "d300", "d500", "d800", "d1000")

fish_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
fish_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
fish_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 1 & d500 == 0 & d800 == 0 & d1000 == 0) %>% count()
fish_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 1 & d800 == 0 & d1000 == 0) %>% count()
fish_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 1 & d1000 == 0) %>% count()
fish_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 0 & d500 == 0 & d800 == 0 & d1000 == 1) %>% count()

fish_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1) %>% count()

139 + 54 + 83 + 177 + 61 + 81 # 595 unique ASVs
fish_euler_df %>% count() # 315 / 668
595 / 667 * 100 # 89.2% unique out of the whole dataset
667 - 595 # 73 shared (including cosmo)
72 / 667 * 100 # 10.8%
(72-2) / 667 * 100 # 10.5% are shared (2 = cosmopolitan)
2 / 667 * 100 # 0.30%

dvm_fish_df1 <- fish_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 1 & d500 == 1 & d800 == 0 & d1000 == 0)
dvm_fish_df2 <- fish_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_fish_df3 <- fish_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 0)
dvm_fish_df4 <- fish_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 0 & d1000 == 0)
dvm_fish_df5 <- fish_euler_df %>% filter(d10 == 1 & d100 == 0 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_fish_df6 <- fish_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_fish_df7 <- fish_euler_df %>% filter(d10 == 1 & d100 == 1 & d300 == 0 & d500 == 1 & d800 == 1 & d1000 == 1)
dvm_fish_df8 <- fish_euler_df %>% filter(d10 == 0 & d100 == 1 & d300 == 0 & d500 == 1 & d800 == 0 & d1000 == 1)
dvm_fish_df9 <- fish_euler_df %>% filter(d10 == 0 & d100 == 0 & d300 == 1 & d500 == 1 & d800 == 1 & d1000 == 1)


dvm_fish_df <- rbind(dvm_fish_df2, dvm_fish_df3, dvm_fish_df4, dvm_fish_df5, dvm_fish_df6, dvm_fish_df7, dvm_fish_df8, dvm_fish_df9) %>% unique()
head(dvm_fish_df)


# extract ASV & TAXA
dvm_fish_asv <- dvm_fish_df %>% select(OTU, Taxa) %>% filter(Taxa != "Other Vertebrates         ")
head(dvm_fish_asv)

# create relative abundance plot
head(fish_taxa_order)
dvm_fish_relabun <- left_join(dvm_fish_asv, fish_taxa_order, by = "OTU")
head(dvm_fish_relabun)
  
dvm_fish_taxa <- dvm_fish_relabun %>% group_by(depth, OTU) %>%
  mutate(RelativeAbundance = (Abundance / sum(Abundance)) * 100, na.rm = TRUE) %>%
  ungroup() %>%
  group_by(depth, diel, OTU) %>%
  summarise(sum_RelativeAbundance = sum(RelativeAbundance, na.rm = TRUE), .groups = "drop") %>%
  ungroup()

# let's get a list of OTUs & TAXA
fish_asv_taxa <- fish_taxa_order %>% select(OTU, Species, TAXA, Abundance) %>% as.data.frame() %>%
  group_by(OTU) %>%
  mutate(total_Abundance = sum(Abundance)) %>%
  select(-Abundance) %>% 
  distinct() %>%
  mutate(read_abundance = paste0("n = ", total_Abundance))

head(fish_asv_taxa)
dvm_fish_plot_df <- left_join(dvm_fish_taxa, fish_asv_taxa, by = "OTU")
head(dvm_fish_plot_df)

dvm_fish_plot_df <- dvm_fish_plot_df %>%
  mutate(Species = case_when(
    is.na(Species) ~ TAXA,
    TRUE ~ Species
  ))

dvm_fish_plot <- dvm_fish_plot_df %>% 
  ggplot(aes(y = depth, x = sum_RelativeAbundance)) +
  geom_point(aes(fill = diel, color = diel), alpha = 0.66, shape = 21, size = 1.5, stroke = 0.25) + 
  geom_path(aes(group = diel, color = diel), alpha = 0.66, linewidth = 1) +
  scale_y_reverse() + 
  scale_color_manual(values = diel_colors) + 
  scale_fill_manual(values = diel_colors) + 
  labs(x = "Relative Abundance (%)",
       y = "Depth (m)",
       title = NULL) +
  facet_wrap(~ Species + OTU + read_abundance, nrow = 3) + 
  theme_custom_paper() +
  theme(plot.title = element_text(size = 8),
        plot.margin = unit(c(t = 0, r = 1, b = -1, l = 1), "mm"),
        plot.title.position = "panel",
        plot.background = element_rect(linewidth = 0.5),
        panel.background = element_rect(linewidth = 0.5),
        panel.spacing.x = unit(2, "mm"),
        panel.spacing.y = unit(1, "mm"),
        legend.position = "bottom", 
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 8, margin = margin(l = 2)),
        legend.key.size = unit(4, 'mm'), 
        strip.text.x = element_text(size = 7),
        strip.background.x = element_blank(),
        strip.clip = "off",
        strip.text=element_text(margin=margin(t = 1.5, b= 1.5)),
        axis.text = element_text(hjust = 1, vjust = 1, size = 6),
        axis.title = element_text(size = 8)) + 
  guides(fill = "none",
         color = guide_legend(title = "Diel")) 

dvm_fish_plot

ggsave("./plots-final/alleuk_fish_dvm.pdf", height = 8, width = 6, units = "in", dpi=300)

```


